// @microsoft/teamsfx-core/build/plugins/resource/identity/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityBicepFile = exports.IdentityBicep = exports.Telemetry = exports.Constants = void 0;
class Constants {
}
exports.Constants = Constants;
Constants.pluginName = "Identity Plugin";
Constants.pluginNameShort = "msi";
Constants.prefix = "teamsfx";
Constants.apiVersion = "2018-11-30";
Constants.deployName = "user-assigned-identity";
Constants.identityName = "identityName";
Constants.identityClientId = "identityClientId";
Constants.identityResourceId = "identityResourceId";
Constants.solution = "solution";
Constants.subscriptionId = "subscriptionId";
Constants.resourceGroupName = "resourceGroupName";
Constants.resourceNameSuffix = "resourceNameSuffix";
Constants.location = "location";
Constants.remoteTeamsAppId = "remoteTeamsAppId";
Constants.resourceProvider = "Microsoft.ManagedIdentity";
class Telemetry {
}
exports.Telemetry = Telemetry;
Telemetry.componentName = "fx-resource-azure-identity";
Telemetry.startSuffix = "-start";
Telemetry.valueYes = "yes";
Telemetry.valueNo = "no";
Telemetry.userError = "user";
Telemetry.systemError = "system";
Telemetry.stage = {
    generateArmTemplates: "generate-arm-templates",
    updateArmTemplates: "update-arm-templates",
};
Telemetry.properties = {
    component: "component",
    success: "success",
    errorCode: "error-code",
    errorType: "error-type",
    errorMessage: "error-message",
    appid: "appid",
};
class IdentityBicep {
}
exports.IdentityBicep = IdentityBicep;
IdentityBicep.identityName = "provisionOutputs.identityOutput.value.identityName";
IdentityBicep.identityClientId = "provisionOutputs.identityOutput.value.identityClientId";
IdentityBicep.identityResourceId = "userAssignedIdentityProvision.outputs.identityResourceId";
IdentityBicep.identityPrincipalId = "userAssignedIdentityProvision.outputs.identityPrincipalId";
class IdentityBicepFile {
}
exports.IdentityBicepFile = IdentityBicepFile;
IdentityBicepFile.moduleTempalteFilename = "identityProvision.template.bicep";
