// @microsoft/teamsfx-core/build/component/resource/aadApp/utils.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertContext = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const migrate_1 = require("../../migrate");
function convertContext(context, inputs) {
    var _a, _b;
    const projectSetting = migrate_1.convertProjectSettingsV3ToV2(context.projectSetting);
    const stateV2 = migrate_1.convertEnvStateV3ToV2(context.envInfo.state);
    (_a = stateV2["fx-resource-aad-app-for-teams"]) !== null && _a !== void 0 ? _a : (stateV2["fx-resource-aad-app-for-teams"] = {});
    const value = teamsfx_api_1.ConfigMap.fromJSON(stateV2);
    const pluginCtx = {
        cryptoProvider: context.cryptoProvider,
        config: new teamsfx_api_1.ConfigMap(),
        logProvider: context.logProvider,
        m365TokenProvider: (_b = context.tokenProvider) === null || _b === void 0 ? void 0 : _b.m365TokenProvider,
        ui: context.userInteraction,
        projectSettings: projectSetting,
        permissionRequestProvider: context.permissionRequestProvider,
        root: inputs.projectPath,
        envInfo: {
            config: {
                manifest: {
                    appName: {
                        short: context.projectSetting.appName,
                    },
                },
            },
            envName: inputs.env,
            state: value,
        },
    };
    return pluginCtx;
}
exports.convertContext = convertContext;
