// @microsoft/teamsfx-core/build/plugins/resource/keyvault/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Telemetry = exports.Constants = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
class Constants {
}
exports.Constants = Constants;
Constants.provisionModuleTemplateFileName = "keyVaultProvision.template.bicep";
Constants.configModuleTemplateFileName = "keyVaultConfiguration.template.bicep";
Constants.SolutionPlugin = {
    id: "solution",
    configKeys: {
        remoteTeamsAppId: "remoteTeamsAppId",
    },
};
Constants.KeyVaultPlugin = {
    pluginName: "fx-resource-key-vault",
    displayName: "Key Vault Plugin",
    shortName: "kv",
};
Constants.KeyVaultBicep = {
    m365ClientSecretReference: "provisionOutputs.keyVaultOutput.value.m365ClientSecretReference",
    botClientSecretReference: "provisionOutputs.keyVaultOutput.value.botClientSecretReference",
};
Constants.Stage = {
    generateArmTemplates: "generate-arm-templates",
    updateArmTemplates: "update-arm-templates",
};
class Telemetry {
}
exports.Telemetry = Telemetry;
Telemetry.component = "component";
Telemetry.errorCode = "error-code";
Telemetry.errorType = "error-type";
Telemetry.errorMessage = "error-message";
Telemetry.userError = "user";
Telemetry.systemError = "system";
Telemetry.isSuccess = "success";
Telemetry.success = "yes";
Telemetry.fail = "no";
Telemetry.appId = "appid";
