// @microsoft/teamsfx-core/build/plugins/resource/identity/v2/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityPluginV2 = void 0;
const tslib_1 = require("tslib");
const typedi_1 = require("typedi");
const __1 = require("..");
const ResourcePluginContainer_1 = require("../../../solution/fx-solution/ResourcePluginContainer");
const utils4v2_1 = require("../../utils4v2");
let IdentityPluginV2 = class IdentityPluginV2 {
    constructor() {
        this.name = "fx-resource-identity";
        this.displayName = "Microsoft Identity";
    }
    activate(projectSettings) {
        const solutionSettings = projectSettings.solutionSettings;
        return this.plugin.activate(solutionSettings);
    }
    async provisionResource(ctx, inputs, envInfo, tokenProvider) {
        return await utils4v2_1.provisionResourceAdapter(ctx, inputs, envInfo, tokenProvider, this.plugin);
    }
    async generateResourceTemplate(ctx, inputs) {
        return utils4v2_1.generateResourceTemplateAdapter(ctx, inputs, this.plugin);
    }
    async updateResourceTemplate(ctx, inputs) {
        return utils4v2_1.updateResourceTemplateAdapter(ctx, inputs, this.plugin);
    }
};
tslib_1.__decorate([
    typedi_1.Inject(ResourcePluginContainer_1.ResourcePlugins.IdentityPlugin),
    tslib_1.__metadata("design:type", __1.IdentityPlugin)
], IdentityPluginV2.prototype, "plugin", void 0);
IdentityPluginV2 = tslib_1.__decorate([
    typedi_1.Service(ResourcePluginContainer_1.ResourcePluginsV2.IdentityPlugin)
], IdentityPluginV2);
exports.IdentityPluginV2 = IdentityPluginV2;
