// @microsoft/teamsfx-core/build/plugins/resource/bot/resources/strings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationTriggers = exports.QuestionBotScenarioToBotCapability = exports.QuestionBotScenarioToPluginActRoles = exports.BotCapabilities = exports.Commands = exports.ConfigNames = exports.PluginBot = exports.PluginFunction = exports.PluginIdentity = exports.PluginSql = exports.PluginSolution = exports.PluginLocalDebug = exports.PluginAAD = exports.CommonStrings = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const localizeUtils_1 = require("../../../../common/localizeUtils");
const question_1 = require("../../../solution/fx-solution/question");
const pluginActRoles_1 = require("../enums/pluginActRoles");
class CommonStrings {
}
exports.CommonStrings = CommonStrings;
CommonStrings.BOT_WORKING_DIR_NAME = "bot";
CommonStrings.AAD_APP = localizeUtils_1.getLocalizedString("plugins.bot.AadApp");
CommonStrings.AAD_CLIENT_SECRET = localizeUtils_1.getLocalizedString("plugins.bot.AadClientSecret");
CommonStrings.APP_STUDIO_BOT_REGISTRATION = localizeUtils_1.getLocalizedString("plugins.bot.AppStudioBotRegistration");
CommonStrings.SHORT_APP_NAME = "short app name";
CommonStrings.AUTH_REDIRECT_URI_SUFFIX = "/public";
CommonStrings.MESSAGE_ENDPOINT_SUFFIX = "/api/messages";
class PluginAAD {
}
exports.PluginAAD = PluginAAD;
PluginAAD.PLUGIN_NAME = "fx-resource-aad-app-for-teams";
PluginAAD.CLIENT_ID = "clientId";
PluginAAD.CLIENT_SECRET = "clientSecret";
PluginAAD.APPLICATION_ID_URIS = "applicationIdUris";
PluginAAD.OAUTH_AUTHORITY = "oauthAuthority";
PluginAAD.TENANT_ID = "tenantId";
class PluginLocalDebug {
}
exports.PluginLocalDebug = PluginLocalDebug;
PluginLocalDebug.PLUGIN_NAME = "fx-resource-local-debug";
PluginLocalDebug.LOCAL_BOT_ENDPOINT = "localBotEndpoint";
PluginLocalDebug.LOCAL_DEBUG_SUFFIX = "-local-debug";
class PluginSolution {
}
exports.PluginSolution = PluginSolution;
PluginSolution.PLUGIN_NAME = "solution";
PluginSolution.SUBSCRIPTION_ID = "subscriptionId";
PluginSolution.RESOURCE_GROUP_NAME = "resourceGroupName";
PluginSolution.LOCATION = "location";
PluginSolution.M365_TENANT_ID = "teamsAppTenantId";
PluginSolution.RESOURCE_NAME_SUFFIX = "resourceNameSuffix";
PluginSolution.REMOTE_TEAMS_APPID = "remoteTeamsAppId";
class PluginSql {
}
exports.PluginSql = PluginSql;
PluginSql.PLUGIN_NAME = "fx-resource-azure-sql";
PluginSql.SQL_ENDPOINT = "sqlEndpoint";
PluginSql.SQL_DATABASE_NAME = "databaseName";
PluginSql.SQL_USERNAME = "sqlUsername";
PluginSql.SQL_PASSWORD = "sqlPassword";
class PluginIdentity {
}
exports.PluginIdentity = PluginIdentity;
PluginIdentity.PLUGIN_NAME = "fx-resource-identity";
PluginIdentity.IDENTITY_ClIENT_ID = "identityClientId";
PluginIdentity.IDENTITY_RESOURCE_ID = "identityResourceId";
class PluginFunction {
}
exports.PluginFunction = PluginFunction;
PluginFunction.PLUGIN_NAME = "fx-resource-function";
PluginFunction.ENDPOINT = "functionEndpoint";
class PluginBot {
}
exports.PluginBot = PluginBot;
PluginBot.PLUGIN_NAME = "fx-resource-bot";
PluginBot.BOT_ID = "botId";
PluginBot.BOT_PASSWORD = "botPassword";
PluginBot.OBJECT_ID = "objectId";
PluginBot.PROGRAMMING_LANGUAGE = "programmingLanguage";
PluginBot.APP_SERVICE_PLAN = "appServicePlan";
PluginBot.SITE_NAME = "siteName";
PluginBot.SKU_NAME = "skuName";
PluginBot.SITE_ENDPOINT = "siteEndpoint";
PluginBot.VALID_DOMAIN = "validDomain";
PluginBot.BOT_CHANNEL_REGISTRATION = "botChannelReg";
PluginBot.BOT_WEB_APP_RESOURCE_ID = "botWebAppResourceId";
PluginBot.RESOURCE_ID = "resourceId";
PluginBot.UNPACK_FLAG = "unPackFlag";
PluginBot.HOST_TYPE = "host-type";
// Bot capabilities, for example: notification, command-and-response.
// Don't mix up with Teams capabilities (tab, bot, etc.)
PluginBot.BOT_CAPABILITIES = "capabilities";
class ConfigNames {
}
exports.ConfigNames = ConfigNames;
ConfigNames.PROGRAMMING_LANGUAGE = "programming language";
ConfigNames.GRAPH_TOKEN = "graph token";
ConfigNames.SUBSCRIPTION_ID = "subscription id";
ConfigNames.ENV = "env";
ConfigNames.BOT_SERVICE_RESOURCE_ID = "bot service resource id";
ConfigNames.RESOURCE_GROUP = "resource group";
ConfigNames.LOCAL_ENDPOINT = "local endpoint";
ConfigNames.AUTH_CLIENT_ID = "auth client id";
ConfigNames.AUTH_CLIENT_SECRET = "auth client secret";
ConfigNames.SITE_ENDPOINT = "site endpoint";
ConfigNames.BOT_ID = "bot id";
ConfigNames.BOT_PASSWORD = "bot password";
ConfigNames.LOCAL_BOT_ID = "local bot id";
ConfigNames.APPSTUDIO_TOKEN = "app studio token";
ConfigNames.MESSAGE_ENDPOINT = "message endpoint";
class Commands {
}
exports.Commands = Commands;
Commands.NPM_INSTALL = "npm install";
Commands.NPM_BUILD = "npm run build";
exports.BotCapabilities = {
    NOTIFICATION: "notification",
    COMMAND_AND_RESPONSE: "command-response",
    WORKFLOW: "workflow",
    BOT: "bot",
    MESSAGE_EXTENSION: "message-extension",
    M365_SEARCH_APP: "m365-search-app",
};
exports.QuestionBotScenarioToPluginActRoles = new Map([
    [question_1.BotScenario.NotificationBot, pluginActRoles_1.PluginActRoles.Notification],
    [question_1.BotScenario.CommandAndResponseBot, pluginActRoles_1.PluginActRoles.CommandAndResponse],
    [question_1.BotScenario.WorkflowBot, pluginActRoles_1.PluginActRoles.Workflow],
]);
exports.QuestionBotScenarioToBotCapability = new Map([
    [question_1.BotScenario.NotificationBot, exports.BotCapabilities.NOTIFICATION],
    [question_1.BotScenario.CommandAndResponseBot, exports.BotCapabilities.COMMAND_AND_RESPONSE],
    [question_1.BotScenario.WorkflowBot, exports.BotCapabilities.WORKFLOW],
]);
exports.NotificationTriggers = {
    HTTP: "http",
    TIMER: "timer",
};
