// @microsoft/teamsfx-core/build/plugins/resource/function/utils/telemetry-helper.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetryHelper = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("../constants");
const enums_1 = require("../enums");
class TelemetryHelper {
    static setContext(ctx) {
        this.ctx = ctx;
    }
    static fillCommonProperty(properties) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        properties[enums_1.TelemetryKey.Component] = constants_1.FunctionPluginInfo.pluginName;
        if (((_b = (_a = this.ctx) === null || _a === void 0 ? void 0 : _a.envInfo) === null || _b === void 0 ? void 0 : _b.state) instanceof Map) {
            properties[enums_1.TelemetryKey.AppId] =
                ((_e = (_d = (_c = this.ctx) === null || _c === void 0 ? void 0 : _c.envInfo) === null || _d === void 0 ? void 0 : _d.state.get(constants_1.DependentPluginInfo.solutionPluginName)) === null || _e === void 0 ? void 0 : _e.get(constants_1.DependentPluginInfo.remoteTeamsAppId)) || constants_1.CommonConstants.emptyString;
        }
        else {
            properties[enums_1.TelemetryKey.AppId] =
                ((_j = (_h = (_g = (_f = this.ctx) === null || _f === void 0 ? void 0 : _f.envInfo) === null || _g === void 0 ? void 0 : _g.state) === null || _h === void 0 ? void 0 : _h.solution) === null || _j === void 0 ? void 0 : _j[constants_1.DependentPluginInfo.remoteTeamsAppId]) ||
                    constants_1.CommonConstants.emptyString;
        }
    }
    static sendStartEvent(eventName, properties = {}, measurements = {}) {
        var _a, _b;
        this.fillCommonProperty(properties);
        (_b = (_a = this.ctx) === null || _a === void 0 ? void 0 : _a.telemetryReporter) === null || _b === void 0 ? void 0 : _b.sendTelemetryEvent(`${eventName}-start`, properties, measurements);
    }
    static sendSuccessEvent(eventName, properties = {}, measurements = {}) {
        var _a, _b;
        this.fillCommonProperty(properties);
        properties[enums_1.TelemetryKey.Success] = enums_1.TelemetryValue.Success;
        (_b = (_a = this.ctx) === null || _a === void 0 ? void 0 : _a.telemetryReporter) === null || _b === void 0 ? void 0 : _b.sendTelemetryEvent(eventName, properties, measurements);
    }
    static sendErrorEvent(eventName, e, properties = {}, measurements = {}) {
        var _a, _b;
        this.fillCommonProperty(properties);
        properties[enums_1.TelemetryKey.Success] = enums_1.TelemetryValue.Fail;
        properties[enums_1.TelemetryKey.ErrorMessage] = e.message;
        properties[enums_1.TelemetryKey.ErrorCode] = e.name;
        if (e instanceof teamsfx_api_1.SystemError) {
            properties[enums_1.TelemetryKey.ErrorType] = enums_1.TelemetryValue.SystemError;
        }
        else if (e instanceof teamsfx_api_1.UserError) {
            properties[enums_1.TelemetryKey.ErrorType] = enums_1.TelemetryValue.UserError;
        }
        (_b = (_a = this.ctx) === null || _a === void 0 ? void 0 : _a.telemetryReporter) === null || _b === void 0 ? void 0 : _b.sendTelemetryErrorEvent(eventName, properties, measurements, [
            enums_1.TelemetryKey.ErrorMessage,
        ]);
    }
    static sendResultEvent(eventName, result, properties = {}, measurements = {}) {
        result.match(() => this.sendSuccessEvent(eventName, properties, measurements), (e) => this.sendErrorEvent(eventName, e, properties, measurements));
    }
    static sendScaffoldFallbackEvent(message, properties = {}, measurements = {}) {
        var _a, _b;
        this.fillCommonProperty(properties);
        properties[enums_1.TelemetryKey.ErrorMessage] = message;
        (_b = (_a = this.ctx) === null || _a === void 0 ? void 0 : _a.telemetryReporter) === null || _b === void 0 ? void 0 : _b.sendTelemetryEvent(enums_1.FunctionEvent.scaffoldFallback, properties, measurements);
    }
    static sendGeneralEvent(eventName, properties = {}, measurements = {}) {
        var _a, _b;
        this.fillCommonProperty(properties);
        (_b = (_a = this.ctx) === null || _a === void 0 ? void 0 : _a.telemetryReporter) === null || _b === void 0 ? void 0 : _b.sendTelemetryEvent(eventName, properties, measurements);
    }
}
exports.TelemetryHelper = TelemetryHelper;
