// @microsoft/teamsfx-core/build/plugins/resource/frontend/dotnet/enum.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceType = exports.DotnetConfigKey = void 0;
var DotnetConfigKey;
(function (DotnetConfigKey) {
    /* Config from solution */
    DotnetConfigKey["resourceGroupName"] = "resourceGroupName";
    DotnetConfigKey["subscriptionId"] = "subscriptionId";
    DotnetConfigKey["resourceNameSuffix"] = "resourceNameSuffix";
    DotnetConfigKey["location"] = "location";
    DotnetConfigKey["credential"] = "credential";
    DotnetConfigKey["teamsAppName"] = "teamsAppName";
    /* Config exported by Dotnet plugin */
    DotnetConfigKey["webAppName"] = "webAppName";
    DotnetConfigKey["appServicePlanName"] = "appServicePlanName";
    DotnetConfigKey["webAppEndpoint"] = "webAppEndpoint";
    DotnetConfigKey["webAppDomain"] = "webAppDomain";
    DotnetConfigKey["projectFilePath"] = "projectFilePath";
    DotnetConfigKey["webAppResourceId"] = "webAppResourceId";
    /* Intermediate */
    DotnetConfigKey["site"] = "site";
})(DotnetConfigKey = exports.DotnetConfigKey || (exports.DotnetConfigKey = {}));
var ResourceType;
(function (ResourceType) {
    ResourceType["webApp"] = "Azure Web App";
    ResourceType["appServicePlan"] = "Azure App Service Plan";
})(ResourceType = exports.ResourceType || (exports.ResourceType = {}));
