// @microsoft/teamsfx-core/build/core/SolutionPluginContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSolutionPluginByName = exports.getSolutionPluginV2ByName = exports.getAllSolutionPlugins = exports.getAllSolutionPluginsV2 = exports.SolutionPluginsV2 = exports.SolutionPlugins = void 0;
require("reflect-metadata");
const typedi_1 = require("typedi");
exports.SolutionPlugins = {
    AzureTeamsSolution: "AzureTeamsSolution",
};
exports.SolutionPluginsV2 = {
    AzureTeamsSolutionV2: "AzureTeamsSolutionV2",
};
/**
 * @returns all registered resource plugins V2
 */
function getAllSolutionPluginsV2() {
    const plugins = [];
    for (const k in exports.SolutionPluginsV2) {
        const plugin = typedi_1.Container.get(k);
        if (plugin) {
            plugins.push(plugin);
        }
    }
    return plugins;
}
exports.getAllSolutionPluginsV2 = getAllSolutionPluginsV2;
/**
 * @returns all registered resource plugins
 */
function getAllSolutionPlugins() {
    const plugins = [];
    for (const k in exports.SolutionPlugins) {
        const plugin = typedi_1.Container.get(k);
        if (plugin) {
            plugins.push(plugin);
        }
    }
    return plugins;
}
exports.getAllSolutionPlugins = getAllSolutionPlugins;
function getSolutionPluginV2ByName(name) {
    const solutions = getAllSolutionPluginsV2().filter((s) => s.name === name);
    if (solutions.length > 0)
        return solutions[0];
    return undefined;
}
exports.getSolutionPluginV2ByName = getSolutionPluginV2ByName;
function getSolutionPluginByName(name) {
    const solutions = getAllSolutionPlugins().filter((s) => s.name === name);
    if (solutions.length > 0)
        return solutions[0];
    return undefined;
}
exports.getSolutionPluginByName = getSolutionPluginByName;
