// @microsoft/teamsfx-core/build/plugins/resource/function/v2/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionPluginV2 = void 0;
const tslib_1 = require("tslib");
const typedi_1 = require("typedi");
const __1 = require("../..");
const ResourcePluginContainer_1 = require("../../../solution/fx-solution/ResourcePluginContainer");
const utils4v2_1 = require("../../utils4v2");
let FunctionPluginV2 = class FunctionPluginV2 {
    constructor() {
        this.name = "fx-resource-function";
        this.displayName = "Azure Function";
    }
    activate(projectSettings) {
        const solutionSettings = projectSettings.solutionSettings;
        return this.plugin.activate(solutionSettings);
    }
    async scaffoldSourceCode(ctx, inputs) {
        return await utils4v2_1.scaffoldSourceCodeAdapter(ctx, inputs, this.plugin);
    }
    async generateResourceTemplate(ctx, inputs) {
        return await utils4v2_1.generateResourceTemplateAdapter(ctx, inputs, this.plugin);
    }
    async updateResourceTemplate(ctx, inputs) {
        return await utils4v2_1.updateResourceTemplateAdapter(ctx, inputs, this.plugin);
    }
    async provisionResource(ctx, inputs, envInfo, tokenProvider) {
        return await utils4v2_1.provisionResourceAdapter(ctx, inputs, envInfo, tokenProvider, this.plugin);
    }
    async configureResource(ctx, inputs, envInfo, tokenProvider) {
        return await utils4v2_1.configureResourceAdapter(ctx, inputs, envInfo, tokenProvider, this.plugin);
    }
    async deploy(ctx, inputs, envInfo, tokenProvider) {
        return await utils4v2_1.deployAdapter(ctx, inputs, envInfo, tokenProvider, this.plugin);
    }
    async executeUserTask(ctx, inputs, func, localSettings, envInfo, tokenProvider) {
        return await utils4v2_1.executeUserTaskAdapter(ctx, inputs, func, localSettings, envInfo, tokenProvider, this.plugin);
    }
    async getQuestionsForUserTask(ctx, inputs, func, envInfo, tokenProvider) {
        return await utils4v2_1.getQuestionsForUserTaskAdapter(ctx, inputs, func, envInfo, tokenProvider, this.plugin);
    }
};
tslib_1.__decorate([
    typedi_1.Inject(ResourcePluginContainer_1.ResourcePlugins.FunctionPlugin),
    tslib_1.__metadata("design:type", __1.FunctionPlugin)
], FunctionPluginV2.prototype, "plugin", void 0);
FunctionPluginV2 = tslib_1.__decorate([
    typedi_1.Service(ResourcePluginContainer_1.ResourcePluginsV2.FunctionPlugin)
], FunctionPluginV2);
exports.FunctionPluginV2 = FunctionPluginV2;
