// @microsoft/teamsfx-core/build/plugins/resource/function/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionBicepFile = exports.FunctionArmOutput = exports.FunctionBicep = exports.QuestionValidationFunc = exports.Commands = exports.AzureInfo = exports.DefaultProvisionConfigs = exports.FunctionAppSettingKeys = exports.DependentPluginInfo = exports.DefaultValues = exports.RegularExpr = exports.FunctionPluginPathInfo = exports.FunctionPluginInfo = exports.CommonConstants = void 0;
const tools_1 = require("../../../common/tools");
const enums_1 = require("./enums");
class CommonConstants {
}
exports.CommonConstants = CommonConstants;
CommonConstants.emptyString = "";
CommonConstants.versionSep = ".";
CommonConstants.msInOneSecond = 1000;
CommonConstants.zipTimeMSGranularity = 2 * CommonConstants.msInOneSecond;
CommonConstants.latestTrustMtime = new Date(2000, 1, 1).getTime();
class FunctionPluginInfo {
}
exports.FunctionPluginInfo = FunctionPluginInfo;
FunctionPluginInfo.alias = "BE";
FunctionPluginInfo.pluginName = "fx-resource-function";
FunctionPluginInfo.displayName = "Function";
FunctionPluginInfo.expectDotnetSDKs = ["3.1", "5.0"];
FunctionPluginInfo.FunctionPluginPersistentConfig = [
    enums_1.FunctionConfigKey.functionAppName,
    enums_1.FunctionConfigKey.storageAccountName,
    enums_1.FunctionConfigKey.appServicePlanName,
    enums_1.FunctionConfigKey.functionEndpoint,
    enums_1.FunctionConfigKey.functionAppResourceId,
];
FunctionPluginInfo.templateBaseGroupName = "function-base";
FunctionPluginInfo.templateBaseScenarioName = "default";
FunctionPluginInfo.templateTriggerGroupName = "function-triggers";
class FunctionPluginPathInfo {
}
exports.FunctionPluginPathInfo = FunctionPluginPathInfo;
FunctionPluginPathInfo.solutionFolderName = "api";
FunctionPluginPathInfo.templateZipExt = ".zip";
FunctionPluginPathInfo.templateZipNameSep = ".";
FunctionPluginPathInfo.functionExtensionsFolderName = "bin";
FunctionPluginPathInfo.functionExtensionsFileName = "extensions.csproj";
FunctionPluginPathInfo.funcDeploymentFolderName = ".deployment";
FunctionPluginPathInfo.funcDeploymentInfoFileName = "deployment.json";
FunctionPluginPathInfo.funcDeploymentZipCacheFileName = "deployment.zip";
FunctionPluginPathInfo.funcIgnoreFileName = ".funcignore";
FunctionPluginPathInfo.gitIgnoreFileName = ".gitignore";
FunctionPluginPathInfo.npmPackageFolderName = "node_modules";
class RegularExpr {
}
exports.RegularExpr = RegularExpr;
RegularExpr.validFunctionNamePattern = /^[a-zA-Z][\w-]{0,126}$/;
// See https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules.
RegularExpr.validAppServicePlanNamePattern = /^[a-zA-Z0-9\-]{1,40}$/;
RegularExpr.validFunctionAppNamePattern = /^[a-zA-Z0-9][a-zA-Z0-9\-]{0,58}[a-zA-Z0-9]$/;
RegularExpr.validStorageAccountNamePattern = /^[a-z0-9]{3,24}$/;
RegularExpr.validResourceSuffixPattern = /[0-9a-z]{1,16}/;
RegularExpr.allCharToBeSkippedInName = /[^a-zA-Z0-9]/g;
RegularExpr.replaceTemplateExtName = /\.tpl$/;
RegularExpr.replaceTemplateFileNamePlaceholder = /entryname/g;
class DefaultValues {
}
exports.DefaultValues = DefaultValues;
DefaultValues.helpLink = "";
DefaultValues.issueLink = "https://github.com/OfficeDev/TeamsFx/issues/new";
DefaultValues.functionName = "getUserProfile";
DefaultValues.functionLanguage = enums_1.FunctionLanguage.JavaScript;
DefaultValues.functionTriggerType = "HTTPTrigger";
DefaultValues.maxTryCount = 3;
DefaultValues.scaffoldTryCount = 1;
DefaultValues.scaffoldTimeoutInMs = 10 * 1000;
DefaultValues.deployTimeoutInMs = 10 * 60 * 1000;
class DependentPluginInfo {
}
exports.DependentPluginInfo = DependentPluginInfo;
DependentPluginInfo.solutionPluginFullName = "fx-solution-azure";
DependentPluginInfo.solutionPluginName = "solution";
DependentPluginInfo.resourceGroupName = "resourceGroupName";
DependentPluginInfo.subscriptionId = "subscriptionId";
DependentPluginInfo.resourceNameSuffix = "resourceNameSuffix";
DependentPluginInfo.location = "location";
DependentPluginInfo.programmingLanguage = "programmingLanguage";
DependentPluginInfo.remoteTeamsAppId = "remoteTeamsAppId";
DependentPluginInfo.aadPluginName = "fx-resource-aad-app-for-teams";
DependentPluginInfo.aadClientId = "clientId";
DependentPluginInfo.aadClientSecret = "clientSecret";
DependentPluginInfo.oauthHost = "oauthHost";
DependentPluginInfo.tenantId = "tenantId";
DependentPluginInfo.applicationIdUris = "applicationIdUris";
DependentPluginInfo.sqlPluginName = "fx-resource-azure-sql";
DependentPluginInfo.databaseName = "databaseName";
DependentPluginInfo.sqlEndpoint = "sqlEndpoint";
DependentPluginInfo.identityPluginName = "fx-resource-identity";
DependentPluginInfo.identityClientId = "identityClientId";
DependentPluginInfo.identityResourceId = "identityResourceId";
DependentPluginInfo.frontendPluginName = "fx-resource-frontend-hosting";
DependentPluginInfo.frontendEndpoint = "endpoint";
DependentPluginInfo.frontendDomain = "domain";
DependentPluginInfo.apimPluginName = "fx-resource-apim";
DependentPluginInfo.apimAppId = "apimClientAADClientId";
class FunctionAppSettingKeys {
}
exports.FunctionAppSettingKeys = FunctionAppSettingKeys;
FunctionAppSettingKeys.clientId = "M365_CLIENT_ID";
FunctionAppSettingKeys.clientSecret = "M365_CLIENT_SECRET";
FunctionAppSettingKeys.oauthHost = "M365_AUTHORITY_HOST";
FunctionAppSettingKeys.tenantId = "M365_TENANT_ID";
FunctionAppSettingKeys.identityId = "IDENTITY_ID";
FunctionAppSettingKeys.databaseName = "SQL_DATABASE_NAME";
FunctionAppSettingKeys.sqlEndpoint = "SQL_ENDPOINT";
FunctionAppSettingKeys.allowedAppIds = "ALLOWED_APP_IDS";
FunctionAppSettingKeys.functionEndpoint = "API_ENDPOINT";
FunctionAppSettingKeys.applicationIdUris = "M365_APPLICATION_ID_URI";
class DefaultProvisionConfigs {
}
exports.DefaultProvisionConfigs = DefaultProvisionConfigs;
DefaultProvisionConfigs.allowAppIdSep = ";";
DefaultProvisionConfigs.nameSuffix = "be";
DefaultProvisionConfigs.siteIdentityTypeUserAssigned = "UserAssigned";
DefaultProvisionConfigs.appServicePlansConfig = (location) => ({
    location: location,
    kind: "functionapp",
    sku: {
        name: "Y1",
        tier: "Dynamic",
        size: "Y1",
        family: "Y",
        capacity: 0,
    },
});
DefaultProvisionConfigs.functionAppStaticSettings = {
    ALLOWED_APP_IDS: tools_1.getAllowedAppIds().join(";"),
    FUNCTIONS_EXTENSION_VERSION: "~3",
    WEBSITE_RUN_FROM_PACKAGE: "1",
};
DefaultProvisionConfigs.storageConfig = (location) => ({
    sku: {
        name: "Standard_LRS",
        tier: "Standard",
    },
    kind: "StorageV2",
    location: location,
    enableHttpsTrafficOnly: true,
});
DefaultProvisionConfigs.functionAppConfig = (location) => ({
    kind: "functionapp",
    location: location,
    clientAffinityEnabled: false,
});
class AzureInfo {
}
exports.AzureInfo = AzureInfo;
AzureInfo.resourceNameLenMax = 24;
AzureInfo.suffixLenMax = 12;
AzureInfo.zipDeployURL = (functionAppName) => `https://${functionAppName}.scm.azurewebsites.net/api/zipdeploy`;
AzureInfo.runFromPackageSettingKey = "WEBSITE_RUN_FROM_PACKAGE";
AzureInfo.runFromPackageEnabled = "1";
AzureInfo.requiredResourceProviders = ["Microsoft.Web", "Microsoft.Storage"];
class Commands {
}
exports.Commands = Commands;
Commands.npmInstall = "npm install";
Commands.npmInstallProd = "npm install --only=prod";
Commands.npmBuild = "npm run build";
class QuestionValidationFunc {
}
exports.QuestionValidationFunc = QuestionValidationFunc;
QuestionValidationFunc.validateFunctionName = "validateFunctionName";
class FunctionBicep {
}
exports.FunctionBicep = FunctionBicep;
FunctionBicep.functionEndpoint = "provisionOutputs.functionOutput.value.functionEndpoint";
FunctionBicep.functionAppResourceId = "provisionOutputs.functionOutput.value.functionAppResourceId";
class FunctionArmOutput {
}
exports.FunctionArmOutput = FunctionArmOutput;
FunctionArmOutput.Endpoint = "function_functionEndpoint";
FunctionArmOutput.AppResourceId = "function_appResourceId";
class FunctionBicepFile {
}
exports.FunctionBicepFile = FunctionBicepFile;
FunctionBicepFile.provisionModuleTemplateFileName = "functionProvision.template.bicep";
FunctionBicepFile.configuraitonTemplateFileName = "functionConfiguration.template.bicep";
