// @microsoft/teamsfx-core/build/plugins/resource/function/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomizedTask = exports.FunctionEvent = exports.TelemetryValue = exports.TelemetryKey = exports.ResourceType = exports.QuestionKey = exports.FunctionConfigKey = exports.FunctionLanguage = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
var FunctionLanguage;
(function (FunctionLanguage) {
    FunctionLanguage["JavaScript"] = "javascript";
    FunctionLanguage["TypeScript"] = "typescript";
    // CSharp = "csharp"
})(FunctionLanguage = exports.FunctionLanguage || (exports.FunctionLanguage = {}));
var FunctionConfigKey;
(function (FunctionConfigKey) {
    FunctionConfigKey["resourceGroupName"] = "resourceGroupName";
    FunctionConfigKey["subscriptionId"] = "subscriptionId";
    FunctionConfigKey["credential"] = "credential";
    FunctionConfigKey["location"] = "location";
    FunctionConfigKey["resourceNameSuffix"] = "resourceNameSuffix";
    FunctionConfigKey["functionLanguage"] = "programmingLanguage";
    /* Config exported by Function plugin */
    FunctionConfigKey["functionAppName"] = "functionAppName";
    FunctionConfigKey["defaultFunctionName"] = "defaultFunctionName";
    FunctionConfigKey["storageAccountName"] = "storageAccountName";
    FunctionConfigKey["appServicePlanName"] = "appServicePlanName";
    FunctionConfigKey["functionEndpoint"] = "functionEndpoint";
    FunctionConfigKey["functionAppResourceId"] = "functionAppResourceId";
    /* Intermediate  */
    FunctionConfigKey["functionName"] = "functionName";
    FunctionConfigKey["skipDeploy"] = "skipDeploy";
    FunctionConfigKey["site"] = "site";
})(FunctionConfigKey = exports.FunctionConfigKey || (exports.FunctionConfigKey = {}));
var QuestionKey;
(function (QuestionKey) {
    QuestionKey["functionName"] = "function-name";
    QuestionKey["programmingLanguage"] = "programming-language";
    QuestionKey["nodeVersion"] = "node-version";
    QuestionKey["stage"] = "stage";
})(QuestionKey = exports.QuestionKey || (exports.QuestionKey = {}));
var ResourceType;
(function (ResourceType) {
    ResourceType["storageAccount"] = "Azure storage account";
    ResourceType["appServicePlan"] = "Azure App Service plan";
    ResourceType["functionApp"] = "Azure function app";
})(ResourceType = exports.ResourceType || (exports.ResourceType = {}));
var TelemetryKey;
(function (TelemetryKey) {
    TelemetryKey["Component"] = "component";
    TelemetryKey["Success"] = "success";
    TelemetryKey["ErrorType"] = "error-type";
    TelemetryKey["ErrorMessage"] = "error-message";
    TelemetryKey["ErrorCode"] = "error-code";
    TelemetryKey["AppId"] = "appid";
    TelemetryKey["OSArch"] = "os-arch";
    TelemetryKey["OSRelease"] = "os-release";
})(TelemetryKey = exports.TelemetryKey || (exports.TelemetryKey = {}));
var TelemetryValue;
(function (TelemetryValue) {
    TelemetryValue["Success"] = "yes";
    TelemetryValue["Fail"] = "no";
    TelemetryValue["UserError"] = "user";
    TelemetryValue["SystemError"] = "system";
})(TelemetryValue = exports.TelemetryValue || (exports.TelemetryValue = {}));
var FunctionEvent;
(function (FunctionEvent) {
    FunctionEvent["preScaffold"] = "pre-scaffold";
    FunctionEvent["scaffold"] = "scaffold";
    FunctionEvent["postScaffold"] = "post-scaffold";
    FunctionEvent["preProvision"] = "pre-provision";
    FunctionEvent["provision"] = "provision";
    FunctionEvent["postProvision"] = "post-provision";
    FunctionEvent["preDeploy"] = "pre-deploy";
    FunctionEvent["deploy"] = "deploy";
    FunctionEvent["postDeploy"] = "post-deploy";
    FunctionEvent["getQuestions"] = "get-questions";
    FunctionEvent["executeUserTask"] = "execute-user-task";
    FunctionEvent["callFunc"] = "call-func";
    FunctionEvent["scaffoldFallback"] = "scaffold-fallback";
    FunctionEvent["skipDeploy"] = "skip-deploy";
    FunctionEvent["DeploymentInfoNotFound"] = "deployment-info-not-found";
    FunctionEvent["generateArmTemplates"] = "generate-arm-templates";
    FunctionEvent["updateArmTemplates"] = "update-arm-templates";
    FunctionEvent["addResource"] = "add-resource";
})(FunctionEvent = exports.FunctionEvent || (exports.FunctionEvent = {}));
var CustomizedTask;
(function (CustomizedTask) {
    CustomizedTask["addResource"] = "addResource";
    CustomizedTask["addFeature"] = "addFeature";
})(CustomizedTask = exports.CustomizedTask || (exports.CustomizedTask = {}));
