// @microsoft/teamsfx-core/build/plugins/resource/function/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionPluginResultFactory = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("./constants");
class FxResultFactory {
    static UserError(errorMessage, name, helpLink, innerError, stack) {
        return teamsfx_api_1.err(new teamsfx_api_1.UserError({
            name: name,
            message: errorMessage[0],
            displayMessage: errorMessage[1],
            source: this.source,
            helpLink: helpLink !== null && helpLink !== void 0 ? helpLink : this.defaultHelpLink,
            error: innerError,
        }));
    }
    static SystemError(errorMessage, name, issueLink, innerError, stack) {
        return teamsfx_api_1.err(new teamsfx_api_1.SystemError({
            name: name,
            message: errorMessage[0],
            displayMessage: errorMessage[1],
            source: this.source,
            issueLink: issueLink !== null && issueLink !== void 0 ? issueLink : this.defaultIssueLink,
            error: innerError,
        }));
    }
    static Success(result) {
        return teamsfx_api_1.ok(result);
    }
}
FxResultFactory.source = constants_1.CommonConstants.emptyString;
FxResultFactory.defaultHelpLink = constants_1.CommonConstants.emptyString;
FxResultFactory.defaultIssueLink = constants_1.CommonConstants.emptyString;
class FunctionPluginResultFactory extends FxResultFactory {
}
exports.FunctionPluginResultFactory = FunctionPluginResultFactory;
FunctionPluginResultFactory.source = constants_1.FunctionPluginInfo.alias;
FunctionPluginResultFactory.defaultHelpLink = constants_1.DefaultValues.helpLink;
FunctionPluginResultFactory.defaultIssueLink = constants_1.DefaultValues.issueLink;
