// @microsoft/teamsfx-core/build/plugins/resource/sql/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqlPluginImpl = void 0;
const tslib_1 = require("tslib");
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const errors_1 = require("./errors");
const results_1 = require("./results");
const dialogUtils_1 = require("./utils/dialogUtils");
const config_1 = require("./config");
const sqlClient_1 = require("./sqlClient");
const contextUtils_1 = require("./utils/contextUtils");
const commonUtils_1 = require("./utils/commonUtils");
const constants_1 = require("./constants");
const message_1 = require("./utils/message");
const telemetryUtils_1 = require("./utils/telemetryUtils");
const questions_1 = require("./questions");
const path_1 = tslib_1.__importDefault(require("path"));
const folder_1 = require("../../../folder");
const constants_2 = require("../../../common/constants");
const fs = tslib_1.__importStar(require("fs-extra"));
const common_1 = require("../../../common");
const ResourcePluginContainer_1 = require("../../solution/fx-solution/ResourcePluginContainer");
const adaptor_1 = require("../../solution/fx-solution/v2/adaptor");
const tools_1 = require("../../../common/tools");
const managementClient_1 = require("./managementClient");
class SqlPluginImpl {
    constructor() {
        this.config = new config_1.SqlConfig();
    }
    async loadConfig(ctx) {
        this.loadConfigSubscription(ctx);
        this.loadConfigResourceGroup(ctx);
        this.config.resourceNameSuffix = contextUtils_1.ContextUtils.getConfig(ctx, constants_1.Constants.solution, constants_1.Constants.solutionConfigKey.resourceNameSuffix);
        this.config.location = contextUtils_1.ContextUtils.getConfig(ctx, constants_1.Constants.solution, constants_1.Constants.solutionConfigKey.location);
        this.config.tenantId = contextUtils_1.ContextUtils.getConfig(ctx, constants_1.Constants.solution, constants_1.Constants.solutionConfigKey.tenantId);
        this.loadConfigSql(ctx);
        this.loadDatabases(ctx);
    }
    async getQuestions(stage, ctx) {
        var _a;
        if (stage === teamsfx_api_1.Stage.provision && ((_a = ctx.answers) === null || _a === void 0 ? void 0 : _a.platform) === teamsfx_api_1.Platform.CLI_HELP) {
            const sqlNode = this.buildQuestionNode();
            return teamsfx_api_1.ok(sqlNode);
        }
        return teamsfx_api_1.ok(undefined);
    }
    async preProvision(ctx) {
        var _a, _b;
        (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.info(message_1.Message.startPreProvision);
        this.removeDatabases(ctx);
        await this.loadConfig(ctx);
        await managementClient_1.SqlMgrClient.create(ctx.azureAccountProvider, this.config);
        dialogUtils_1.DialogUtils.init(ctx.ui);
        telemetryUtils_1.TelemetryUtils.init(ctx.telemetryReporter);
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.preProvision + constants_1.Telemetry.startSuffix);
        await this.loadSkipAddingUser(ctx);
        await this.checkSqlExisting(ctx);
        if (!this.config.existSql) {
            await this.askInputs(ctx);
            this.config.admin = ctx.answers[constants_1.Constants.questionKey.adminName];
            this.config.adminPassword = ctx.answers[constants_1.Constants.questionKey.adminPassword];
            if (!this.config.admin || !this.config.adminPassword) {
                throw results_1.SqlResultFactory.SystemError(errors_1.ErrorMessage.SqlInputError.name, errors_1.ErrorMessage.SqlInputError.message());
            }
        }
        await this.parseLoginToken(ctx);
        this.setPasswordContext(ctx);
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.preProvision, true);
        (_b = ctx.logProvider) === null || _b === void 0 ? void 0 : _b.info(message_1.Message.endPreProvision);
        return teamsfx_api_1.ok(undefined);
    }
    async postProvision(ctx) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.info(message_1.Message.startPostProvision);
        await this.loadConfig(ctx);
        dialogUtils_1.DialogUtils.init(ctx.ui, dialogUtils_1.ProgressTitle.PostProvision, Object.keys(dialogUtils_1.ConfigureMessage).length);
        telemetryUtils_1.TelemetryUtils.init(ctx.telemetryReporter);
        const telemetryProperties = {
            [constants_1.Telemetry.properties.skipAddingUser]: this.config.skipAddingUser
                ? constants_1.Telemetry.valueYes
                : constants_1.Telemetry.valueNo,
            [constants_1.Telemetry.properties.dbCount]: this.config.databases.length.toString(),
        };
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.postProvision + constants_1.Telemetry.startSuffix, undefined, telemetryProperties);
        ctx.config.delete(constants_1.Constants.adminPassword);
        await managementClient_1.SqlMgrClient.create(ctx.azureAccountProvider, this.config);
        (_b = ctx.logProvider) === null || _b === void 0 ? void 0 : _b.info(message_1.Message.addFirewall);
        await this.AddFireWallRules(managementClient_1.SqlMgrClient);
        await ((_c = dialogUtils_1.DialogUtils.progressBar) === null || _c === void 0 ? void 0 : _c.start());
        await ((_d = dialogUtils_1.DialogUtils.progressBar) === null || _d === void 0 ? void 0 : _d.next(dialogUtils_1.ConfigureMessage.postProvisionAddAadmin));
        await this.CheckAndSetAadAdmin(ctx, managementClient_1.SqlMgrClient);
        this.getIdentity(ctx);
        if (!this.config.skipAddingUser) {
            await ((_e = dialogUtils_1.DialogUtils.progressBar) === null || _e === void 0 ? void 0 : _e.next(dialogUtils_1.ConfigureMessage.postProvisionAddUser));
            // azure sql does not support service principal admin to add databse user currently, so just notice developer if so.
            if (this.config.aadAdminType === commonUtils_1.UserType.User) {
                const sqlClient = await sqlClient_1.SqlClient.create(ctx.azureAccountProvider, this.config);
                (_f = ctx.logProvider) === null || _f === void 0 ? void 0 : _f.info(message_1.Message.addDatabaseUser(this.config.identity));
                await this.addDatabaseUser(ctx, sqlClient, managementClient_1.SqlMgrClient);
            }
            else {
                const message = errors_1.ErrorMessage.ServicePrincipalWarning(this.config.identity, this.config.databaseName);
                (_g = ctx.logProvider) === null || _g === void 0 ? void 0 : _g.warning(`[${constants_1.Constants.pluginName}] ${message}. You can follow ${constants_1.HelpLinks.default} to add database user ${this.config.identity}`);
            }
        }
        else {
            (_h = ctx.logProvider) === null || _h === void 0 ? void 0 : _h.warning(`[${constants_1.Constants.pluginName}] Skip adding database user. You can follow ${constants_1.HelpLinks.default} to add database user ${this.config.identity}`);
        }
        await managementClient_1.SqlMgrClient.deleteLocalFirewallRule();
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.postProvision, true, telemetryProperties);
        (_j = ctx.logProvider) === null || _j === void 0 ? void 0 : _j.info(message_1.Message.endPostProvision);
        await ((_k = dialogUtils_1.DialogUtils.progressBar) === null || _k === void 0 ? void 0 : _k.end(true));
        return teamsfx_api_1.ok(undefined);
    }
    async updateArmTemplates(ctx) {
        telemetryUtils_1.TelemetryUtils.init(ctx.telemetryReporter);
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.updateArmTemplates + constants_1.Telemetry.startSuffix);
        const result = {
            Reference: {
                sqlResourceId: constants_1.AzureSqlBicep.sqlResourceId,
                sqlEndpoint: constants_1.AzureSqlBicep.sqlEndpoint,
                databaseName: constants_1.AzureSqlBicep.databaseName,
            },
        };
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.updateArmTemplates, true);
        return teamsfx_api_1.ok(result);
    }
    async addDatabaseUser(ctx, sqlClient, managementClient) {
        var _a;
        let retryCount = 0;
        const databaseWithUser = {};
        this.config.databases.forEach((element) => {
            databaseWithUser[element] = false;
        });
        while (true) {
            try {
                for (const database in databaseWithUser) {
                    if (!databaseWithUser[database]) {
                        await sqlClient.addDatabaseUser(database);
                        databaseWithUser[database] = true;
                    }
                }
                return;
            }
            catch (error) {
                if (!sqlClient_1.SqlClient.isFireWallError(error === null || error === void 0 ? void 0 : error.innerError) ||
                    retryCount >= constants_1.Constants.maxRetryTimes) {
                    throw error;
                }
                else {
                    retryCount++;
                    (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.warning(`[${constants_1.Constants.pluginName}] Retry adding new firewall rule to access azure sql, because the local IP address has changed after added firewall rule for it. [Retry time: ${retryCount}]`);
                    await managementClient.addLocalFirewallRule();
                }
            }
        }
    }
    async generateArmTemplates(ctx) {
        telemetryUtils_1.TelemetryUtils.init(ctx.telemetryReporter);
        const telemetryProperties = {
            [constants_1.Telemetry.properties.dbOnly]: constants_1.Telemetry.valueNo,
        };
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.generateArmTemplates + constants_1.Telemetry.startSuffix, undefined, telemetryProperties);
        const plugins = ResourcePluginContainer_1.getActivatedV2ResourcePlugins(ctx.projectSettings).map((p) => new adaptor_1.NamedArmResourcePluginAdaptor(p));
        const pluginCtx = { plugins: plugins.map((obj) => obj.name) };
        const bicepTemplateDirectory = path_1.default.join(folder_1.getTemplatesFolder(), "plugins", "resource", "sql", "bicep");
        const provisionOrchestration = await tools_1.generateBicepFromFile(path_1.default.join(bicepTemplateDirectory, constants_1.AzureSqlBicepFile.moduleTemplateFileName), pluginCtx);
        const provisionModules = await tools_1.generateBicepFromFile(path_1.default.join(bicepTemplateDirectory, constants_1.AzureSqlBicepFile.ProvisionModuleTemplateFileName), pluginCtx);
        const result = {
            Provision: {
                Orchestration: provisionOrchestration,
                Modules: { azureSql: provisionModules },
            },
            Parameters: JSON.parse(await fs.readFile(path_1.default.join(bicepTemplateDirectory, constants_2.Bicep.ParameterFileName), constants_2.ConstantString.UTF8Encoding)),
            Reference: {
                sqlResourceId: constants_1.AzureSqlBicep.sqlResourceId,
                sqlEndpoint: constants_1.AzureSqlBicep.sqlEndpoint,
                databaseName: constants_1.AzureSqlBicep.databaseName,
            },
        };
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.generateArmTemplates, true, telemetryProperties);
        return teamsfx_api_1.ok(result);
    }
    async generateNewDatabaseBicepSnippet(ctx) {
        telemetryUtils_1.TelemetryUtils.init(ctx.telemetryReporter);
        const telemetryProperties = {
            [constants_1.Telemetry.properties.dbOnly]: constants_1.Telemetry.valueYes,
        };
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.generateArmTemplates + constants_1.Telemetry.startSuffix, undefined, telemetryProperties);
        const suffix = tools_1.getUuid().substring(0, 6);
        const compileCtx = {
            suffix: suffix,
        };
        const bicepTemplateDirectory = path_1.default.join(folder_1.getTemplatesFolder(), "plugins", "resource", "sql", "bicep");
        const provisionOrchestration = await tools_1.generateBicepFromFile(path_1.default.join(bicepTemplateDirectory, constants_1.AzureSqlBicepFile.newDatabaseOrchestrationTemplateFileName), compileCtx);
        const provisionModules = await tools_1.generateBicepFromFile(path_1.default.join(bicepTemplateDirectory, constants_1.AzureSqlBicepFile.newDatabaseProvisionTemplateFileName), compileCtx);
        const result = {
            Provision: {
                Orchestration: provisionOrchestration,
                Modules: { azureSql: provisionModules },
            },
            Reference: {
                sqlResourceId: constants_1.AzureSqlBicep.sqlResourceId,
                sqlEndpoint: constants_1.AzureSqlBicep.sqlEndpoint,
                databaseName: constants_1.AzureSqlBicep.databaseName,
            },
        };
        telemetryUtils_1.TelemetryUtils.sendEvent(constants_1.Telemetry.stage.generateArmTemplates, true, telemetryProperties);
        return teamsfx_api_1.ok(result);
    }
    setPasswordContext(ctx) {
        ctx.config.set(constants_1.Constants.admin, this.config.admin);
        ctx.config.set(constants_1.Constants.adminPassword, this.config.adminPassword);
    }
    buildQuestionNode() {
        const sqlNode = new teamsfx_api_1.QTreeNode({
            type: "group",
        });
        sqlNode.addChild(new teamsfx_api_1.QTreeNode(questions_1.adminNameQuestion));
        sqlNode.addChild(new teamsfx_api_1.QTreeNode(questions_1.adminPasswordQuestion));
        sqlNode.addChild(new teamsfx_api_1.QTreeNode(questions_1.confirmPasswordQuestion));
        return sqlNode;
    }
    async AddFireWallRules(client) {
        await client.addLocalFirewallRule();
    }
    async CheckAndSetAadAdmin(ctx, client) {
        var _a, _b;
        let existAdmin = false;
        existAdmin = await client.existAadAdmin();
        if (!existAdmin) {
            (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.info(message_1.Message.addSqlAadAdmin);
            await client.addAADadmin();
        }
        else {
            (_b = ctx.logProvider) === null || _b === void 0 ? void 0 : _b.info(message_1.Message.skipAddAadAdmin);
        }
    }
    getIdentity(ctx) {
        const identityConfig = ctx.envInfo.state.get(constants_1.Constants.identityPlugin);
        this.config.identity = identityConfig.get(constants_1.Constants.identityName);
        if (!this.config.identity) {
            const error = results_1.SqlResultFactory.SystemError(errors_1.ErrorMessage.SqlGetConfigError.name, errors_1.ErrorMessage.SqlGetConfigError.message(constants_1.Constants.identityPlugin, constants_1.Constants.identityName));
            throw error;
        }
    }
    async loadSkipAddingUser(ctx) {
        var _a, _b;
        const skipAddingUser = (_a = ctx.envInfo.config) === null || _a === void 0 ? void 0 : _a[constants_1.Constants.skipAddingSqlUser];
        if (skipAddingUser === undefined) {
            this.config.skipAddingUser = (await ((_b = ctx.azureAccountProvider) === null || _b === void 0 ? void 0 : _b.getIdentityCredentialAsync()))
                ? false
                : true;
        }
        else {
            this.config.skipAddingUser = skipAddingUser;
        }
    }
    async checkSqlExisting(ctx) {
        this.config.admin = ctx.config.get(constants_1.Constants.admin);
        this.config.adminPassword = ctx.config.get(constants_1.Constants.adminPassword);
        this.config.sqlEndpoint = ctx.config.get(constants_1.Constants.sqlEndpoint);
        if (this.config.sqlEndpoint && this.config.azureSubscriptionId) {
            this.config.existSql = await managementClient_1.SqlMgrClient.existAzureSQL();
        }
    }
    async askInputs(ctx) {
        const node = this.buildQuestionNode();
        const res = await teamsfx_api_1.traverse(node, ctx.answers, ctx.ui);
        if (res.isErr()) {
            throw results_1.SqlResultFactory.UserError(errors_1.ErrorMessage.SqlAskInputError.name, errors_1.ErrorMessage.SqlAskInputError.message(), res.error);
        }
    }
    async parseLoginToken(ctx) {
        var _a;
        // get login user info to set aad admin in sql
        try {
            const credential = await ctx.azureAccountProvider.getAccountCredentialAsync();
            const token = await credential.getToken();
            const accessToken = token.accessToken;
            const tokenInfo = commonUtils_1.parseToken(accessToken);
            this.config.aadAdmin = tokenInfo.name;
            this.config.aadAdminObjectId = tokenInfo.objectId;
            this.config.aadAdminType = tokenInfo.userType;
            (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.debug(message_1.Message.adminName(tokenInfo.name));
        }
        catch (error) {
            throw results_1.SqlResultFactory.SystemError(errors_1.ErrorMessage.SqlUserInfoError.name, errors_1.ErrorMessage.SqlUserInfoError.message(), error);
        }
    }
    loadConfigResourceGroup(ctx) {
        this.config.sqlResourceId = ctx.config.get(constants_1.Constants.sqlResourceId);
        if (this.config.sqlResourceId) {
            try {
                this.config.resourceGroup = common_1.getResourceGroupNameFromResourceId(this.config.sqlResourceId);
            }
            catch (error) {
                throw results_1.SqlResultFactory.UserError(errors_1.ErrorMessage.SqlInvalidConfigError.name, errors_1.ErrorMessage.SqlInvalidConfigError.message(this.config.sqlResourceId, error.message), error);
            }
        }
    }
    loadConfigSubscription(ctx) {
        this.config.sqlResourceId = ctx.config.get(constants_1.Constants.sqlResourceId);
        if (this.config.sqlResourceId) {
            try {
                this.config.azureSubscriptionId = common_1.getSubscriptionIdFromResourceId(this.config.sqlResourceId);
            }
            catch (error) {
                throw results_1.SqlResultFactory.UserError(errors_1.ErrorMessage.SqlInvalidConfigError.name, errors_1.ErrorMessage.SqlInvalidConfigError.message(this.config.sqlResourceId, error.message), error);
            }
        }
    }
    loadConfigSql(ctx) {
        this.config.sqlEndpoint = ctx.config.get(constants_1.Constants.sqlEndpoint);
        this.config.databaseName = ctx.config.get(constants_1.Constants.databaseName);
        if (this.config.sqlEndpoint) {
            this.config.sqlServer = this.config.sqlEndpoint.split(".")[0];
        }
    }
    loadDatabases(ctx) {
        ctx.config.forEach((v, k) => {
            if (k.startsWith(constants_1.Constants.databaseName)) {
                if (!this.config.databases.includes(v)) {
                    this.config.databases.push(v);
                }
            }
        });
    }
    removeDatabases(ctx) {
        ctx.config.forEach((v, k) => {
            if (k.startsWith(constants_1.Constants.databaseName) && k !== constants_1.Constants.databaseName) {
                ctx.config.delete(k);
            }
        });
    }
}
exports.SqlPluginImpl = SqlPluginImpl;
