// @microsoft/teamsfx-core/build/plugins/solution/fx-solution/v2/createEnv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEnv = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const core_1 = require("../../../../core");
const environment_1 = require("../../../../core/environment");
const arm_1 = require("../arm");
const utils_1 = require("./utils");
async function createEnv(ctx, inputs) {
    if (utils_1.isAzureProject(ctx.projectSetting.solutionSettings)) {
        const solutionContext = {
            root: inputs.projectPath || "",
            envInfo: environment_1.newEnvInfo(inputs.targetEnvName),
            projectSettings: ctx.projectSetting,
            answers: inputs,
            logProvider: ctx.logProvider,
            telemetryReporter: ctx.telemetryReporter,
            cryptoProvider: ctx.cryptoProvider,
            permissionRequestProvider: ctx.permissionRequestProvider,
            ui: ctx.userInteraction,
        };
        try {
            if (inputs.copy === true) {
                await arm_1.copyParameterJson(inputs.projectPath, ctx.projectSetting.appName, inputs.targetEnvName, inputs.sourceEnvName);
            }
            else {
                await arm_1.getParameterJson(solutionContext);
            }
        }
        catch (e) {
            return teamsfx_api_1.err(core_1.CopyFileError(e));
        }
    }
    return teamsfx_api_1.ok(teamsfx_api_1.Void);
}
exports.createEnv = createEnv;
