// @microsoft/teamsfx-core/build/plugins/solution/fx-solution/v2/scaffolding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scaffoldReadme = exports.scaffoldByPlugins = exports.scaffoldSourceCode = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const question_1 = require("../question");
const executor_1 = require("./executor");
const utils_1 = require("./utils");
const projectSettingsHelper_1 = require("../../../../common/projectSettingsHelper");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const constants_1 = require("../constants");
const ResourcePluginContainer_1 = require("../ResourcePluginContainer");
const typedi_1 = require("typedi");
const scaffolding_1 = require("../debug/scaffolding");
const folder_1 = require("../../../../folder");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const tools_1 = require("./../../../../common/tools");
async function scaffoldSourceCode(ctx, inputs) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    (_a = ctx.telemetryReporter) === null || _a === void 0 ? void 0 : _a.sendTelemetryEvent(constants_1.SolutionTelemetryEvent.CreateStart, {
        [constants_1.SolutionTelemetryProperty.Component]: constants_1.SolutionTelemetryComponentName,
    });
    if (inputs.projectPath === undefined) {
        return teamsfx_api_1.err(new teamsfx_api_1.SystemError("Solution", constants_1.SolutionError.InternelError, "projectPath is undefined"));
    }
    const lang = inputs[question_1.AzureSolutionQuestionNames.ProgrammingLanguage];
    if (lang) {
        ctx.projectSetting.programmingLanguage = lang;
    }
    const solutionSettings = utils_1.getAzureSolutionSettings(ctx);
    const fillinRes = utils_1.fillInSolutionSettings(ctx.projectSetting, inputs);
    if (fillinRes.isErr())
        return teamsfx_api_1.err(fillinRes.error);
    const plugins = utils_1.getSelectedPlugins(ctx.projectSetting);
    let thunks = plugins
        .filter((plugin) => !!plugin.scaffoldSourceCode)
        .map((plugin) => {
        return {
            pluginName: `${plugin.name}`,
            taskName: "scaffoldSourceCode",
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            thunk: () => plugin.scaffoldSourceCode(ctx, inputs),
        };
    });
    ///SPFx plugin will be executed last, so remove it from the thunks.
    const SPFxPlugin = typedi_1.Container.get(ResourcePluginContainer_1.ResourcePluginsV2.SpfxPlugin);
    if (thunks.map((p) => p.pluginName === SPFxPlugin.name).length > 0) {
        thunks = thunks.filter((p) => p.pluginName !== SPFxPlugin.name);
    }
    const result = await executor_1.executeConcurrently(thunks, ctx.logProvider);
    if (result.kind === "success") {
        const capabilities = (solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.capabilities) || [];
        const azureResources = (solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.azureResources) || [];
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const scaffoldLocalDebugSettingsResult = await scaffolding_1.scaffoldLocalDebugSettings(ctx, inputs);
        if (scaffoldLocalDebugSettingsResult.isErr()) {
            return scaffoldLocalDebugSettingsResult;
        }
        if (!projectSettingsHelper_1.isVSProject(ctx.projectSetting)) {
            await scaffoldReadme(capabilities, azureResources, inputs.projectPath);
        }
        if (utils_1.isAzureProject(solutionSettings)) {
            if (!tools_1.isAadManifestEnabled()) {
                await fs_extra_1.default.writeJSON(`${inputs.projectPath}/permissions.json`, constants_1.DEFAULT_PERMISSION_REQUEST, {
                    spaces: 4,
                });
            }
            (_b = ctx.telemetryReporter) === null || _b === void 0 ? void 0 : _b.sendTelemetryEvent(constants_1.SolutionTelemetryEvent.Create, {
                [constants_1.SolutionTelemetryProperty.Component]: constants_1.SolutionTelemetryComponentName,
                [constants_1.SolutionTelemetryProperty.Success]: constants_1.SolutionTelemetrySuccess.Yes,
                [constants_1.SolutionTelemetryProperty.Resources]: ((solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.azureResources) || []).join(";"),
                [constants_1.SolutionTelemetryProperty.Capabilities]: ((solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.capabilities) || []).join(";"),
                [constants_1.SolutionTelemetryProperty.ProgrammingLanguage]: (_d = (_c = ctx.projectSetting) === null || _c === void 0 ? void 0 : _c.programmingLanguage) !== null && _d !== void 0 ? _d : "",
                [constants_1.SolutionTelemetryProperty.HostType]: "azure",
            });
        }
        else {
            //For SPFx plugin, execute it alone lastly
            if (SPFxPlugin.scaffoldSourceCode) {
                const spfxRes = await SPFxPlugin.scaffoldSourceCode(ctx, inputs);
                if (spfxRes.isErr()) {
                    return teamsfx_api_1.err(spfxRes.error);
                }
                (_e = ctx.telemetryReporter) === null || _e === void 0 ? void 0 : _e.sendTelemetryEvent(constants_1.SolutionTelemetryEvent.Create, {
                    [constants_1.SolutionTelemetryProperty.Component]: constants_1.SolutionTelemetryComponentName,
                    [constants_1.SolutionTelemetryProperty.Success]: constants_1.SolutionTelemetrySuccess.Yes,
                    [constants_1.SolutionTelemetryProperty.Capabilities]: ((solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.capabilities) || []).join(";"),
                    [constants_1.SolutionTelemetryProperty.ProgrammingLanguage]: (_g = (_f = ctx.projectSetting) === null || _f === void 0 ? void 0 : _f.programmingLanguage) !== null && _g !== void 0 ? _g : "",
                    [constants_1.SolutionTelemetryProperty.HostType]: "spfx",
                });
            }
        }
        ctx.userInteraction.showMessage("info", `Success: ${localizeUtils_1.getLocalizedString("core.create.successNotice")}`, false);
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    }
    else {
        (_h = ctx.telemetryReporter) === null || _h === void 0 ? void 0 : _h.sendTelemetryErrorEvent(constants_1.SolutionTelemetryEvent.Create, {
            [constants_1.SolutionTelemetryProperty.Component]: constants_1.SolutionTelemetryComponentName,
            [constants_1.SolutionTelemetryProperty.Success]: constants_1.SolutionTelemetrySuccess.No,
            [constants_1.SolutionTelemetryProperty.ErrorCode]: result.error.name,
            [constants_1.SolutionTelemetryProperty.ErrorMessage]: result.error.message,
            [constants_1.SolutionTelemetryProperty.HostType]: utils_1.isAzureProject(solutionSettings) ? "azure" : "spfx",
        });
        return teamsfx_api_1.err(result.error);
    }
}
exports.scaffoldSourceCode = scaffoldSourceCode;
async function scaffoldByPlugins(ctx, inputs, localSettings, plugins, concurrent = true) {
    var _a, _b, _c;
    if (plugins.length === 0)
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    (_a = ctx.logProvider) === null || _a === void 0 ? void 0 : _a.info(`start scaffolding ${plugins.map((p) => p.name).join(",")}.....`);
    const thunks = plugins
        .filter((plugin) => !!plugin.scaffoldSourceCode)
        .map((plugin) => {
        return {
            pluginName: `${plugin.name}`,
            taskName: "scaffoldSourceCode",
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            thunk: () => plugin.scaffoldSourceCode(ctx, inputs),
        };
    });
    const result = await executor_1.executeConcurrently(thunks, ctx.logProvider, concurrent);
    const solutionSettings = utils_1.getAzureSolutionSettings(ctx);
    if (result.kind === "success") {
        const capabilities = (solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.capabilities) || [];
        const azureResources = (solutionSettings === null || solutionSettings === void 0 ? void 0 : solutionSettings.azureResources) || [];
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        await scaffoldReadme(capabilities, azureResources, inputs.projectPath);
        const isVs = projectSettingsHelper_1.isVSProject(ctx.projectSetting);
        if (isVs) {
            ctx.logProvider.info(`Success: ${localizeUtils_1.getLocalizedString("core.create.successNotice")}`);
        }
        else {
            ctx.userInteraction.showMessage("info", `Success: ${localizeUtils_1.getLocalizedString("core.create.successNotice")}`, false);
        }
        (_b = ctx.logProvider) === null || _b === void 0 ? void 0 : _b.info(`finish scaffolding ${plugins.map((p) => p.name).join(",")}!`);
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    }
    else {
        (_c = ctx.logProvider) === null || _c === void 0 ? void 0 : _c.info(`failed to scaffold ${plugins.map((p) => p.name).join(",")}!`);
        return teamsfx_api_1.err(result.error);
    }
}
exports.scaffoldByPlugins = scaffoldByPlugins;
async function scaffoldReadme(capabilities, azureResources, projectPath) {
    capabilities = capabilities || [];
    azureResources = azureResources || [];
    const hasBot = capabilities.includes(question_1.BotOptionItem.id);
    const hasMsgExt = capabilities.includes(question_1.MessageExtensionItem.id);
    const hasTab = capabilities.includes(question_1.TabOptionItem.id);
    if (hasTab && (hasBot || hasMsgExt)) {
        const readme = path_1.default.join(folder_1.getTemplatesFolder(), "plugins", "solution", "README.md");
        if (await fs_extra_1.default.pathExists(readme)) {
            await fs_extra_1.default.copy(readme, `${projectPath}/${teamsfx_api_1.AutoGeneratedReadme}`);
        }
    }
}
exports.scaffoldReadme = scaffoldReadme;
