// @azure/arm-resources/esm/operations/providerResourceTypes.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/providerResourceTypesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ProviderResourceTypes. */
var ProviderResourceTypes = /** @class */ (function () {
    /**
     * Create a ProviderResourceTypes.
     * @param {ResourceManagementClientContext} client Reference to the service client.
     */
    function ProviderResourceTypes(client) {
        this.client = client;
    }
    ProviderResourceTypes.prototype.list = function (resourceProviderNamespace, options, callback) {
        return this.client.sendOperationRequest({
            resourceProviderNamespace: resourceProviderNamespace,
            options: options
        }, listOperationSpec, callback);
    };
    return ProviderResourceTypes;
}());
export { ProviderResourceTypes };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/providers/{resourceProviderNamespace}/resourceTypes",
    urlParameters: [Parameters.resourceProviderNamespace, Parameters.subscriptionId],
    queryParameters: [Parameters.expand, Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {
            bodyMapper: Mappers.ProviderResourceTypeListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
