// @azure/arm-resources/esm/operations/tagsOperations.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/tagsOperationsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a TagsOperations. */
var TagsOperations = /** @class */ (function () {
    /**
     * Create a TagsOperations.
     * @param {ResourceManagementClientContext} client Reference to the service client.
     */
    function TagsOperations(client) {
        this.client = client;
    }
    TagsOperations.prototype.deleteValue = function (tagName, tagValue, options, callback) {
        return this.client.sendOperationRequest({
            tagName: tagName,
            tagValue: tagValue,
            options: options
        }, deleteValueOperationSpec, callback);
    };
    TagsOperations.prototype.createOrUpdateValue = function (tagName, tagValue, options, callback) {
        return this.client.sendOperationRequest({
            tagName: tagName,
            tagValue: tagValue,
            options: options
        }, createOrUpdateValueOperationSpec, callback);
    };
    TagsOperations.prototype.createOrUpdate = function (tagName, options, callback) {
        return this.client.sendOperationRequest({
            tagName: tagName,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    TagsOperations.prototype.deleteMethod = function (tagName, options, callback) {
        return this.client.sendOperationRequest({
            tagName: tagName,
            options: options
        }, deleteMethodOperationSpec, callback);
    };
    TagsOperations.prototype.list = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, listOperationSpec, callback);
    };
    TagsOperations.prototype.createOrUpdateAtScope = function (scope, parameters, options, callback) {
        return this.client.sendOperationRequest({
            scope: scope,
            parameters: parameters,
            options: options
        }, createOrUpdateAtScopeOperationSpec, callback);
    };
    TagsOperations.prototype.updateAtScope = function (scope, parameters, options, callback) {
        return this.client.sendOperationRequest({
            scope: scope,
            parameters: parameters,
            options: options
        }, updateAtScopeOperationSpec, callback);
    };
    TagsOperations.prototype.getAtScope = function (scope, options, callback) {
        return this.client.sendOperationRequest({
            scope: scope,
            options: options
        }, getAtScopeOperationSpec, callback);
    };
    TagsOperations.prototype.deleteAtScope = function (scope, options, callback) {
        return this.client.sendOperationRequest({
            scope: scope,
            options: options
        }, deleteAtScopeOperationSpec, callback);
    };
    TagsOperations.prototype.listNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listNextOperationSpec, callback);
    };
    return TagsOperations;
}());
export { TagsOperations };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var deleteValueOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/tagNames/{tagName}/tagValues/{tagValue}",
    urlParameters: [Parameters.tagName, Parameters.tagValue, Parameters.subscriptionId],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var createOrUpdateValueOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/tagNames/{tagName}/tagValues/{tagValue}",
    urlParameters: [Parameters.tagName, Parameters.tagValue, Parameters.subscriptionId],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {
            bodyMapper: Mappers.TagValue
        },
        201: {
            bodyMapper: Mappers.TagValue
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/tagNames/{tagName}",
    urlParameters: [Parameters.tagName, Parameters.subscriptionId],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {
            bodyMapper: Mappers.TagDetails
        },
        201: {
            bodyMapper: Mappers.TagDetails
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/tagNames/{tagName}",
    urlParameters: [Parameters.tagName, Parameters.subscriptionId],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/tagNames",
    urlParameters: [Parameters.subscriptionId],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {
            bodyMapper: Mappers.TagsListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var createOrUpdateAtScopeOperationSpec = {
    httpMethod: "PUT",
    path: "{scope}/providers/Microsoft.Resources/tags/default",
    urlParameters: [Parameters.scope],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.TagsResource), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.TagsResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var updateAtScopeOperationSpec = {
    httpMethod: "PATCH",
    path: "{scope}/providers/Microsoft.Resources/tags/default",
    urlParameters: [Parameters.scope],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.TagsPatchResource), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.TagsResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getAtScopeOperationSpec = {
    httpMethod: "GET",
    path: "{scope}/providers/Microsoft.Resources/tags/default",
    urlParameters: [Parameters.scope],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {
            bodyMapper: Mappers.TagsResource
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var deleteAtScopeOperationSpec = {
    httpMethod: "DELETE",
    path: "{scope}/providers/Microsoft.Resources/tags/default",
    urlParameters: [Parameters.scope],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [Parameters.nextPageLink],
    queryParameters: [Parameters.apiVersion],
    headerParameters: [Parameters.acceptLanguage],
    responses: {
        200: {
            bodyMapper: Mappers.TagsListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
