// @azure/arm-sql/esm/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
export var acceptLanguage = {
    parameterPath: "acceptLanguage",
    mapper: {
        serializedName: "accept-language",
        defaultValue: 'en-US',
        type: {
            name: "String"
        }
    }
};
export var administratorName0 = {
    parameterPath: "administratorName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "administratorName",
        defaultValue: 'activeDirectory',
        type: {
            name: "String"
        }
    }
};
export var administratorName1 = {
    parameterPath: "administratorName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "administratorName",
        defaultValue: 'ActiveDirectory',
        type: {
            name: "String"
        }
    }
};
export var apiVersion0 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2014-04-01',
        type: {
            name: "String"
        }
    }
};
export var apiVersion1 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2015-05-01-preview',
        type: {
            name: "String"
        }
    }
};
export var apiVersion2 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2017-10-01-preview',
        type: {
            name: "String"
        }
    }
};
export var apiVersion3 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2018-06-01-preview',
        type: {
            name: "String"
        }
    }
};
export var apiVersion4 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2017-03-01-preview',
        type: {
            name: "String"
        }
    }
};
export var backupName = {
    parameterPath: "backupName",
    mapper: {
        required: true,
        serializedName: "backupName",
        type: {
            name: "String"
        }
    }
};
export var baselineName = {
    parameterPath: "baselineName",
    mapper: {
        required: true,
        serializedName: "baselineName",
        type: {
            name: "Enum",
            allowedValues: [
                "master",
                "default"
            ]
        }
    }
};
export var blobAuditingPolicyName = {
    parameterPath: "blobAuditingPolicyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "blobAuditingPolicyName",
        defaultValue: 'default',
        type: {
            name: "String"
        }
    }
};
export var columnName = {
    parameterPath: "columnName",
    mapper: {
        required: true,
        serializedName: "columnName",
        type: {
            name: "String"
        }
    }
};
export var communicationLinkName = {
    parameterPath: "communicationLinkName",
    mapper: {
        required: true,
        serializedName: "communicationLinkName",
        type: {
            name: "String"
        }
    }
};
export var connectionPolicyName = {
    parameterPath: "connectionPolicyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "connectionPolicyName",
        defaultValue: 'default',
        type: {
            name: "String"
        }
    }
};
export var continuationToken = {
    parameterPath: [
        "options",
        "continuationToken"
    ],
    mapper: {
        serializedName: "continuationToken",
        type: {
            name: "String"
        }
    }
};
export var createTimeMax = {
    parameterPath: [
        "options",
        "createTimeMax"
    ],
    mapper: {
        serializedName: "createTimeMax",
        type: {
            name: "DateTime"
        }
    }
};
export var createTimeMin = {
    parameterPath: [
        "options",
        "createTimeMin"
    ],
    mapper: {
        serializedName: "createTimeMin",
        type: {
            name: "DateTime"
        }
    }
};
export var credentialName = {
    parameterPath: "credentialName",
    mapper: {
        required: true,
        serializedName: "credentialName",
        type: {
            name: "String"
        }
    }
};
export var databaseName = {
    parameterPath: "databaseName",
    mapper: {
        required: true,
        serializedName: "databaseName",
        type: {
            name: "String"
        }
    }
};
export var databaseState = {
    parameterPath: [
        "options",
        "databaseState"
    ],
    mapper: {
        serializedName: "databaseState",
        type: {
            name: "String"
        }
    }
};
export var dataMaskingPolicyName = {
    parameterPath: "dataMaskingPolicyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "dataMaskingPolicyName",
        defaultValue: 'Default',
        type: {
            name: "String"
        }
    }
};
export var dataMaskingRuleName = {
    parameterPath: "dataMaskingRuleName",
    mapper: {
        required: true,
        serializedName: "dataMaskingRuleName",
        type: {
            name: "String"
        }
    }
};
export var dnsAliasName = {
    parameterPath: "dnsAliasName",
    mapper: {
        required: true,
        serializedName: "dnsAliasName",
        type: {
            name: "String"
        }
    }
};
export var elasticPoolName = {
    parameterPath: "elasticPoolName",
    mapper: {
        required: true,
        serializedName: "elasticPoolName",
        type: {
            name: "String"
        }
    }
};
export var encryptionProtectorName = {
    parameterPath: "encryptionProtectorName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "encryptionProtectorName",
        defaultValue: 'current',
        type: {
            name: "String"
        }
    }
};
export var endTime = {
    parameterPath: "endTime",
    mapper: {
        required: true,
        serializedName: "endTime",
        type: {
            name: "String"
        }
    }
};
export var endTimeMax = {
    parameterPath: [
        "options",
        "endTimeMax"
    ],
    mapper: {
        serializedName: "endTimeMax",
        type: {
            name: "DateTime"
        }
    }
};
export var endTimeMin = {
    parameterPath: [
        "options",
        "endTimeMin"
    ],
    mapper: {
        serializedName: "endTimeMin",
        type: {
            name: "DateTime"
        }
    }
};
export var expandChildren = {
    parameterPath: [
        "options",
        "expandChildren"
    ],
    mapper: {
        serializedName: "expandChildren",
        type: {
            name: "Boolean"
        }
    }
};
export var extensionName = {
    parameterPath: "extensionName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "extensionName",
        defaultValue: 'import',
        type: {
            name: "String"
        }
    }
};
export var failoverGroupName = {
    parameterPath: "failoverGroupName",
    mapper: {
        required: true,
        serializedName: "failoverGroupName",
        type: {
            name: "String"
        }
    }
};
export var filter0 = {
    parameterPath: "filter",
    mapper: {
        required: true,
        serializedName: "$filter",
        type: {
            name: "String"
        }
    }
};
export var filter1 = {
    parameterPath: [
        "options",
        "filter"
    ],
    mapper: {
        serializedName: "$filter",
        type: {
            name: "String"
        }
    }
};
export var firewallRuleName = {
    parameterPath: "firewallRuleName",
    mapper: {
        required: true,
        serializedName: "firewallRuleName",
        type: {
            name: "String"
        }
    }
};
export var geoBackupPolicyName = {
    parameterPath: "geoBackupPolicyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "geoBackupPolicyName",
        defaultValue: 'Default',
        type: {
            name: "String"
        }
    }
};
export var groupName = {
    parameterPath: "groupName",
    mapper: {
        required: true,
        serializedName: "groupName",
        type: {
            name: "String"
        }
    }
};
export var include = {
    parameterPath: [
        "options",
        "include"
    ],
    mapper: {
        serializedName: "include",
        type: {
            name: "String"
        }
    }
};
export var includeDisabledRecommendations = {
    parameterPath: [
        "options",
        "includeDisabledRecommendations"
    ],
    mapper: {
        serializedName: "includeDisabledRecommendations",
        type: {
            name: "Boolean"
        }
    }
};
export var instancePoolName = {
    parameterPath: "instancePoolName",
    mapper: {
        required: true,
        serializedName: "instancePoolName",
        type: {
            name: "String"
        }
    }
};
export var isActive = {
    parameterPath: [
        "options",
        "isActive"
    ],
    mapper: {
        serializedName: "isActive",
        type: {
            name: "Boolean"
        }
    }
};
export var jobAgentName = {
    parameterPath: "jobAgentName",
    mapper: {
        required: true,
        serializedName: "jobAgentName",
        type: {
            name: "String"
        }
    }
};
export var jobExecutionId = {
    parameterPath: "jobExecutionId",
    mapper: {
        required: true,
        serializedName: "jobExecutionId",
        type: {
            name: "Uuid"
        }
    }
};
export var jobName = {
    parameterPath: "jobName",
    mapper: {
        required: true,
        serializedName: "jobName",
        type: {
            name: "String"
        }
    }
};
export var jobVersion = {
    parameterPath: "jobVersion",
    mapper: {
        required: true,
        serializedName: "jobVersion",
        type: {
            name: "Number"
        }
    }
};
export var keyName = {
    parameterPath: "keyName",
    mapper: {
        required: true,
        serializedName: "keyName",
        type: {
            name: "String"
        }
    }
};
export var linkId = {
    parameterPath: "linkId",
    mapper: {
        required: true,
        serializedName: "linkId",
        type: {
            name: "String"
        }
    }
};
export var locationName = {
    parameterPath: "locationName",
    mapper: {
        required: true,
        serializedName: "locationName",
        type: {
            name: "String"
        }
    }
};
export var longTermRetentionDatabaseName = {
    parameterPath: "longTermRetentionDatabaseName",
    mapper: {
        required: true,
        serializedName: "longTermRetentionDatabaseName",
        type: {
            name: "String"
        }
    }
};
export var longTermRetentionServerName = {
    parameterPath: "longTermRetentionServerName",
    mapper: {
        required: true,
        serializedName: "longTermRetentionServerName",
        type: {
            name: "String"
        }
    }
};
export var managedInstanceName = {
    parameterPath: "managedInstanceName",
    mapper: {
        required: true,
        serializedName: "managedInstanceName",
        type: {
            name: "String"
        }
    }
};
export var nextPageLink = {
    parameterPath: "nextPageLink",
    mapper: {
        required: true,
        serializedName: "nextLink",
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export var onlyLatestPerDatabase = {
    parameterPath: [
        "options",
        "onlyLatestPerDatabase"
    ],
    mapper: {
        serializedName: "onlyLatestPerDatabase",
        type: {
            name: "Boolean"
        }
    }
};
export var operationId = {
    parameterPath: "operationId",
    mapper: {
        required: true,
        serializedName: "operationId",
        type: {
            name: "Uuid"
        }
    }
};
export var policyName = {
    parameterPath: "policyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "policyName",
        defaultValue: 'default',
        type: {
            name: "String"
        }
    }
};
export var privateEndpointConnectionName = {
    parameterPath: "privateEndpointConnectionName",
    mapper: {
        required: true,
        serializedName: "privateEndpointConnectionName",
        type: {
            name: "String"
        }
    }
};
export var recommendedElasticPoolName = {
    parameterPath: "recommendedElasticPoolName",
    mapper: {
        required: true,
        serializedName: "recommendedElasticPoolName",
        type: {
            name: "String"
        }
    }
};
export var recoverableDatabaseName = {
    parameterPath: "recoverableDatabaseName",
    mapper: {
        required: true,
        serializedName: "recoverableDatabaseName",
        type: {
            name: "String"
        }
    }
};
export var replicaType = {
    parameterPath: [
        "options",
        "replicaType"
    ],
    mapper: {
        serializedName: "replicaType",
        type: {
            name: "String"
        }
    }
};
export var resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        required: true,
        serializedName: "resourceGroupName",
        type: {
            name: "String"
        }
    }
};
export var restorableDroppedDatabaseId = {
    parameterPath: "restorableDroppedDatabaseId",
    mapper: {
        required: true,
        serializedName: "restorableDroppedDatabaseId",
        type: {
            name: "String"
        }
    }
};
export var restorableDroppededDatabaseId = {
    parameterPath: "restorableDroppededDatabaseId",
    mapper: {
        required: true,
        serializedName: "restorableDroppededDatabaseId",
        type: {
            name: "String"
        }
    }
};
export var restoreDetailsName = {
    parameterPath: "restoreDetailsName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "restoreDetailsName",
        defaultValue: 'Default',
        type: {
            name: "String"
        }
    }
};
export var restorePointName = {
    parameterPath: "restorePointName",
    mapper: {
        required: true,
        serializedName: "restorePointName",
        type: {
            name: "String"
        }
    }
};
export var ruleId = {
    parameterPath: "ruleId",
    mapper: {
        required: true,
        serializedName: "ruleId",
        type: {
            name: "String"
        }
    }
};
export var scanId = {
    parameterPath: "scanId",
    mapper: {
        required: true,
        serializedName: "scanId",
        type: {
            name: "String"
        }
    }
};
export var schemaName = {
    parameterPath: "schemaName",
    mapper: {
        required: true,
        serializedName: "schemaName",
        type: {
            name: "String"
        }
    }
};
export var securityAlertPolicyName0 = {
    parameterPath: "securityAlertPolicyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "securityAlertPolicyName",
        defaultValue: 'default',
        type: {
            name: "String"
        }
    }
};
export var securityAlertPolicyName1 = {
    parameterPath: "securityAlertPolicyName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "securityAlertPolicyName",
        defaultValue: 'Default',
        type: {
            name: "String"
        }
    }
};
export var sensitivityLabelSource0 = {
    parameterPath: "sensitivityLabelSource",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "sensitivityLabelSource",
        defaultValue: 'recommended',
        type: {
            name: "String"
        }
    }
};
export var sensitivityLabelSource1 = {
    parameterPath: "sensitivityLabelSource",
    mapper: {
        required: true,
        serializedName: "sensitivityLabelSource",
        type: {
            name: "Enum",
            allowedValues: [
                "current",
                "recommended"
            ]
        }
    }
};
export var sensitivityLabelSource2 = {
    parameterPath: "sensitivityLabelSource",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "sensitivityLabelSource",
        defaultValue: 'current',
        type: {
            name: "String"
        }
    }
};
export var serverName = {
    parameterPath: "serverName",
    mapper: {
        required: true,
        serializedName: "serverName",
        type: {
            name: "String"
        }
    }
};
export var serviceObjectiveName = {
    parameterPath: "serviceObjectiveName",
    mapper: {
        required: true,
        serializedName: "serviceObjectiveName",
        type: {
            name: "String"
        }
    }
};
export var serviceTierAdvisorName = {
    parameterPath: "serviceTierAdvisorName",
    mapper: {
        required: true,
        serializedName: "serviceTierAdvisorName",
        type: {
            name: "String"
        }
    }
};
export var skip = {
    parameterPath: [
        "options",
        "skip"
    ],
    mapper: {
        serializedName: "$skip",
        type: {
            name: "Number"
        }
    }
};
export var skipToken = {
    parameterPath: [
        "options",
        "skipToken"
    ],
    mapper: {
        serializedName: "$skipToken",
        type: {
            name: "String"
        }
    }
};
export var startTime = {
    parameterPath: "startTime",
    mapper: {
        required: true,
        serializedName: "startTime",
        type: {
            name: "String"
        }
    }
};
export var stepName = {
    parameterPath: "stepName",
    mapper: {
        required: true,
        serializedName: "stepName",
        type: {
            name: "String"
        }
    }
};
export var subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        required: true,
        serializedName: "subscriptionId",
        type: {
            name: "String"
        }
    }
};
export var syncAgentName = {
    parameterPath: "syncAgentName",
    mapper: {
        required: true,
        serializedName: "syncAgentName",
        type: {
            name: "String"
        }
    }
};
export var syncGroupName = {
    parameterPath: "syncGroupName",
    mapper: {
        required: true,
        serializedName: "syncGroupName",
        type: {
            name: "String"
        }
    }
};
export var syncMemberName = {
    parameterPath: "syncMemberName",
    mapper: {
        required: true,
        serializedName: "syncMemberName",
        type: {
            name: "String"
        }
    }
};
export var tableName = {
    parameterPath: "tableName",
    mapper: {
        required: true,
        serializedName: "tableName",
        type: {
            name: "String"
        }
    }
};
export var targetGroupName = {
    parameterPath: "targetGroupName",
    mapper: {
        required: true,
        serializedName: "targetGroupName",
        type: {
            name: "String"
        }
    }
};
export var targetId = {
    parameterPath: "targetId",
    mapper: {
        required: true,
        serializedName: "targetId",
        type: {
            name: "Uuid"
        }
    }
};
export var top = {
    parameterPath: [
        "options",
        "top"
    ],
    mapper: {
        serializedName: "$top",
        type: {
            name: "Number"
        }
    }
};
export var transparentDataEncryptionName = {
    parameterPath: "transparentDataEncryptionName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "transparentDataEncryptionName",
        defaultValue: 'current',
        type: {
            name: "String"
        }
    }
};
export var type = {
    parameterPath: "type",
    mapper: {
        required: true,
        serializedName: "type",
        type: {
            name: "String"
        }
    }
};
export var usageName = {
    parameterPath: "usageName",
    mapper: {
        required: true,
        serializedName: "usageName",
        type: {
            name: "String"
        }
    }
};
export var virtualClusterName = {
    parameterPath: "virtualClusterName",
    mapper: {
        required: true,
        serializedName: "virtualClusterName",
        type: {
            name: "String"
        }
    }
};
export var virtualNetworkRuleName = {
    parameterPath: "virtualNetworkRuleName",
    mapper: {
        required: true,
        serializedName: "virtualNetworkRuleName",
        type: {
            name: "String"
        }
    }
};
export var vulnerabilityAssessmentName = {
    parameterPath: "vulnerabilityAssessmentName",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "vulnerabilityAssessmentName",
        defaultValue: 'default',
        type: {
            name: "String"
        }
    }
};
