// @microsoft/teamsfx-core/build/plugins/resource/frontend/resources/steps.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployProgress = exports.PreDeployProgress = exports.PostProvisionProgress = exports.ScaffoldProgress = void 0;
const messages_1 = require("../resources/messages");
exports.ScaffoldProgress = {
    title: messages_1.Messages.ScaffoldProgressTitle,
    steps: {
        Scaffold: messages_1.Messages.ProgressScaffold,
    },
};
exports.PostProvisionProgress = {
    title: messages_1.Messages.PostProvisionProgressTitle,
    steps: {
        EnableStaticWebsite: messages_1.Messages.ProgressEnableStorageStaticWebsite,
    },
};
exports.PreDeployProgress = {
    title: messages_1.Messages.PreDeployProgressTitle,
    steps: {
        CheckStorage: messages_1.Messages.ProgressCheckStorage,
    },
};
exports.DeployProgress = {
    title: messages_1.Messages.DeployProgressTitle,
    steps: {
        NPMInstall: messages_1.Messages.ProgressNPMInstall,
        Build: messages_1.Messages.ProgressBuild,
        getSrcAndDest: messages_1.Messages.ProgressGetSrcAndDest,
        Clear: messages_1.Messages.ProgressClear,
        Upload: messages_1.Messages.ProgressUpload,
    },
};
