// @microsoft/teamsfx-core/build/plugins/resource/bot/v2/enum.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.HostType = exports.BicepModules = exports.Runtime = exports.ProgrammingLanguage = void 0;
var ProgrammingLanguage;
(function (ProgrammingLanguage) {
    ProgrammingLanguage["Js"] = "js";
    ProgrammingLanguage["Ts"] = "ts";
    ProgrammingLanguage["Csharp"] = "csharp";
})(ProgrammingLanguage = exports.ProgrammingLanguage || (exports.ProgrammingLanguage = {}));
var Runtime;
(function (Runtime) {
    Runtime["Node"] = "node";
    Runtime["Dotnet"] = "dotnet";
})(Runtime = exports.Runtime || (exports.Runtime = {}));
var BicepModules;
(function (BicepModules) {
    BicepModules["Functions"] = "botFunction";
})(BicepModules = exports.BicepModules || (exports.BicepModules = {}));
var HostType;
(function (HostType) {
    HostType["AppService"] = "app-service";
    HostType["Functions"] = "azure-functions";
})(HostType = exports.HostType || (exports.HostType = {}));
