// @microsoft/teamsfx-core/build/plugins/resource/bot/v2/mapping.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectFileName = exports.getTriggerScenarios = exports.getLanguage = exports.getServiceType = exports.getRuntime = exports.getPlatformRuntime = exports.moduleMap = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const interfaces_1 = require("../../../../common/azure-hosting/interfaces");
const constants_1 = require("../constants");
const question_1 = require("../question");
const enum_1 = require("./enum");
const runtimeMap = new Map([
    [enum_1.ProgrammingLanguage.Js, enum_1.Runtime.Node],
    [enum_1.ProgrammingLanguage.Ts, enum_1.Runtime.Node],
    [enum_1.ProgrammingLanguage.Csharp, enum_1.Runtime.Dotnet],
]);
const serviceMap = new Map([
    [enum_1.HostType.AppService, interfaces_1.ServiceType.AppService],
    [enum_1.HostType.Functions, interfaces_1.ServiceType.Functions],
]);
const langMap = new Map([
    ["javascript", enum_1.ProgrammingLanguage.Js],
    ["typescript", enum_1.ProgrammingLanguage.Ts],
    ["csharp", enum_1.ProgrammingLanguage.Csharp],
]);
const triggerScenariosMap = new Map([
    [
        question_1.FunctionsHttpTriggerOptionItem.id,
        [
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_BASE_SCENARIO_NAME,
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_TRIGGER_HTTP_SCENARIO_NAME,
        ],
    ],
    [
        question_1.FunctionsTimerTriggerOptionItem.id,
        [
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_BASE_SCENARIO_NAME,
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_TRIGGER_TIMER_SCENARIO_NAME,
        ],
    ],
    [
        question_1.FunctionsHttpAndTimerTriggerOptionItem.id,
        [
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_BASE_SCENARIO_NAME,
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_TRIGGER_HTTP_SCENARIO_NAME,
            constants_1.TemplateProjectsScenarios.NOTIFICATION_FUNCTION_TRIGGER_TIMER_SCENARIO_NAME,
        ],
    ],
    [question_1.AppServiceOptionItem.id, [constants_1.TemplateProjectsScenarios.NOTIFICATION_RESTIFY_SCENARIO_NAME]],
    [question_1.AppServiceOptionItemForVS.id, [constants_1.TemplateProjectsScenarios.NOTIFICATION_WEBAPI_SCENARIO_NAME]],
]);
const PlatformRuntimeMap = new Map([
    [teamsfx_api_1.Platform.VS, enum_1.Runtime.Dotnet],
    [teamsfx_api_1.Platform.VSCode, enum_1.Runtime.Node],
    [teamsfx_api_1.Platform.CLI, enum_1.Runtime.Node],
    [teamsfx_api_1.Platform.CLI_HELP, enum_1.Runtime.Node],
]);
function getKeyNotFoundInMapErrorMsg(key) {
    return `The key ${key} is not found in map.`;
}
const projectFileMap = new Map([
    [enum_1.Runtime.Node, (_) => "package.json"],
    [enum_1.Runtime.Dotnet, (appName) => `${appName}.csproj`],
]);
exports.moduleMap = {
    [interfaces_1.ServiceType.Functions]: enum_1.BicepModules.Functions,
};
function getPlatformRuntime(platform) {
    const runtime = PlatformRuntimeMap.get(platform);
    if (runtime) {
        return runtime;
    }
    throw new Error(getKeyNotFoundInMapErrorMsg(platform));
}
exports.getPlatformRuntime = getPlatformRuntime;
function getRuntime(lang) {
    const runtime = runtimeMap.get(lang);
    if (runtime) {
        return runtime;
    }
    throw new Error(getKeyNotFoundInMapErrorMsg(lang));
}
exports.getRuntime = getRuntime;
function getServiceType(hostType) {
    const serviceType = serviceMap.get(hostType !== null && hostType !== void 0 ? hostType : "");
    if (serviceType) {
        return serviceType;
    }
    throw new Error(getKeyNotFoundInMapErrorMsg(hostType));
}
exports.getServiceType = getServiceType;
function getLanguage(lang) {
    var _a;
    const language = langMap.get((_a = lang === null || lang === void 0 ? void 0 : lang.toLowerCase()) !== null && _a !== void 0 ? _a : "");
    if (language) {
        return language;
    }
    throw new Error(getKeyNotFoundInMapErrorMsg(lang));
}
exports.getLanguage = getLanguage;
function getTriggerScenarios(trigger) {
    const scenarios = triggerScenariosMap.get(trigger);
    if (scenarios) {
        return scenarios;
    }
    throw new Error(getKeyNotFoundInMapErrorMsg(trigger));
}
exports.getTriggerScenarios = getTriggerScenarios;
function getProjectFileName(runtime, appName) {
    const projectFileName = projectFileMap.get(runtime);
    if (projectFileName) {
        return projectFileName(appName);
    }
    throw new Error(getKeyNotFoundInMapErrorMsg(runtime));
}
exports.getProjectFileName = getProjectFileName;
