// @microsoft/teamsfx-core/build/plugins/resource/bot/errors.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.runWithExceptionCatching = exports.wrapError = exports.RegisterResourceProviderError = exports.CommandExecutionError = exports.MessageEndpointUpdatingError = exports.PackDirExistenceError = exports.ConfigValidationError = exports.ConfigUpdatingError = exports.UnzipError = exports.ProvisionError = exports.TemplateZipFallbackError = exports.CreateAADSecretError = exports.CreateAADAppError = exports.AADAppCheckingError = exports.checkPrecondition = exports.checkAndThrowIfMissing = exports.SomethingMissingError = exports.PreconditionError = exports.PluginError = exports.isPluginError = exports.isErrorWithMessage = exports.isHttpError = exports.ErrorType = void 0;
const errorCodes_1 = require("../aad/errorCodes");
const errors_1 = require("../aad/errors");
const constants_1 = require("./constants");
const messages_1 = require("./resources/messages");
const localizeUtils_1 = require("../../../common/localizeUtils");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const result_1 = require("./result");
const logger_1 = require("./logger");
const telemetry_helper_1 = require("./utils/telemetry-helper");
const hostingError_1 = require("../../../common/azure-hosting/hostingError");
const progressBars_1 = require("./progressBars");
exports.ErrorType = {
    USER: "User",
    SYSTEM: "System",
};
function isHttpError(e) {
    return e instanceof Object && "response" in e;
}
exports.isHttpError = isHttpError;
function isErrorWithMessage(e) {
    return e instanceof Object && "message" in e;
}
exports.isErrorWithMessage = isErrorWithMessage;
function isPluginError(e) {
    return e instanceof Object && "innerError" in e;
}
exports.isPluginError = isPluginError;
function resolveInnerError(target, helpLinkMap) {
    var _a, _b, _c, _d;
    if (!target.innerError)
        return;
    const statusCode = isHttpError(target.innerError) ? (_a = target.innerError.response) === null || _a === void 0 ? void 0 : _a.status : 500;
    if (statusCode) {
        if (statusCode >= 400 && statusCode < 500) {
            target.errorType = exports.ErrorType.USER;
        }
        else {
            target.errorType = exports.ErrorType.SYSTEM;
        }
    }
    if (isHttpError(target.innerError)) {
        const errorCode = (_d = (_c = (_b = target.innerError.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.error) === null || _d === void 0 ? void 0 : _d.code;
        if (errorCode) {
            const helpLink = helpLinkMap.get(errorCode);
            if (helpLink)
                target.helpLink = helpLink;
        }
    }
}
class PluginError extends Error {
    constructor(type, name, details, suggestions, innerError, helpLink) {
        super(details[0]);
        this.name = name;
        this.details = details;
        this.suggestions = suggestions;
        this.errorType = type;
        this.innerError = innerError;
        this.helpLink = helpLink;
        Object.setPrototypeOf(this, PluginError.prototype);
    }
    genMessage() {
        let msg = `${this.details[0]} `;
        if (this.suggestions.length > 0) {
            msg += localizeUtils_1.getDefaultString("plugins.bot.ErrorSuggestions", this.suggestions.join(" "));
        }
        return msg;
    }
    genDisplayMessage() {
        let msg = `${this.details[1]} `;
        if (this.suggestions.length > 0) {
            msg += localizeUtils_1.getLocalizedString("plugins.bot.ErrorSuggestions", this.suggestions.join(" "));
        }
        return msg;
    }
}
exports.PluginError = PluginError;
class PreconditionError extends PluginError {
    constructor(message, suggestions) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.PRECONDITION_ERROR, message, suggestions);
    }
}
exports.PreconditionError = PreconditionError;
class SomethingMissingError extends PreconditionError {
    constructor(something) {
        super(messages_1.Messages.SomethingIsMissing(something), [messages_1.Messages.RetryTheCurrentStep]);
    }
}
exports.SomethingMissingError = SomethingMissingError;
function checkAndThrowIfMissing(name, value) {
    if (!value) {
        throw new SomethingMissingError(name);
    }
    return value;
}
exports.checkAndThrowIfMissing = checkAndThrowIfMissing;
function checkPrecondition(message, value) {
    if (!value) {
        throw new PreconditionError(message, []);
    }
    return value;
}
exports.checkPrecondition = checkPrecondition;
class AADAppCheckingError extends PluginError {
    constructor(innerError) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.CALL_APPSTUDIO_API_ERROR, messages_1.Messages.FailToCallAppStudioForCheckingAADApp, [messages_1.Messages.RetryTheCurrentStep], innerError);
    }
}
exports.AADAppCheckingError = AADAppCheckingError;
class CreateAADAppError extends PluginError {
    constructor(innerError) {
        super(exports.ErrorType.USER, errors_1.CreateAppError.name, errors_1.CreateAppError.message(), [], innerError);
        resolveInnerError(this, errorCodes_1.GraphErrorCodes);
    }
}
exports.CreateAADAppError = CreateAADAppError;
class CreateAADSecretError extends PluginError {
    constructor(innerError) {
        super(exports.ErrorType.USER, errors_1.CreateSecretError.name, errors_1.CreateSecretError.message(), [], innerError);
        resolveInnerError(this, errorCodes_1.GraphErrorCodes);
    }
}
exports.CreateAADSecretError = CreateAADSecretError;
class TemplateZipFallbackError extends PluginError {
    constructor() {
        super(exports.ErrorType.USER, "TemplateZipFallbackError", [
            localizeUtils_1.getDefaultString("plugins.bot.TemplateZipFallbackError"),
            localizeUtils_1.getLocalizedString("plugins.bot.TemplateZipFallbackError"),
        ], [messages_1.Messages.CheckOutputLogAndTryToFix, messages_1.Messages.RetryTheCurrentStep]);
    }
}
exports.TemplateZipFallbackError = TemplateZipFallbackError;
class ProvisionError extends PluginError {
    constructor(resource, innerError) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.PROVISION_ERROR, messages_1.Messages.FailToProvisionSomeResource(resource), [messages_1.Messages.CheckOutputLogAndTryToFix, messages_1.Messages.RetryTheCurrentStep], innerError);
    }
}
exports.ProvisionError = ProvisionError;
class UnzipError extends PluginError {
    constructor(path) {
        super(exports.ErrorType.USER, "UnzipError", [localizeUtils_1.getDefaultString("plugins.bot.UnzipError"), localizeUtils_1.getLocalizedString("plugins.bot.UnzipError")], [
            messages_1.Messages.CheckOutputLogAndTryToFix,
            messages_1.Messages.ReopenWorkingDir(path),
            messages_1.Messages.RetryTheCurrentStep,
        ]);
    }
}
exports.UnzipError = UnzipError;
class ConfigUpdatingError extends PluginError {
    constructor(configName, innerError) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.CONFIG_UPDATING_ERROR, messages_1.Messages.FailToUpdateConfigs(configName), [messages_1.Messages.CheckOutputLogAndTryToFix, messages_1.Messages.RetryTheCurrentStep], innerError);
    }
}
exports.ConfigUpdatingError = ConfigUpdatingError;
class ConfigValidationError extends PluginError {
    constructor(name, value) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.CONFIG_VALIDATION_ERROR, messages_1.Messages.SomethingIsInvalidWithValue(name, value), [messages_1.Messages.RecoverConfig, messages_1.Messages.RecreateTheProject[1]]);
    }
}
exports.ConfigValidationError = ConfigValidationError;
class PackDirExistenceError extends PluginError {
    constructor() {
        super(exports.ErrorType.USER, constants_1.ErrorNames.PACK_DIR_EXISTENCE_ERROR, messages_1.Messages.SomethingIsNotExisting("pack directory"), [messages_1.Messages.RecreateTheProject[1]]);
    }
}
exports.PackDirExistenceError = PackDirExistenceError;
class MessageEndpointUpdatingError extends PluginError {
    constructor(endpoint, innerError) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.MSG_ENDPOINT_UPDATING_ERROR, messages_1.Messages.FailToUpdateMessageEndpoint(endpoint), [messages_1.Messages.CheckOutputLogAndTryToFix, messages_1.Messages.RetryTheCurrentStep], innerError);
    }
}
exports.MessageEndpointUpdatingError = MessageEndpointUpdatingError;
class CommandExecutionError extends PluginError {
    constructor(cmd, path, innerError) {
        super(exports.ErrorType.USER, constants_1.ErrorNames.COMMAND_EXECUTION_ERROR, messages_1.Messages.CommandExecutionFailed(cmd), [
            messages_1.Messages.RunFailedCommand(cmd, path),
            messages_1.Messages.CheckCommandOutputAndTryToFixIt,
            messages_1.Messages.RetryTheCurrentStep,
        ], innerError);
    }
}
exports.CommandExecutionError = CommandExecutionError;
class RegisterResourceProviderError extends PluginError {
    constructor(innerError) {
        super(exports.ErrorType.USER, "RegisterResourceProviderError", [
            localizeUtils_1.getDefaultString("plugins.bot.RegisterResourceProviderError"),
            localizeUtils_1.getLocalizedString("plugins.bot.RegisterResourceProviderError"),
        ], [
            messages_1.Messages.RegisterRequiredRP(constants_1.AzureConstants.requiredResourceProviders),
            messages_1.Messages.CheckOutputLogAndTryToFix,
        ], innerError);
    }
}
exports.RegisterResourceProviderError = RegisterResourceProviderError;
//! context and name are only for telemetry, they may be empty if sendTelemetry is false
function wrapError(e, context, sendTelemetry, name) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    let errorMsg = isErrorWithMessage(e) ? e.message : "";
    const innerError = isPluginError(e) ? e.innerError : undefined;
    if (innerError) {
        errorMsg += localizeUtils_1.getLocalizedString("plugins.bot.DetailedError", isErrorWithMessage(innerError) ? innerError.message : "");
        if (isHttpError(innerError)) {
            if ((_b = (_a = innerError.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.errorMessage) {
                errorMsg += localizeUtils_1.getLocalizedString("plugins.bot.DetailedErrorReason", (_d = (_c = innerError.response) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d.errorMessage);
            }
            else if ((_g = (_f = (_e = innerError.response) === null || _e === void 0 ? void 0 : _e.data) === null || _f === void 0 ? void 0 : _f.error) === null || _g === void 0 ? void 0 : _g.message) {
                // For errors return from Graph API
                errorMsg += localizeUtils_1.getLocalizedString("plugins.bot.DetailedErrorReason", (_k = (_j = (_h = innerError.response) === null || _h === void 0 ? void 0 : _h.data) === null || _j === void 0 ? void 0 : _j.error) === null || _k === void 0 ? void 0 : _k.message);
            }
            else if ((_m = (_l = innerError.response) === null || _l === void 0 ? void 0 : _l.data) === null || _m === void 0 ? void 0 : _m.errors) {
                // For errors return from App Studio API
                errorMsg += localizeUtils_1.getLocalizedString("plugins.bot.DetailedErrorReason", JSON.stringify((_p = (_o = innerError.response) === null || _o === void 0 ? void 0 : _o.data) === null || _p === void 0 ? void 0 : _p.errors));
            }
        }
    }
    logger_1.Logger.error(errorMsg);
    if (e instanceof teamsfx_api_1.UserError || e instanceof teamsfx_api_1.SystemError) {
        const res = teamsfx_api_1.err(e);
        sendTelemetry && telemetry_helper_1.telemetryHelper.sendResultEvent(context, name, res);
        return res;
    }
    if (e instanceof PluginError || e instanceof hostingError_1.CommonHostingError) {
        const message = e.genMessage() + errorMsg;
        const displayMessage = e.genDisplayMessage() + errorMsg;
        const result = e instanceof PluginError && e.errorType === exports.ErrorType.SYSTEM
            ? result_1.FxBotPluginResultFactory.SystemError(e.name, [message, displayMessage], e.innerError)
            : result_1.FxBotPluginResultFactory.UserError(e.name, [message, displayMessage], e.innerError, e instanceof PluginError ? e.helpLink : "");
        sendTelemetry && telemetry_helper_1.telemetryHelper.sendResultEvent(context, name, result);
        return result;
    }
    else {
        // Unrecognized Exception.
        const UnhandledErrorCode = "UnhandledError";
        sendTelemetry &&
            telemetry_helper_1.telemetryHelper.sendResultEvent(context, name, result_1.FxBotPluginResultFactory.SystemError(UnhandledErrorCode, [
                localizeUtils_1.getDefaultString("plugins.bot.UnhandledError", errorMsg),
                localizeUtils_1.getLocalizedString("plugins.bot.UnhandledError", errorMsg),
            ], isPluginError(e) ? e.innerError : undefined));
        return result_1.FxBotPluginResultFactory.SystemError(UnhandledErrorCode, [errorMsg, errorMsg], innerError);
    }
}
exports.wrapError = wrapError;
async function runWithExceptionCatching(context, fn, sendTelemetry, name) {
    try {
        sendTelemetry && telemetry_helper_1.telemetryHelper.sendStartEvent(context, name);
        const res = await fn();
        sendTelemetry && telemetry_helper_1.telemetryHelper.sendResultEvent(context, name, res);
        return res;
    }
    catch (e) {
        await progressBars_1.ProgressBarFactory.closeProgressBar(false); // Close all progress bars.
        return wrapError(e, context, sendTelemetry, name);
    }
}
exports.runWithExceptionCatching = runWithExceptionCatching;
