// @microsoft/teamsfx-core/build/plugins/resource/bot/configs/localDebugConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalDebugConfig = void 0;
const localSettingsConstants_1 = require("../../../../common/localSettingsConstants");
const strings_1 = require("../resources/strings");
class LocalDebugConfig {
    botAADCreated() {
        if (this.localBotId && this.localBotPassword) {
            return true;
        }
        return false;
    }
    async restoreConfigFromContext(context) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
        this.localEndpoint = (_b = (_a = context.localSettings) === null || _a === void 0 ? void 0 : _a.bot) === null || _b === void 0 ? void 0 : _b.get(localSettingsConstants_1.LocalSettingsBotKeys.BotEndpoint);
        this.localRedirectUri = this.localEndpoint
            ? `${this.localEndpoint}${strings_1.CommonStrings.AUTH_REDIRECT_URI_SUFFIX}`
            : undefined;
        this.localBotId = (_d = (_c = context.localSettings) === null || _c === void 0 ? void 0 : _c.bot) === null || _d === void 0 ? void 0 : _d.get(localSettingsConstants_1.LocalSettingsBotKeys.BotId);
        this.localBotPassword = (_f = (_e = context.localSettings) === null || _e === void 0 ? void 0 : _e.bot) === null || _f === void 0 ? void 0 : _f.get(localSettingsConstants_1.LocalSettingsBotKeys.BotPassword);
        this.localObjectId = (_h = (_g = context.localSettings) === null || _g === void 0 ? void 0 : _g.bot) === null || _h === void 0 ? void 0 : _h.get(localSettingsConstants_1.LocalSettingsBotKeys.BotAadObjectId);
        this.localRedirectUri = (_k = (_j = context.localSettings) === null || _j === void 0 ? void 0 : _j.bot) === null || _k === void 0 ? void 0 : _k.get(localSettingsConstants_1.LocalSettingsBotKeys.BotRedirectUri);
        // To respect existing Bot AAD.
        this.localBotId = (_m = (_l = context.envInfo.config.bot) === null || _l === void 0 ? void 0 : _l.appId) !== null && _m !== void 0 ? _m : this.localBotId;
        this.localBotPassword = (_p = (_o = context.envInfo.config.bot) === null || _o === void 0 ? void 0 : _o.appPassword) !== null && _p !== void 0 ? _p : this.localBotPassword;
    }
    saveConfigIntoContext(context) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        (_b = (_a = context.localSettings) === null || _a === void 0 ? void 0 : _a.bot) === null || _b === void 0 ? void 0 : _b.set(localSettingsConstants_1.LocalSettingsBotKeys.BotId, this.localBotId);
        (_d = (_c = context.localSettings) === null || _c === void 0 ? void 0 : _c.bot) === null || _d === void 0 ? void 0 : _d.set(localSettingsConstants_1.LocalSettingsBotKeys.BotPassword, this.localBotPassword);
        (_f = (_e = context.localSettings) === null || _e === void 0 ? void 0 : _e.bot) === null || _f === void 0 ? void 0 : _f.set(localSettingsConstants_1.LocalSettingsBotKeys.BotAadObjectId, this.localObjectId);
        (_h = (_g = context.localSettings) === null || _g === void 0 ? void 0 : _g.bot) === null || _h === void 0 ? void 0 : _h.set(localSettingsConstants_1.LocalSettingsBotKeys.BotRedirectUri, this.localRedirectUri);
    }
}
exports.LocalDebugConfig = LocalDebugConfig;
