// @microsoft/teamsfx-core/build/component/resource/keyVault.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyVaultResource = void 0;
const tslib_1 = require("tslib");
require("reflect-metadata");
const typedi_1 = require("typedi");
const constants_1 = require("../constants");
const azureResource_1 = require("./azureResource");
let KeyVaultResource = class KeyVaultResource extends azureResource_1.AzureResource {
    constructor() {
        super(...arguments);
        this.name = constants_1.ComponentNames.KeyVault;
        this.bicepModuleName = "keyVault";
        this.outputs = constants_1.KeyVaultOutputs;
        this.finalOutputKeys = ["keyVaultResourceId", "m365ClientSecretReference", "botClientSecretReference"];
        this.templateContext = {
            identity: {
                principalId: constants_1.IdentityOutputs.identityPrincipalId.bicepVariable,
            },
        };
    }
};
KeyVaultResource = tslib_1.__decorate([
    typedi_1.Service(constants_1.ComponentNames.KeyVault)
], KeyVaultResource);
exports.KeyVaultResource = KeyVaultResource;
