// @azure/arm-apimanagement/esm/operations/gatewayHostnameConfiguration.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/gatewayHostnameConfigurationMappers";
import * as Parameters from "../models/parameters";
/** Class representing a GatewayHostnameConfiguration. */
var GatewayHostnameConfiguration = /** @class */ (function () {
    /**
     * Create a GatewayHostnameConfiguration.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function GatewayHostnameConfiguration(client) {
        this.client = client;
    }
    GatewayHostnameConfiguration.prototype.listByService = function (resourceGroupName, serviceName, gatewayId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            options: options
        }, listByServiceOperationSpec, callback);
    };
    GatewayHostnameConfiguration.prototype.getEntityTag = function (resourceGroupName, serviceName, gatewayId, hcId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            hcId: hcId,
            options: options
        }, getEntityTagOperationSpec, callback);
    };
    GatewayHostnameConfiguration.prototype.get = function (resourceGroupName, serviceName, gatewayId, hcId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            hcId: hcId,
            options: options
        }, getOperationSpec, callback);
    };
    GatewayHostnameConfiguration.prototype.createOrUpdate = function (resourceGroupName, serviceName, gatewayId, hcId, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            hcId: hcId,
            parameters: parameters,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    GatewayHostnameConfiguration.prototype.deleteMethod = function (resourceGroupName, serviceName, gatewayId, hcId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            hcId: hcId,
            options: options
        }, deleteMethodOperationSpec, callback);
    };
    GatewayHostnameConfiguration.prototype.listByServiceNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByServiceNextOperationSpec, callback);
    };
    return GatewayHostnameConfiguration;
}());
export { GatewayHostnameConfiguration };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByServiceOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/hostnameConfigurations",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.top,
        Parameters.skip,
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.GatewayHostnameConfigurationCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var getEntityTagOperationSpec = {
    httpMethod: "HEAD",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/hostnameConfigurations/{hcId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.hcId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            headersMapper: Mappers.GatewayHostnameConfigurationGetEntityTagHeaders
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/hostnameConfigurations/{hcId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.hcId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.GatewayHostnameConfigurationContract,
            headersMapper: Mappers.GatewayHostnameConfigurationGetHeaders
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/hostnameConfigurations/{hcId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.hcId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.GatewayHostnameConfigurationContract), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.GatewayHostnameConfigurationContract,
            headersMapper: Mappers.GatewayHostnameConfigurationCreateOrUpdateHeaders
        },
        201: {
            bodyMapper: Mappers.GatewayHostnameConfigurationContract,
            headersMapper: Mappers.GatewayHostnameConfigurationCreateOrUpdateHeaders
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/hostnameConfigurations/{hcId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.hcId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var listByServiceNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.GatewayHostnameConfigurationCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
