// @azure/arm-apimanagement/esm/operations/gatewayApi.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/gatewayApiMappers";
import * as Parameters from "../models/parameters";
/** Class representing a GatewayApi. */
var GatewayApi = /** @class */ (function () {
    /**
     * Create a GatewayApi.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function GatewayApi(client) {
        this.client = client;
    }
    GatewayApi.prototype.listByService = function (resourceGroupName, serviceName, gatewayId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            options: options
        }, listByServiceOperationSpec, callback);
    };
    GatewayApi.prototype.getEntityTag = function (resourceGroupName, serviceName, gatewayId, apiId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            apiId: apiId,
            options: options
        }, getEntityTagOperationSpec, callback);
    };
    GatewayApi.prototype.createOrUpdate = function (resourceGroupName, serviceName, gatewayId, apiId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            apiId: apiId,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    GatewayApi.prototype.deleteMethod = function (resourceGroupName, serviceName, gatewayId, apiId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            gatewayId: gatewayId,
            apiId: apiId,
            options: options
        }, deleteMethodOperationSpec, callback);
    };
    GatewayApi.prototype.listByServiceNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByServiceNextOperationSpec, callback);
    };
    return GatewayApi;
}());
export { GatewayApi };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByServiceOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/apis",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.top,
        Parameters.skip,
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApiCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var getEntityTagOperationSpec = {
    httpMethod: "HEAD",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/apis/{apiId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.apiId1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            headersMapper: Mappers.GatewayApiGetEntityTagHeaders
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/apis/{apiId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.apiId1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: [
            "options",
            "parameters"
        ],
        mapper: Mappers.AssociationContract
    },
    responses: {
        200: {
            bodyMapper: Mappers.ApiContract
        },
        201: {
            bodyMapper: Mappers.ApiContract
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/gateways/{gatewayId}/apis/{apiId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.gatewayId,
        Parameters.apiId1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var listByServiceNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApiCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
