// @azure/arm-storage/esm/operations/objectReplicationPoliciesOperations.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/objectReplicationPoliciesOperationsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ObjectReplicationPoliciesOperations. */
var ObjectReplicationPoliciesOperations = /** @class */ (function () {
    /**
     * Create a ObjectReplicationPoliciesOperations.
     * @param {StorageManagementClientContext} client Reference to the service client.
     */
    function ObjectReplicationPoliciesOperations(client) {
        this.client = client;
    }
    ObjectReplicationPoliciesOperations.prototype.list = function (resourceGroupName, accountName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            accountName: accountName,
            options: options
        }, listOperationSpec, callback);
    };
    ObjectReplicationPoliciesOperations.prototype.get = function (resourceGroupName, accountName, objectReplicationPolicyId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            accountName: accountName,
            objectReplicationPolicyId: objectReplicationPolicyId,
            options: options
        }, getOperationSpec, callback);
    };
    ObjectReplicationPoliciesOperations.prototype.createOrUpdate = function (resourceGroupName, accountName, objectReplicationPolicyId, properties, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            accountName: accountName,
            objectReplicationPolicyId: objectReplicationPolicyId,
            properties: properties,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    ObjectReplicationPoliciesOperations.prototype.deleteMethod = function (resourceGroupName, accountName, objectReplicationPolicyId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            accountName: accountName,
            objectReplicationPolicyId: objectReplicationPolicyId,
            options: options
        }, deleteMethodOperationSpec, callback);
    };
    return ObjectReplicationPoliciesOperations;
}());
export { ObjectReplicationPoliciesOperations };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ObjectReplicationPolicies
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies/{objectReplicationPolicyId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.objectReplicationPolicyId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ObjectReplicationPolicy
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies/{objectReplicationPolicyId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.objectReplicationPolicyId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "properties",
        mapper: __assign(__assign({}, Mappers.ObjectReplicationPolicy), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ObjectReplicationPolicy
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies/{objectReplicationPolicyId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.objectReplicationPolicyId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
