// @azure/arm-storage/esm/operations/privateLinkResources.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/privateLinkResourcesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a PrivateLinkResources. */
var PrivateLinkResources = /** @class */ (function () {
    /**
     * Create a PrivateLinkResources.
     * @param {StorageManagementClientContext} client Reference to the service client.
     */
    function PrivateLinkResources(client) {
        this.client = client;
    }
    PrivateLinkResources.prototype.listByStorageAccount = function (resourceGroupName, accountName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            accountName: accountName,
            options: options
        }, listByStorageAccountOperationSpec, callback);
    };
    return PrivateLinkResources;
}());
export { PrivateLinkResources };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByStorageAccountOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/privateLinkResources",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.PrivateLinkResourceListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
