// @microsoft/teamsfx-core/build/component/feature/sso.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.canAddSso = exports.SSO = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
require("reflect-metadata");
const typedi_1 = require("typedi");
const localizeUtils_1 = require("../../common/localizeUtils");
const utils_1 = require("../../common/utils");
const telemetry_1 = require("../../core/telemetry");
const plugins_1 = require("../../plugins");
require("../connection/azureWebAppConfig");
const constants_1 = require("../constants");
const debug_1 = require("../debug");
const utils_2 = require("../resource/appManifest/utils");
require("../resource/azureSql");
require("../resource/identity");
const utils_3 = require("../utils");
const workflow_1 = require("../workflow");
let SSO = class SSO {
    constructor() {
        this.name = "sso";
    }
    async add(context, inputs) {
        context.telemetryReporter.sendTelemetryEvent(plugins_1.SolutionTelemetryEvent.AddSsoStart, {
            [plugins_1.SolutionTelemetryProperty.Component]: plugins_1.SolutionTelemetryComponentName,
        });
        const updates = getUpdateComponents(context.projectSetting, inputs.stage === teamsfx_api_1.Stage.create);
        // generate manifest
        const aadApp = typedi_1.Container.get(constants_1.ComponentNames.AadApp);
        {
            const res = await aadApp.generateManifest(context, inputs);
            if (res.isErr()) {
                return teamsfx_api_1.err(telemetry_1.sendErrorTelemetryThenReturnError(plugins_1.SolutionTelemetryEvent.AddSso, res.error, context.telemetryReporter));
            }
        }
        // config sso
        if (updates.aad) {
            context.projectSetting.components.push({
                name: constants_1.ComponentNames.AadApp,
                provision: true,
                deploy: true,
            });
        }
        if (updates.tab) {
            const teamsTabComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.TeamsTab);
            teamsTabComponent.sso = true;
        }
        if (updates.bot) {
            const teamsBotComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.TeamsBot);
            teamsBotComponent.sso = true;
        }
        // generate bicep
        {
            const res = await aadApp.generateBicep(context, inputs);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
            const bicepRes = await utils_3.bicepUtils.persistBiceps(inputs.projectPath, utils_1.convertToAlphanumericOnly(context.projectSetting.appName), res.value);
            if (bicepRes.isErr()) {
                return teamsfx_api_1.err(telemetry_1.sendErrorTelemetryThenReturnError(plugins_1.SolutionTelemetryEvent.AddSso, bicepRes.error, context.telemetryReporter));
            }
        }
        // generate auth files
        if (inputs.stage === teamsfx_api_1.Stage.addFeature &&
            inputs[plugins_1.AzureSolutionQuestionNames.Features] !== plugins_1.TabOptionItem.id) {
            const isExistingTabAppRes = await utils_2.manifestUtils.isExistingTab(inputs, context);
            if (isExistingTabAppRes.isErr())
                return teamsfx_api_1.err(isExistingTabAppRes.error);
            const res = await aadApp.generateAuthFiles(context, inputs, updates.tab || isExistingTabAppRes.value, updates.bot);
            if (res.isErr()) {
                return teamsfx_api_1.err(telemetry_1.sendErrorTelemetryThenReturnError(plugins_1.SolutionTelemetryEvent.AddSso, res.error, context.telemetryReporter));
            }
        }
        // update app manifest
        {
            const capabilities = [{ name: "WebApplicationInfo" }];
            const appManifest = typedi_1.Container.get(constants_1.ComponentNames.AppManifest);
            const res = await appManifest.addCapability(inputs, capabilities);
            if (res.isErr()) {
                return teamsfx_api_1.err(telemetry_1.sendErrorTelemetryThenReturnError(plugins_1.SolutionTelemetryEvent.AddSso, res.error, context.telemetryReporter));
            }
        }
        // local debug settings
        {
            const res = await debug_1.generateLocalDebugSettings(context, inputs);
            if (res.isErr()) {
                return teamsfx_api_1.err(telemetry_1.sendErrorTelemetryThenReturnError(plugins_1.SolutionTelemetryEvent.AddSso, res.error, context.telemetryReporter));
            }
        }
        // generate config bicep
        {
            const res = await utils_3.generateConfigBiceps(context, inputs);
            if (res.isErr()) {
                return teamsfx_api_1.err(telemetry_1.sendErrorTelemetryThenReturnError(plugins_1.SolutionTelemetryEvent.AddSso, res.error, context.telemetryReporter));
            }
        }
        // show notification
        if (inputs.platform == teamsfx_api_1.Platform.VSCode) {
            context.userInteraction
                .showMessage("info", localizeUtils_1.getLocalizedString("core.addSso.learnMore", plugins_1.AddSsoParameters.LearnMore), false, plugins_1.AddSsoParameters.LearnMore)
                .then((result) => {
                var _a;
                const userSelected = result.isOk() ? result.value : undefined;
                if (userSelected === plugins_1.AddSsoParameters.LearnMore) {
                    (_a = context.userInteraction) === null || _a === void 0 ? void 0 : _a.openUrl(plugins_1.AddSsoParameters.LearnMoreUrl);
                    context.telemetryReporter.sendTelemetryEvent(plugins_1.SolutionTelemetryEvent.AddSsoReadme, {
                        [plugins_1.SolutionTelemetryProperty.Component]: plugins_1.SolutionTelemetryComponentName,
                    });
                }
            });
        }
        else if (inputs.platform == teamsfx_api_1.Platform.CLI) {
            await context.userInteraction.showMessage("info", localizeUtils_1.getLocalizedString("core.addSso.learnMore", plugins_1.AddSsoParameters.LearnMoreUrl), false);
        }
        context.telemetryReporter.sendTelemetryEvent(plugins_1.SolutionTelemetryEvent.AddSso, {
            [plugins_1.SolutionTelemetryProperty.Component]: plugins_1.SolutionTelemetryComponentName,
            [plugins_1.SolutionTelemetryProperty.Success]: constants_1.TelemetryConstants.values.yes,
            [plugins_1.SolutionTelemetryProperty.AddTabSso]: updates.tab
                ? constants_1.TelemetryConstants.values.yes
                : constants_1.TelemetryConstants.values.no,
            [plugins_1.SolutionTelemetryProperty.AddBotSso]: updates.bot
                ? constants_1.TelemetryConstants.values.yes
                : constants_1.TelemetryConstants.values.no,
        });
        return teamsfx_api_1.ok({
            func: plugins_1.AddSsoParameters.AddSso,
            capabilities: [
                ...(updates.tab ? [plugins_1.AddSsoParameters.Tab] : []),
                ...(updates.bot ? [plugins_1.AddSsoParameters.Bot] : []),
            ],
        });
    }
};
SSO = tslib_1.__decorate([
    typedi_1.Service("sso")
], SSO);
exports.SSO = SSO;
function getUpdateComponents(projectSetting, isCreateStage) {
    if (isCreateStage) {
        return {
            tab: true,
            aad: true,
        };
    }
    let needsBot = false;
    let needsTab = false;
    const aadComponent = workflow_1.getComponent(projectSetting, constants_1.ComponentNames.AadApp);
    const teamsBotComponent = workflow_1.getComponent(projectSetting, constants_1.ComponentNames.TeamsBot);
    if (teamsBotComponent && !teamsBotComponent.sso) {
        if (teamsBotComponent.capabilities &&
            teamsBotComponent.capabilities.length === 1 &&
            teamsBotComponent.capabilities.includes("message-extension")) {
            needsBot = false;
        }
        else {
            needsBot = teamsBotComponent.hosting !== constants_1.ComponentNames.Function;
        }
    }
    const teamsTabComponent = workflow_1.getComponent(projectSetting, constants_1.ComponentNames.TeamsTab);
    if (teamsTabComponent && !teamsTabComponent.sso) {
        needsTab = true;
    }
    return {
        bot: needsBot,
        tab: needsTab,
        aad: !aadComponent,
    };
}
function canAddSso(projectSettings, returnError = false) {
    // TODO: support existing tab app with sso in v3
    const update = getUpdateComponents(projectSettings, false);
    if (update.tab || update.bot || update.aad) {
        return true;
    }
    else {
        const aadComponent = workflow_1.getComponent(projectSettings, constants_1.ComponentNames.AadApp);
        const teamsBotComponent = workflow_1.getComponent(projectSettings, constants_1.ComponentNames.TeamsBot);
        if (teamsBotComponent) {
            if (teamsBotComponent.capabilities &&
                teamsBotComponent.capabilities.length === 1 &&
                teamsBotComponent.capabilities.includes("message-extension")) {
                return returnError
                    ? teamsfx_api_1.err(new teamsfx_api_1.SystemError(plugins_1.SolutionSource, plugins_1.SolutionError.AddSsoNotSupported, localizeUtils_1.getLocalizedString("core.addSso.onlyMeNotSupport")))
                    : false;
            }
            if (teamsBotComponent.hosting === constants_1.ComponentNames.Function) {
                return returnError
                    ? teamsfx_api_1.err(new teamsfx_api_1.SystemError(plugins_1.SolutionSource, plugins_1.SolutionError.AddSsoNotSupported, localizeUtils_1.getLocalizedString("core.addSso.functionNotSupport")))
                    : false;
            }
        }
        if (aadComponent) {
            return returnError
                ? teamsfx_api_1.err(new teamsfx_api_1.SystemError(plugins_1.SolutionSource, plugins_1.SolutionError.SsoEnabled, localizeUtils_1.getLocalizedString("core.addSso.ssoEnabled")))
                : false;
        }
        return false;
    }
}
exports.canAddSso = canAddSso;
