// @microsoft/teamsfx-core/build/common/deps-checker/constant/helpLink.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ngrokInstallHelpLink = exports.dotnetNotSupportTargetVersionHelpLink = exports.dotnetManualInstallHelpLink = exports.dotnetFailToInstallHelpLink = exports.dotnetExplanationHelpLink = exports.nodeNotSupportedForSPFxHelpLink = exports.nodeNotSupportedForFunctionsHelpLink = exports.nodeNotSupportedForAzureHelpLink = exports.nodeNotFoundHelpLink = exports.functionDepsVersionsLink = exports.validationSettingsHelpLink = exports.defaultHelpLink = void 0;
exports.defaultHelpLink = "https://aka.ms/teamsfx-envchecker-help";
exports.validationSettingsHelpLink = `${exports.defaultHelpLink}#teams-toolkit-validation-settings`;
exports.functionDepsVersionsLink = "https://aka.ms/functions-node-versions";
exports.nodeNotFoundHelpLink = `${exports.defaultHelpLink}#nodenotfound`;
exports.nodeNotSupportedForAzureHelpLink = `${exports.defaultHelpLink}#nodenotsupported-azure-hosting`;
exports.nodeNotSupportedForFunctionsHelpLink = `${exports.defaultHelpLink}#nodenotsupported-azure-functions`;
exports.nodeNotSupportedForSPFxHelpLink = `${exports.defaultHelpLink}#nodenotsupported-spfx-hosting`;
exports.dotnetExplanationHelpLink = `${exports.defaultHelpLink}#how-to-install-net-sdk`;
exports.dotnetFailToInstallHelpLink = `${exports.defaultHelpLink}#failtoinstalldotnet`;
exports.dotnetManualInstallHelpLink = `${exports.defaultHelpLink}#dotnetnotfound`;
exports.dotnetNotSupportTargetVersionHelpLink = `${exports.defaultHelpLink}#dotnetnotsupporttargetversion`;
exports.ngrokInstallHelpLink = `${exports.defaultHelpLink}#failtoinstallngrok`;
