// @microsoft/teamsfx-core/build/plugins/resource/frontend/error-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorFactory = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("./constants");
class ErrorFactory {
    static UserError(name, message, innerError, stack, helpLink) {
        return new teamsfx_api_1.UserError({
            name,
            message: message[0],
            displayMessage: message[1],
            source: this.source,
            helpLink,
            error: innerError,
        });
    }
    static SystemError(name, message, innerError, stack, issueLink = this.issueLink) {
        return new teamsfx_api_1.SystemError({
            name,
            message: message[0],
            displayMessage: message[1],
            source: this.source,
            issueLink,
            error: innerError,
        });
    }
}
exports.ErrorFactory = ErrorFactory;
ErrorFactory.source = constants_1.FrontendPluginInfo.ShortName;
ErrorFactory.issueLink = constants_1.FrontendPluginInfo.IssueLink;
ErrorFactory.helpLink = constants_1.FrontendPluginInfo.HelpLink;
