// @microsoft/teamsfx-core/build/plugins/resource/appstudio/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.M365_MANIFEST_VERSION = exports.M365_SCHEMA = exports.STATIC_TABS_MAX_ITEMS = exports.TEAMS_APP_SHORT_NAME_MAX_LENGTH = exports.DEFAULT_DEVELOPER = exports.WEB_APPLICATION_INFO_MULTI_ENV = exports.BOTS_TPL_EXISTING_APP = exports.BOTS_TPL_FOR_COMMAND_AND_RESPONSE = exports.BOTS_TPL_FOR_NOTIFICATION = exports.BOTS_TPL_FOR_MULTI_ENV = exports.COMPOSE_EXTENSIONS_TPL_EXISTING_APP = exports.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV_M365 = exports.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV = exports.CONFIGURABLE_TABS_TPL_EXISTING_APP = exports.CONFIGURABLE_TABS_TPL_FOR_MULTI_ENV = exports.STATIC_TABS_TPL_EXISTING_APP = exports.STATIC_TABS_TPL_FOR_MULTI_ENV = exports.TEAMS_APP_MANIFEST_TEMPLATE_V3 = exports.SOLUTION_USERINFO = exports.SOLUTION = exports.MANIFEST_RESOURCES = exports.DEFAULT_OUTLINE_PNG_FILENAME = exports.DEFAULT_COLOR_PNG_FILENAME = exports.OUTLINE_TEMPLATE = exports.COLOR_TEMPLATE = exports.LOCAL_BOT_ID = exports.BOT_ID = exports.FRONTEND_INDEX_PATH = exports.FRONTEND_DOMAIN = exports.FRONTEND_ENDPOINT = exports.MANIFEST_TEMPLATE_CONSOLIDATE = exports.MANIFEST_LOCAL = exports.MANIFEST_TEMPLATE = exports.REMOTE_MANIFEST = exports.ErrorMessages = exports.APP_STUDIO_API_NAMES = exports.Constants = void 0;
class Constants {
}
exports.Constants = Constants;
Constants.MANIFEST_FILE = "manifest.json";
Constants.PLUGIN_NAME = "AppStudioPlugin";
Constants.BUILD_OR_PUBLISH_QUESTION = "build-or-publish";
Constants.INCLUDE_APP_MANIFEST = "include-app-manifest";
Constants.READ_MORE = "Read more";
Constants.VIEW_DEVELOPER_PORTAL = "View in Developer Portal";
Constants.DEVELOPER_PORTAL_APP_PACKAGE_URL = "https://dev.teams.microsoft.com/apps/%s/app-package";
Constants.PUBLISH_GUIDE = "https://aka.ms/teamsfx-publish";
Constants.TEAMS_ADMIN_PORTAL = "https://aka.ms/teamsfx-mtac";
Constants.TEAMS_MANAGE_APP_DOC = "https://aka.ms/teamsfx-mtac-doc";
Constants.TEAMS_APP_ID = "teamsAppId";
Constants.TEAMS_APP_UPDATED_AT = "teamsAppUpdatedAt";
Constants.PERMISSIONS = {
    name: "Teams App",
    noPermission: "No permission",
    admin: "Administrator",
    operative: "Operative",
    type: "M365",
};
Constants.CORRELATION_ID = "X-Correlation-ID";
class APP_STUDIO_API_NAMES {
}
exports.APP_STUDIO_API_NAMES = APP_STUDIO_API_NAMES;
APP_STUDIO_API_NAMES.CREATE_APP = "create-app";
APP_STUDIO_API_NAMES.GET_APP = "get-app";
APP_STUDIO_API_NAMES.PUBLISH_APP = "publish-app";
APP_STUDIO_API_NAMES.GET_PUBLISHED_APP = "get-published-app";
APP_STUDIO_API_NAMES.UPDATE_PUBLISHED_APP = "update-published-app";
APP_STUDIO_API_NAMES.UPDATE_OWNER = "update-owner";
class ErrorMessages {
}
exports.ErrorMessages = ErrorMessages;
ErrorMessages.GetConfigError = (configName, plugin) => `Failed to get configuration value "${configName}" for ${plugin}.`;
ErrorMessages.GrantPermissionFailed = "Response is empty or user is not added.";
ErrorMessages.TeamsAppNotFound = (teamsAppId) => `Cannot find Teams App with id: ${teamsAppId}. Maybe your current M365 account doesn't not have permission, or the Teams App has already been deleted.`;
/**
 * Config keys that are useful for generating remote teams app manifest
 */
exports.REMOTE_MANIFEST = "manifest.source.json";
exports.MANIFEST_TEMPLATE = "manifest.remote.template.json";
exports.MANIFEST_LOCAL = "manifest.local.template.json";
exports.MANIFEST_TEMPLATE_CONSOLIDATE = "manifest.template.json";
exports.FRONTEND_ENDPOINT = "endpoint";
exports.FRONTEND_DOMAIN = "domain";
exports.FRONTEND_INDEX_PATH = "indexPath";
exports.BOT_ID = "botId";
exports.LOCAL_BOT_ID = "localBotId";
exports.COLOR_TEMPLATE = "plugins/resource/appstudio/defaultIcon.png";
exports.OUTLINE_TEMPLATE = "plugins/resource/appstudio/defaultOutline.png";
exports.DEFAULT_COLOR_PNG_FILENAME = "color.png";
exports.DEFAULT_OUTLINE_PNG_FILENAME = "outline.png";
exports.MANIFEST_RESOURCES = "resources";
/**
 * Config Keys that are useful for remote collaboration
 */
exports.SOLUTION = "solution";
exports.SOLUTION_USERINFO = "userinfo";
exports.TEAMS_APP_MANIFEST_TEMPLATE_V3 = `{
  "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.14/MicrosoftTeams.schema.json",
  "manifestVersion": "1.14",
  "version": "1.0.0",
  "id": "{{state.fx-resource-appstudio.teamsAppId}}",
  "packageName": "com.microsoft.teams.extension",
  "developer": {
      "name": "Teams App, Inc.",
      "websiteUrl": "{{state.fx-resource-frontend-hosting.endpoint}}",
      "privacyUrl": "{{state.fx-resource-frontend-hosting.endpoint}}{{state.fx-resource-frontend-hosting.indexPath}}/privacy",
      "termsOfUseUrl": "{{state.fx-resource-frontend-hosting.endpoint}}{{state.fx-resource-frontend-hosting.indexPath}}/termsofuse"
  },
  "icons": {
      "color": "{{config.manifest.icons.color}}",
      "outline": "{{config.manifest.icons.outline}}"
  },
  "name": {
      "short": "{{config.manifest.appName.short}}",
      "full": "{{config.manifest.appName.full}}"
  },
  "description": {
      "short": "{{config.manifest.description.short}}",
      "full": "{{config.manifest.description.full}}"
  },
  "accentColor": "#FFFFFF",
  "bots": [],
  "composeExtensions": [],
  "configurableTabs": [],
  "staticTabs": [],
  "permissions": [
      "identity",
      "messageTeamMembers"
  ],
  "validDomains": []
}`;
exports.STATIC_TABS_TPL_FOR_MULTI_ENV = [
    {
        entityId: "index",
        name: "Personal Tab",
        contentUrl: "{{state.fx-resource-frontend-hosting.endpoint}}{{state.fx-resource-frontend-hosting.indexPath}}/tab",
        websiteUrl: "{{state.fx-resource-frontend-hosting.endpoint}}{{state.fx-resource-frontend-hosting.indexPath}}/tab",
        scopes: ["personal"],
    },
];
exports.STATIC_TABS_TPL_EXISTING_APP = [
    {
        entityId: "index",
        name: "Personal Tab",
        contentUrl: "{{config.manifest.tabContentUrl}}",
        websiteUrl: "{{config.manifest.tabWebsiteUrl}}",
        scopes: ["personal"],
    },
];
exports.CONFIGURABLE_TABS_TPL_FOR_MULTI_ENV = [
    {
        configurationUrl: "{{state.fx-resource-frontend-hosting.endpoint}}{{state.fx-resource-frontend-hosting.indexPath}}/config",
        canUpdateConfiguration: true,
        scopes: ["team", "groupchat"],
    },
];
exports.CONFIGURABLE_TABS_TPL_EXISTING_APP = [
    {
        configurationUrl: "{{config.manifest.tabConfigurationUrl}}",
        canUpdateConfiguration: true,
        scopes: ["team", "groupchat"],
    },
];
exports.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        commands: [
            {
                id: "createCard",
                context: ["compose"],
                description: "Command to run action to create a Card from Compose Box",
                title: "Create Card",
                type: "action",
                parameters: [
                    {
                        name: "title",
                        title: "Card title",
                        description: "Title for the card",
                        inputType: "text",
                    },
                    {
                        name: "subTitle",
                        title: "Subtitle",
                        description: "Subtitle for the card",
                        inputType: "text",
                    },
                    {
                        name: "text",
                        title: "Text",
                        description: "Text for the card",
                        inputType: "textarea",
                    },
                ],
            },
            {
                id: "shareMessage",
                context: ["message"],
                description: "Test command to run action on message context (message sharing)",
                title: "Share Message",
                type: "action",
                parameters: [
                    {
                        name: "includeImage",
                        title: "Include Image",
                        description: "Include image in Hero Card",
                        inputType: "toggle",
                    },
                ],
            },
            {
                id: "searchQuery",
                context: ["compose", "commandBox"],
                description: "Test command to run query",
                title: "Search",
                type: "query",
                parameters: [
                    {
                        name: "searchQuery",
                        title: "Search Query",
                        description: "Your search query",
                        inputType: "text",
                    },
                ],
            },
        ],
        messageHandlers: [
            {
                type: "link",
                value: {
                    domains: ["*.botframework.com"],
                },
            },
        ],
    },
];
exports.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV_M365 = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        commands: [
            {
                id: "searchQuery",
                context: ["compose", "commandBox"],
                description: "Test command to run query",
                title: "Search",
                type: "query",
                parameters: [
                    {
                        name: "searchQuery",
                        title: "Search Query",
                        description: "Your search query",
                        inputType: "text",
                    },
                ],
            },
        ],
    },
];
exports.COMPOSE_EXTENSIONS_TPL_EXISTING_APP = [
    {
        botId: "{{config.manifest.botId}}",
        commands: [],
        messageHandlers: [
            {
                type: "link",
                value: {
                    domains: ["*.botframework.com"],
                },
            },
        ],
    },
];
exports.BOTS_TPL_FOR_MULTI_ENV = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        scopes: ["personal", "team", "groupchat"],
        supportsFiles: false,
        isNotificationOnly: false,
        commandLists: [
            {
                scopes: ["personal", "team", "groupchat"],
                commands: [
                    {
                        title: "welcome",
                        description: "Resend welcome card of this Bot",
                    },
                    {
                        title: "learn",
                        description: "Learn about Adaptive Card and Bot Command",
                    },
                ],
            },
        ],
    },
];
exports.BOTS_TPL_FOR_NOTIFICATION = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        scopes: ["personal", "team", "groupchat"],
        supportsFiles: false,
        isNotificationOnly: false,
    },
];
exports.BOTS_TPL_FOR_COMMAND_AND_RESPONSE = [
    {
        botId: "{{state.fx-resource-bot.botId}}",
        scopes: ["personal", "team", "groupchat"],
        supportsFiles: false,
        isNotificationOnly: false,
        commandLists: [
            {
                scopes: ["personal", "team", "groupchat"],
                commands: [
                    {
                        title: "helloWorld",
                        description: "A helloworld command to send a welcome message",
                    },
                ],
            },
        ],
    },
];
exports.BOTS_TPL_EXISTING_APP = [
    {
        botId: "{{config.manifest.botId}}",
        scopes: ["personal", "team", "groupchat"],
        supportsFiles: false,
        isNotificationOnly: false,
        commandLists: [
            {
                scopes: ["personal", "team", "groupchat"],
                commands: [],
            },
        ],
    },
];
exports.WEB_APPLICATION_INFO_MULTI_ENV = {
    id: "{{state.fx-resource-aad-app-for-teams.clientId}}",
    resource: "{{state.fx-resource-aad-app-for-teams.applicationIdUris}}",
};
// Default values for the developer fields in manifest.
exports.DEFAULT_DEVELOPER = {
    name: "Teams App, Inc.",
    websiteUrl: "https://www.example.com",
    privacyUrl: "https://www.example.com/termofuse",
    termsOfUseUrl: "https://www.example.com/privacy",
};
exports.TEAMS_APP_SHORT_NAME_MAX_LENGTH = 30;
exports.STATIC_TABS_MAX_ITEMS = 16;
exports.M365_SCHEMA = "https://developer.microsoft.com/en-us/json-schemas/teams/v1.14/MicrosoftTeams.schema.json";
exports.M365_MANIFEST_VERSION = "1.14";
