// @microsoft/teamsfx-core/build/plugins/resource/appstudio/manifestTemplate.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteCapability = exports.updateCapability = exports.addCapabilities = exports.capabilityExceedLimit = exports.getCapabilities = exports.saveManifest = exports.loadManifest = exports.init = exports.getManifestTemplatePath = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs-extra"));
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const tools_1 = require("../../../common/tools");
const errors_1 = require("./errors");
const results_1 = require("./results");
const constants_1 = require("./constants");
const question_1 = require("../../solution/fx-solution/question");
const featureFlags_1 = require("../../../common/featureFlags");
async function getManifestTemplatePath(projectRoot) {
    const appDir = await tools_1.getAppDirectory(projectRoot);
    return `${appDir}/${constants_1.MANIFEST_TEMPLATE_CONSOLIDATE}`;
}
exports.getManifestTemplatePath = getManifestTemplatePath;
async function init(projectRoot, appName, existingApp) {
    const newAppPackageFolder = `${projectRoot}/templates/${teamsfx_api_1.AppPackageFolderName}`;
    await fs.ensureDir(newAppPackageFolder);
    const manifestString = constants_1.TEAMS_APP_MANIFEST_TEMPLATE_V3;
    const manifest = JSON.parse(manifestString);
    if (existingApp) {
        manifest.developer = constants_1.DEFAULT_DEVELOPER;
    }
    await saveManifest(projectRoot, manifest);
    return teamsfx_api_1.ok(undefined);
}
exports.init = init;
async function loadManifest(projectRoot) {
    const manifestFilePath = await getManifestTemplatePath(projectRoot);
    if (!(await fs.pathExists(manifestFilePath))) {
        return teamsfx_api_1.err(results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.FileNotFoundError.name, errors_1.AppStudioError.FileNotFoundError.message(manifestFilePath)));
    }
    try {
        const manifest = await fs.readJson(manifestFilePath);
        return teamsfx_api_1.ok(manifest);
    }
    catch (e) {
        if (e.stack && e.stack.startsWith("SyntaxError")) {
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.ManifestLoadFailedError.name, errors_1.AppStudioError.ManifestLoadFailedError.message(`Failed to load manifest file from ${manifestFilePath}, due to ${e.message}`)));
        }
        return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.ManifestLoadFailedError.name, errors_1.AppStudioError.ManifestLoadFailedError.message(`Failed to load manifest file from ${manifestFilePath}, due to ${e.message}`)));
    }
}
exports.loadManifest = loadManifest;
async function saveManifest(projectRoot, manifest) {
    const manifestFilePath = await getManifestTemplatePath(projectRoot);
    await fs.writeFile(manifestFilePath, JSON.stringify(manifest, null, 4));
    return teamsfx_api_1.ok(manifestFilePath);
}
exports.saveManifest = saveManifest;
/**
 * Only works for manifest.template.json
 * @param projectRoot
 * @returns
 */
async function getCapabilities(projectRoot) {
    const manifestRes = await loadManifest(projectRoot);
    if (manifestRes.isErr()) {
        return teamsfx_api_1.err(manifestRes.error);
    }
    const capabilities = [];
    if (manifestRes.value.staticTabs && manifestRes.value.staticTabs.length > 0) {
        capabilities.push("staticTab");
    }
    if (manifestRes.value.configurableTabs && manifestRes.value.configurableTabs.length > 0) {
        capabilities.push("configurableTab");
    }
    if (manifestRes.value.bots && manifestRes.value.bots.length > 0) {
        capabilities.push("Bot");
    }
    if (manifestRes.value.composeExtensions) {
        capabilities.push("MessageExtension");
    }
    return teamsfx_api_1.ok(capabilities);
}
exports.getCapabilities = getCapabilities;
async function capabilityExceedLimit(projectRoot, capability) {
    const manifest = await loadManifest(projectRoot);
    if (manifest.isErr()) {
        return teamsfx_api_1.err(manifest.error);
    }
    let exceed = false;
    switch (capability) {
        case "staticTab":
            exceed =
                manifest.value.staticTabs !== undefined &&
                    manifest.value.staticTabs.length >= constants_1.STATIC_TABS_MAX_ITEMS;
            return teamsfx_api_1.ok(exceed);
        case "configurableTab":
            exceed =
                manifest.value.configurableTabs !== undefined &&
                    manifest.value.configurableTabs.length >= 1;
            return teamsfx_api_1.ok(exceed);
        case "Bot":
            exceed = manifest.value.bots !== undefined && manifest.value.bots.length >= 1;
            return teamsfx_api_1.ok(exceed);
        case "MessageExtension":
            exceed =
                manifest.value.composeExtensions !== undefined &&
                    manifest.value.composeExtensions.length >= 1;
            return teamsfx_api_1.ok(exceed);
        case "WebApplicationInfo":
            return teamsfx_api_1.ok(false);
        default:
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.InvalidCapabilityError.name, errors_1.AppStudioError.InvalidCapabilityError.message(capability)));
    }
}
exports.capabilityExceedLimit = capabilityExceedLimit;
async function addCapabilities(projectRoot, capabilities, inputs) {
    var _a, _b;
    const manifestRes = await loadManifest(projectRoot);
    if (manifestRes.isErr()) {
        return teamsfx_api_1.err(manifestRes.error);
    }
    const manifest = manifestRes.value;
    let staticTabIndex = (_b = (_a = manifest.staticTabs) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
    capabilities.map((capability) => {
        var _a, _b, _c, _d, _e;
        switch (capability.name) {
            case "staticTab":
                if (!manifest.staticTabs) {
                    Object.assign(manifest, { staticTabs: [] });
                }
                if (capability.snippet) {
                    manifest.staticTabs.push(capability.snippet);
                }
                else {
                    if (capability.existingApp) {
                        const template = tools_1.deepCopy(constants_1.STATIC_TABS_TPL_EXISTING_APP[0]);
                        template.entityId = "index" + staticTabIndex;
                        manifest.staticTabs.push(template);
                    }
                    else {
                        const template = tools_1.deepCopy(constants_1.STATIC_TABS_TPL_FOR_MULTI_ENV[0]);
                        template.entityId = "index" + staticTabIndex;
                        manifest.staticTabs.push(template);
                        if (((_a = manifest.validDomains) === null || _a === void 0 ? void 0 : _a.indexOf("{{state.fx-resource-frontend-hosting.domain}}")) == -1) {
                            (_b = manifest.validDomains) === null || _b === void 0 ? void 0 : _b.push("{{state.fx-resource-frontend-hosting.domain}}");
                        }
                    }
                    staticTabIndex++;
                }
                break;
            case "configurableTab":
                if (!manifest.configurableTabs) {
                    Object.assign(manifest, { configurableTabs: [] });
                }
                if (capability.snippet) {
                    manifest.configurableTabs.push(capability.snippet);
                }
                else {
                    if (capability.existingApp) {
                        manifest.configurableTabs = manifest.configurableTabs.concat(constants_1.CONFIGURABLE_TABS_TPL_EXISTING_APP);
                    }
                    else {
                        manifest.configurableTabs = manifest.configurableTabs.concat(constants_1.CONFIGURABLE_TABS_TPL_FOR_MULTI_ENV);
                        if (((_c = manifest.validDomains) === null || _c === void 0 ? void 0 : _c.indexOf("{{state.fx-resource-frontend-hosting.domain}}")) == -1) {
                            (_d = manifest.validDomains) === null || _d === void 0 ? void 0 : _d.push("{{state.fx-resource-frontend-hosting.domain}}");
                        }
                    }
                }
                break;
            case "Bot":
                if (!manifest.bots) {
                    Object.assign(manifest, { bots: [] });
                }
                if (capability.snippet) {
                    manifest.bots.push(capability.snippet);
                }
                else {
                    if (capability.existingApp) {
                        manifest.bots = manifest.bots.concat(constants_1.BOTS_TPL_EXISTING_APP);
                    }
                    else {
                        if (manifest.bots === undefined) {
                            manifest.bots = [];
                        }
                        if (featureFlags_1.isBotNotificationEnabled()) {
                            const scenariosRaw = inputs[question_1.AzureSolutionQuestionNames.Scenarios];
                            const scenarios = Array.isArray(scenariosRaw) ? scenariosRaw : [];
                            if (scenarios.includes(question_1.BotScenario.CommandAndResponseBot) ||
                                scenarios.includes(question_1.BotScenario.WorkflowBot)) {
                                // command and response bot or workflow bot
                                manifest.bots = manifest.bots.concat(constants_1.BOTS_TPL_FOR_COMMAND_AND_RESPONSE);
                            }
                            else if (scenarios.includes(question_1.BotScenario.NotificationBot)) {
                                // notification
                                manifest.bots = manifest.bots.concat(constants_1.BOTS_TPL_FOR_NOTIFICATION);
                            }
                            else {
                                // legacy bot
                                manifest.bots = manifest.bots.concat(constants_1.BOTS_TPL_FOR_MULTI_ENV);
                            }
                        }
                        else {
                            manifest.bots = manifest.bots.concat(constants_1.BOTS_TPL_FOR_MULTI_ENV);
                        }
                        (_e = manifest.validDomains) === null || _e === void 0 ? void 0 : _e.push("{{state.fx-resource-bot.validDomain}}");
                    }
                }
                break;
            case "MessageExtension":
                if (!manifest.composeExtensions) {
                    Object.assign(manifest, { composeExtensions: [] });
                }
                if (capability.snippet) {
                    manifest.composeExtensions.push(capability.snippet);
                }
                else {
                    if (capability.existingApp) {
                        manifest.composeExtensions = manifest.composeExtensions.concat(constants_1.COMPOSE_EXTENSIONS_TPL_EXISTING_APP);
                    }
                    else {
                        manifest.composeExtensions = manifest.composeExtensions.concat(constants_1.COMPOSE_EXTENSIONS_TPL_FOR_MULTI_ENV);
                    }
                }
                break;
            case "WebApplicationInfo":
                if (capability.snippet) {
                    manifest.webApplicationInfo = capability.snippet;
                }
                else {
                    manifest.webApplicationInfo = constants_1.WEB_APPLICATION_INFO_MULTI_ENV;
                }
                break;
        }
    });
    const res = await saveManifest(projectRoot, manifest);
    if (res.isErr()) {
        return teamsfx_api_1.err(res.error);
    }
    return teamsfx_api_1.ok(undefined);
}
exports.addCapabilities = addCapabilities;
async function updateCapability(projectRoot, capability) {
    var _a;
    const manifestRes = await loadManifest(projectRoot);
    if (manifestRes.isErr()) {
        return teamsfx_api_1.err(manifestRes.error);
    }
    const manifest = manifestRes.value;
    switch (capability.name) {
        case "staticTab":
            // find the corresponding static Tab with entity id
            const entityId = capability.snippet.entityId;
            const index = (_a = manifest.staticTabs) === null || _a === void 0 ? void 0 : _a.map((x) => x.entityId).indexOf(entityId);
            if (index !== undefined && index !== -1) {
                manifest.staticTabs[index] = capability.snippet;
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.StaticTabNotExistError.name, errors_1.AppStudioError.StaticTabNotExistError.message(entityId)));
            }
            break;
        case "configurableTab":
            if (manifest.configurableTabs && manifest.configurableTabs.length) {
                manifest.configurableTabs[0] = capability.snippet;
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CapabilityNotExistError.name, errors_1.AppStudioError.CapabilityNotExistError.message(capability.name)));
            }
            break;
        case "Bot":
            if (manifest.bots && manifest.bots.length > 0) {
                manifest.bots[0] = capability.snippet;
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CapabilityNotExistError.name, errors_1.AppStudioError.CapabilityNotExistError.message(capability.name)));
            }
            break;
        case "MessageExtension":
            if (manifest.composeExtensions && manifest.composeExtensions.length > 0) {
                manifest.composeExtensions[0] = capability.snippet;
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CapabilityNotExistError.name, errors_1.AppStudioError.CapabilityNotExistError.message(capability.name)));
            }
            break;
        case "WebApplicationInfo":
            manifest.webApplicationInfo = capability.snippet;
            break;
    }
    const res = await saveManifest(projectRoot, manifest);
    if (res.isErr()) {
        return teamsfx_api_1.err(res.error);
    }
    return teamsfx_api_1.ok(undefined);
}
exports.updateCapability = updateCapability;
async function deleteCapability(projectRoot, capability) {
    var _a;
    const manifestRes = await loadManifest(projectRoot);
    if (manifestRes.isErr()) {
        return teamsfx_api_1.err(manifestRes.error);
    }
    const manifest = manifestRes.value;
    switch (capability.name) {
        case "staticTab":
            // find the corresponding static Tab with entity id
            const entityId = capability.snippet.entityId;
            const index = (_a = manifest.staticTabs) === null || _a === void 0 ? void 0 : _a.map((x) => x.entityId).indexOf(entityId);
            if (index !== undefined && index !== -1) {
                manifest.staticTabs.slice(index, 1);
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.StaticTabNotExistError.name, errors_1.AppStudioError.StaticTabNotExistError.message(entityId)));
            }
            break;
        case "configurableTab":
            if (manifest.configurableTabs && manifest.configurableTabs.length > 0) {
                manifest.configurableTabs = [];
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CapabilityNotExistError.name, errors_1.AppStudioError.CapabilityNotExistError.message(capability.name)));
            }
            break;
        case "Bot":
            if (manifest.bots && manifest.bots.length > 0) {
                manifest.bots = [];
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CapabilityNotExistError.name, errors_1.AppStudioError.CapabilityNotExistError.message(capability.name)));
            }
            break;
        case "MessageExtension":
            if (manifest.composeExtensions && manifest.composeExtensions.length > 0) {
                manifest.composeExtensions = [];
            }
            else {
                return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.CapabilityNotExistError.name, errors_1.AppStudioError.CapabilityNotExistError.message(capability.name)));
            }
            break;
        case "WebApplicationInfo":
            manifest.webApplicationInfo = undefined;
            break;
    }
    const res = await saveManifest(projectRoot, manifest);
    if (res.isErr()) {
        return teamsfx_api_1.err(res.error);
    }
    return teamsfx_api_1.ok(undefined);
}
exports.deleteCapability = deleteCapability;
