// @microsoft/teamsfx-core/build/plugins/resource/appstudio/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetryHandler = exports.renderTemplate = exports.getLocalAppName = exports.getCustomizedKeys = exports.checkAndConfig = exports.replaceConfigValue = void 0;
const tslib_1 = require("tslib");
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const mustache_1 = tslib_1.__importDefault(require("mustache"));
const constants_1 = require("../constants");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const results_1 = require("../results");
function replaceConfigValue(config, id, value) {
    if (config && id && value) {
        const idTag = `{${id}}`;
        while (config.includes(idTag)) {
            config = config.replace(idTag, value);
        }
    }
    return config;
}
exports.replaceConfigValue = replaceConfigValue;
/**
 *
 * @throws Error - when placeholder doesn't have corresponding value
 */
function checkAndConfig(config, id, value) {
    const idTag = `{${id}}`;
    if (value) {
        return replaceConfigValue(config, id, value);
    }
    else {
        if (config.includes(idTag)) {
            throw results_1.AppStudioResultFactory.SystemError("RequiredDataMissing", [
                localizeUtils_1.getDefaultString("plugins.appstudio.dataRequired", idTag),
                localizeUtils_1.getLocalizedString("plugins.appstudio.dataRequired", idTag),
            ]);
        }
        else {
            return config;
        }
    }
}
exports.checkAndConfig = checkAndConfig;
function getCustomizedKeys(prefix, manifest) {
    let keys = [];
    for (const key in manifest) {
        if (manifest.hasOwnProperty(key)) {
            const value = manifest[key];
            if (typeof value === "object") {
                if (Array.isArray(value)) {
                    value.map((item, index) => {
                        keys = keys.concat(getCustomizedKeys(`${prefix}.${key}[${index}]`, item));
                    });
                }
                else {
                    keys = keys.concat(getCustomizedKeys(`${prefix}.${key}`, value));
                }
            }
            else if (typeof value === "string" && value.startsWith("{{config.manifest")) {
                keys.push(`${prefix}.${key}`);
            }
        }
    }
    return keys;
}
exports.getCustomizedKeys = getCustomizedKeys;
function getLocalAppName(appName) {
    const suffix = "-local-debug";
    if (suffix.length + appName.length <= constants_1.TEAMS_APP_SHORT_NAME_MAX_LENGTH) {
        appName = appName + suffix;
    }
    return appName;
}
exports.getLocalAppName = getLocalAppName;
function renderTemplate(manifestString, view) {
    // Unesacped HTML
    mustache_1.default.escape = (value) => value;
    manifestString = mustache_1.default.render(manifestString, view);
    return manifestString;
}
exports.renderTemplate = renderTemplate;
class RetryHandler {
    static async Retry(fn) {
        var _a;
        let retries = this.RETRIES;
        let response;
        while (retries > 0) {
            retries = retries - 1;
            try {
                response = await fn();
                return response;
            }
            catch (e) {
                // Directly throw 404 error, keep trying for other status code e.g. 503 400
                if (retries <= 0 || ((_a = e.response) === null || _a === void 0 ? void 0 : _a.status) == 404) {
                    throw e;
                }
                else {
                    await new Promise((resolve) => setTimeout(resolve, 5000));
                }
            }
        }
    }
}
exports.RetryHandler = RetryHandler;
RetryHandler.RETRIES = 6;
