// @azure/arm-sql/esm/operations/capabilities.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/capabilitiesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a Capabilities. */
var Capabilities = /** @class */ (function () {
    /**
     * Create a Capabilities.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function Capabilities(client) {
        this.client = client;
    }
    Capabilities.prototype.listByLocation = function (locationName, options, callback) {
        return this.client.sendOperationRequest({
            locationName: locationName,
            options: options
        }, listByLocationOperationSpec, callback);
    };
    return Capabilities;
}());
export { Capabilities };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByLocationOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/providers/Microsoft.Sql/locations/{locationName}/capabilities",
    urlParameters: [
        Parameters.locationName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.include,
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.LocationCapabilities
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
