// @azure/arm-sql/esm/operations/databaseUsages.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/databaseUsagesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a DatabaseUsages. */
var DatabaseUsages = /** @class */ (function () {
    /**
     * Create a DatabaseUsages.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function DatabaseUsages(client) {
        this.client = client;
    }
    DatabaseUsages.prototype.listByDatabase = function (resourceGroupName, serverName, databaseName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            databaseName: databaseName,
            options: options
        }, listByDatabaseOperationSpec, callback);
    };
    return DatabaseUsages;
}());
export { DatabaseUsages };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByDatabaseOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/usages",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.databaseName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.DatabaseUsageListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
