// @azure/arm-sql/esm/operations/elasticPoolDatabaseActivities.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/elasticPoolDatabaseActivitiesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ElasticPoolDatabaseActivities. */
var ElasticPoolDatabaseActivities = /** @class */ (function () {
    /**
     * Create a ElasticPoolDatabaseActivities.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ElasticPoolDatabaseActivities(client) {
        this.client = client;
    }
    ElasticPoolDatabaseActivities.prototype.listByElasticPool = function (resourceGroupName, serverName, elasticPoolName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            elasticPoolName: elasticPoolName,
            options: options
        }, listByElasticPoolOperationSpec, callback);
    };
    return ElasticPoolDatabaseActivities;
}());
export { ElasticPoolDatabaseActivities };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByElasticPoolOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/elasticPools/{elasticPoolName}/elasticPoolDatabaseActivity",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.elasticPoolName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ElasticPoolDatabaseActivityListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
