// @azure/arm-sql/esm/operations/extendedServerBlobAuditingPolicies.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/extendedServerBlobAuditingPoliciesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ExtendedServerBlobAuditingPolicies. */
var ExtendedServerBlobAuditingPolicies = /** @class */ (function () {
    /**
     * Create a ExtendedServerBlobAuditingPolicies.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ExtendedServerBlobAuditingPolicies(client) {
        this.client = client;
    }
    ExtendedServerBlobAuditingPolicies.prototype.get = function (resourceGroupName, serverName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            options: options
        }, getOperationSpec, callback);
    };
    /**
     * Creates or updates an extended server's blob auditing policy.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serverName The name of the server.
     * @param parameters Properties of extended blob auditing policy
     * @param [options] The optional parameters
     * @returns Promise<Models.ExtendedServerBlobAuditingPoliciesCreateOrUpdateResponse>
     */
    ExtendedServerBlobAuditingPolicies.prototype.createOrUpdate = function (resourceGroupName, serverName, parameters, options) {
        return this.beginCreateOrUpdate(resourceGroupName, serverName, parameters, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Creates or updates an extended server's blob auditing policy.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serverName The name of the server.
     * @param parameters Properties of extended blob auditing policy
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ExtendedServerBlobAuditingPolicies.prototype.beginCreateOrUpdate = function (resourceGroupName, serverName, parameters, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            parameters: parameters,
            options: options
        }, beginCreateOrUpdateOperationSpec, options);
    };
    return ExtendedServerBlobAuditingPolicies;
}());
export { ExtendedServerBlobAuditingPolicies };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/extendedAuditingSettings/{blobAuditingPolicyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.blobAuditingPolicyName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ExtendedServerBlobAuditingPolicy
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginCreateOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/extendedAuditingSettings/{blobAuditingPolicyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.blobAuditingPolicyName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ExtendedServerBlobAuditingPolicy), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ExtendedServerBlobAuditingPolicy
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
