// @azure/arm-sql/esm/operations/instanceFailoverGroups.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/instanceFailoverGroupsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a InstanceFailoverGroups. */
var InstanceFailoverGroups = /** @class */ (function () {
    /**
     * Create a InstanceFailoverGroups.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function InstanceFailoverGroups(client) {
        this.client = client;
    }
    InstanceFailoverGroups.prototype.get = function (resourceGroupName, locationName, failoverGroupName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            locationName: locationName,
            failoverGroupName: failoverGroupName,
            options: options
        }, getOperationSpec, callback);
    };
    /**
     * Creates or updates a failover group.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param parameters The failover group parameters.
     * @param [options] The optional parameters
     * @returns Promise<Models.InstanceFailoverGroupsCreateOrUpdateResponse>
     */
    InstanceFailoverGroups.prototype.createOrUpdate = function (resourceGroupName, locationName, failoverGroupName, parameters, options) {
        return this.beginCreateOrUpdate(resourceGroupName, locationName, failoverGroupName, parameters, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Deletes a failover group.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param [options] The optional parameters
     * @returns Promise<msRest.RestResponse>
     */
    InstanceFailoverGroups.prototype.deleteMethod = function (resourceGroupName, locationName, failoverGroupName, options) {
        return this.beginDeleteMethod(resourceGroupName, locationName, failoverGroupName, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    InstanceFailoverGroups.prototype.listByLocation = function (resourceGroupName, locationName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            locationName: locationName,
            options: options
        }, listByLocationOperationSpec, callback);
    };
    /**
     * Fails over from the current primary managed instance to this managed instance.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param [options] The optional parameters
     * @returns Promise<Models.InstanceFailoverGroupsFailoverResponse>
     */
    InstanceFailoverGroups.prototype.failover = function (resourceGroupName, locationName, failoverGroupName, options) {
        return this.beginFailover(resourceGroupName, locationName, failoverGroupName, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Fails over from the current primary managed instance to this managed instance. This operation
     * might result in data loss.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param [options] The optional parameters
     * @returns Promise<Models.InstanceFailoverGroupsForceFailoverAllowDataLossResponse>
     */
    InstanceFailoverGroups.prototype.forceFailoverAllowDataLoss = function (resourceGroupName, locationName, failoverGroupName, options) {
        return this.beginForceFailoverAllowDataLoss(resourceGroupName, locationName, failoverGroupName, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Creates or updates a failover group.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param parameters The failover group parameters.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    InstanceFailoverGroups.prototype.beginCreateOrUpdate = function (resourceGroupName, locationName, failoverGroupName, parameters, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            locationName: locationName,
            failoverGroupName: failoverGroupName,
            parameters: parameters,
            options: options
        }, beginCreateOrUpdateOperationSpec, options);
    };
    /**
     * Deletes a failover group.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    InstanceFailoverGroups.prototype.beginDeleteMethod = function (resourceGroupName, locationName, failoverGroupName, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            locationName: locationName,
            failoverGroupName: failoverGroupName,
            options: options
        }, beginDeleteMethodOperationSpec, options);
    };
    /**
     * Fails over from the current primary managed instance to this managed instance.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    InstanceFailoverGroups.prototype.beginFailover = function (resourceGroupName, locationName, failoverGroupName, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            locationName: locationName,
            failoverGroupName: failoverGroupName,
            options: options
        }, beginFailoverOperationSpec, options);
    };
    /**
     * Fails over from the current primary managed instance to this managed instance. This operation
     * might result in data loss.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param locationName The name of the region where the resource is located.
     * @param failoverGroupName The name of the failover group.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    InstanceFailoverGroups.prototype.beginForceFailoverAllowDataLoss = function (resourceGroupName, locationName, failoverGroupName, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            locationName: locationName,
            failoverGroupName: failoverGroupName,
            options: options
        }, beginForceFailoverAllowDataLossOperationSpec, options);
    };
    InstanceFailoverGroups.prototype.listByLocationNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByLocationNextOperationSpec, callback);
    };
    return InstanceFailoverGroups;
}());
export { InstanceFailoverGroups };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/locations/{locationName}/instanceFailoverGroups/{failoverGroupName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.locationName,
        Parameters.failoverGroupName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.InstanceFailoverGroup
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByLocationOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/locations/{locationName}/instanceFailoverGroups",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.locationName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.InstanceFailoverGroupListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginCreateOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/locations/{locationName}/instanceFailoverGroups/{failoverGroupName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.locationName,
        Parameters.failoverGroupName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.InstanceFailoverGroup), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.InstanceFailoverGroup
        },
        201: {
            bodyMapper: Mappers.InstanceFailoverGroup
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginDeleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/locations/{locationName}/instanceFailoverGroups/{failoverGroupName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.locationName,
        Parameters.failoverGroupName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        202: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginFailoverOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/locations/{locationName}/instanceFailoverGroups/{failoverGroupName}/failover",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.locationName,
        Parameters.failoverGroupName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.InstanceFailoverGroup
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginForceFailoverAllowDataLossOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/locations/{locationName}/instanceFailoverGroups/{failoverGroupName}/forceFailoverAllowDataLoss",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.locationName,
        Parameters.failoverGroupName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.InstanceFailoverGroup
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByLocationNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.InstanceFailoverGroupListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
