// @azure/arm-sql/esm/operations/jobTargetExecutions.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/jobTargetExecutionsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a JobTargetExecutions. */
var JobTargetExecutions = /** @class */ (function () {
    /**
     * Create a JobTargetExecutions.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function JobTargetExecutions(client) {
        this.client = client;
    }
    JobTargetExecutions.prototype.listByJobExecution = function (resourceGroupName, serverName, jobAgentName, jobName, jobExecutionId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            jobAgentName: jobAgentName,
            jobName: jobName,
            jobExecutionId: jobExecutionId,
            options: options
        }, listByJobExecutionOperationSpec, callback);
    };
    JobTargetExecutions.prototype.listByStep = function (resourceGroupName, serverName, jobAgentName, jobName, jobExecutionId, stepName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            jobAgentName: jobAgentName,
            jobName: jobName,
            jobExecutionId: jobExecutionId,
            stepName: stepName,
            options: options
        }, listByStepOperationSpec, callback);
    };
    JobTargetExecutions.prototype.get = function (resourceGroupName, serverName, jobAgentName, jobName, jobExecutionId, stepName, targetId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            jobAgentName: jobAgentName,
            jobName: jobName,
            jobExecutionId: jobExecutionId,
            stepName: stepName,
            targetId: targetId,
            options: options
        }, getOperationSpec, callback);
    };
    JobTargetExecutions.prototype.listByJobExecutionNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByJobExecutionNextOperationSpec, callback);
    };
    JobTargetExecutions.prototype.listByStepNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByStepNextOperationSpec, callback);
    };
    return JobTargetExecutions;
}());
export { JobTargetExecutions };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByJobExecutionOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/jobAgents/{jobAgentName}/jobs/{jobName}/executions/{jobExecutionId}/targets",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.jobAgentName,
        Parameters.jobName,
        Parameters.jobExecutionId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.createTimeMin,
        Parameters.createTimeMax,
        Parameters.endTimeMin,
        Parameters.endTimeMax,
        Parameters.isActive,
        Parameters.skip,
        Parameters.top,
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.JobExecutionListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByStepOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/jobAgents/{jobAgentName}/jobs/{jobName}/executions/{jobExecutionId}/steps/{stepName}/targets",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.jobAgentName,
        Parameters.jobName,
        Parameters.jobExecutionId,
        Parameters.stepName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.createTimeMin,
        Parameters.createTimeMax,
        Parameters.endTimeMin,
        Parameters.endTimeMax,
        Parameters.isActive,
        Parameters.skip,
        Parameters.top,
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.JobExecutionListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/jobAgents/{jobAgentName}/jobs/{jobName}/executions/{jobExecutionId}/steps/{stepName}/targets/{targetId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.jobAgentName,
        Parameters.jobName,
        Parameters.jobExecutionId,
        Parameters.stepName,
        Parameters.targetId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.JobExecution
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByJobExecutionNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.JobExecutionListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByStepNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.JobExecutionListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
