// @azure/arm-sql/esm/operations/managedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/managedRestorableDroppedDatabaseBackupShortTermRetentionPoliciesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies. */
var ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies = /** @class */ (function () {
    /**
     * Create a ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies(client) {
        this.client = client;
    }
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.get = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            restorableDroppedDatabaseId: restorableDroppedDatabaseId,
            options: options
        }, getOperationSpec, callback);
    };
    /**
     * Sets a database's long term retention policy.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param restorableDroppedDatabaseId
     * @param parameters The long term retention policy info.
     * @param [options] The optional parameters
     * @returns
     * Promise<Models.ManagedRestorableDroppedDatabaseBackupShortTermRetentionPoliciesCreateOrUpdateResponse>
     */
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.createOrUpdate = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, parameters, options) {
        return this.beginCreateOrUpdate(resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, parameters, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Sets a database's long term retention policy.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param restorableDroppedDatabaseId
     * @param parameters The long term retention policy info.
     * @param [options] The optional parameters
     * @returns
     * Promise<Models.ManagedRestorableDroppedDatabaseBackupShortTermRetentionPoliciesUpdateResponse>
     */
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.update = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, parameters, options) {
        return this.beginUpdate(resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, parameters, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.listByRestorableDroppedDatabase = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            restorableDroppedDatabaseId: restorableDroppedDatabaseId,
            options: options
        }, listByRestorableDroppedDatabaseOperationSpec, callback);
    };
    /**
     * Sets a database's long term retention policy.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param restorableDroppedDatabaseId
     * @param parameters The long term retention policy info.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.beginCreateOrUpdate = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, parameters, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            restorableDroppedDatabaseId: restorableDroppedDatabaseId,
            parameters: parameters,
            options: options
        }, beginCreateOrUpdateOperationSpec, options);
    };
    /**
     * Sets a database's long term retention policy.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param restorableDroppedDatabaseId
     * @param parameters The long term retention policy info.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.beginUpdate = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, parameters, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            restorableDroppedDatabaseId: restorableDroppedDatabaseId,
            parameters: parameters,
            options: options
        }, beginUpdateOperationSpec, options);
    };
    ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies.prototype.listByRestorableDroppedDatabaseNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByRestorableDroppedDatabaseNextOperationSpec, callback);
    };
    return ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies;
}());
export { ManagedRestorableDroppedDatabaseBackupShortTermRetentionPolicies };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/restorableDroppedDatabases/{restorableDroppedDatabaseId}/backupShortTermRetentionPolicies/{policyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.restorableDroppedDatabaseId,
        Parameters.policyName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedBackupShortTermRetentionPolicy
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByRestorableDroppedDatabaseOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/restorableDroppedDatabases/{restorableDroppedDatabaseId}/backupShortTermRetentionPolicies",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.restorableDroppedDatabaseId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedBackupShortTermRetentionPolicyListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginCreateOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/restorableDroppedDatabases/{restorableDroppedDatabaseId}/backupShortTermRetentionPolicies/{policyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.restorableDroppedDatabaseId,
        Parameters.policyName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ManagedBackupShortTermRetentionPolicy), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ManagedBackupShortTermRetentionPolicy
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginUpdateOperationSpec = {
    httpMethod: "PATCH",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/restorableDroppedDatabases/{restorableDroppedDatabaseId}/backupShortTermRetentionPolicies/{policyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.restorableDroppedDatabaseId,
        Parameters.policyName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ManagedBackupShortTermRetentionPolicy), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ManagedBackupShortTermRetentionPolicy
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByRestorableDroppedDatabaseNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedBackupShortTermRetentionPolicyListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
