// @azure/arm-sql/esm/operations/managedDatabaseSecurityAlertPolicies.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/managedDatabaseSecurityAlertPoliciesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ManagedDatabaseSecurityAlertPolicies. */
var ManagedDatabaseSecurityAlertPolicies = /** @class */ (function () {
    /**
     * Create a ManagedDatabaseSecurityAlertPolicies.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ManagedDatabaseSecurityAlertPolicies(client) {
        this.client = client;
    }
    ManagedDatabaseSecurityAlertPolicies.prototype.get = function (resourceGroupName, managedInstanceName, databaseName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            databaseName: databaseName,
            options: options
        }, getOperationSpec, callback);
    };
    ManagedDatabaseSecurityAlertPolicies.prototype.createOrUpdate = function (resourceGroupName, managedInstanceName, databaseName, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            databaseName: databaseName,
            parameters: parameters,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    ManagedDatabaseSecurityAlertPolicies.prototype.listByDatabase = function (resourceGroupName, managedInstanceName, databaseName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            databaseName: databaseName,
            options: options
        }, listByDatabaseOperationSpec, callback);
    };
    ManagedDatabaseSecurityAlertPolicies.prototype.listByDatabaseNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByDatabaseNextOperationSpec, callback);
    };
    return ManagedDatabaseSecurityAlertPolicies;
}());
export { ManagedDatabaseSecurityAlertPolicies };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/databases/{databaseName}/securityAlertPolicies/{securityAlertPolicyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.databaseName,
        Parameters.securityAlertPolicyName0,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedDatabaseSecurityAlertPolicy
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/databases/{databaseName}/securityAlertPolicies/{securityAlertPolicyName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.databaseName,
        Parameters.securityAlertPolicyName0,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ManagedDatabaseSecurityAlertPolicy), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ManagedDatabaseSecurityAlertPolicy
        },
        201: {
            bodyMapper: Mappers.ManagedDatabaseSecurityAlertPolicy
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByDatabaseOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/databases/{databaseName}/securityAlertPolicies",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.databaseName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedDatabaseSecurityAlertPolicyListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByDatabaseNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedDatabaseSecurityAlertPolicyListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
