// @azure/arm-sql/esm/operations/managedInstanceAdministrators.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/managedInstanceAdministratorsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ManagedInstanceAdministrators. */
var ManagedInstanceAdministrators = /** @class */ (function () {
    /**
     * Create a ManagedInstanceAdministrators.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ManagedInstanceAdministrators(client) {
        this.client = client;
    }
    ManagedInstanceAdministrators.prototype.listByInstance = function (resourceGroupName, managedInstanceName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            options: options
        }, listByInstanceOperationSpec, callback);
    };
    ManagedInstanceAdministrators.prototype.get = function (resourceGroupName, managedInstanceName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            options: options
        }, getOperationSpec, callback);
    };
    /**
     * Creates or updates a managed instance administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param parameters The requested administrator parameters.
     * @param [options] The optional parameters
     * @returns Promise<Models.ManagedInstanceAdministratorsCreateOrUpdateResponse>
     */
    ManagedInstanceAdministrators.prototype.createOrUpdate = function (resourceGroupName, managedInstanceName, parameters, options) {
        return this.beginCreateOrUpdate(resourceGroupName, managedInstanceName, parameters, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Deletes a managed instance administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param [options] The optional parameters
     * @returns Promise<msRest.RestResponse>
     */
    ManagedInstanceAdministrators.prototype.deleteMethod = function (resourceGroupName, managedInstanceName, options) {
        return this.beginDeleteMethod(resourceGroupName, managedInstanceName, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Creates or updates a managed instance administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param parameters The requested administrator parameters.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ManagedInstanceAdministrators.prototype.beginCreateOrUpdate = function (resourceGroupName, managedInstanceName, parameters, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            parameters: parameters,
            options: options
        }, beginCreateOrUpdateOperationSpec, options);
    };
    /**
     * Deletes a managed instance administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param managedInstanceName The name of the managed instance.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ManagedInstanceAdministrators.prototype.beginDeleteMethod = function (resourceGroupName, managedInstanceName, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            options: options
        }, beginDeleteMethodOperationSpec, options);
    };
    ManagedInstanceAdministrators.prototype.listByInstanceNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByInstanceNextOperationSpec, callback);
    };
    return ManagedInstanceAdministrators;
}());
export { ManagedInstanceAdministrators };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByInstanceOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/administrators",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceAdministratorListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/administrators/{administratorName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.administratorName1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceAdministrator
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginCreateOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/administrators/{administratorName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.administratorName1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ManagedInstanceAdministrator), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceAdministrator
        },
        201: {
            bodyMapper: Mappers.ManagedInstanceAdministrator
        },
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginDeleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/administrators/{administratorName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.administratorName1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        202: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByInstanceNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ManagedInstanceAdministratorListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
