// @azure/arm-sql/esm/operations/recoverableManagedDatabases.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/recoverableManagedDatabasesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a RecoverableManagedDatabases. */
var RecoverableManagedDatabases = /** @class */ (function () {
    /**
     * Create a RecoverableManagedDatabases.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function RecoverableManagedDatabases(client) {
        this.client = client;
    }
    RecoverableManagedDatabases.prototype.listByInstance = function (resourceGroupName, managedInstanceName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            options: options
        }, listByInstanceOperationSpec, callback);
    };
    RecoverableManagedDatabases.prototype.get = function (resourceGroupName, managedInstanceName, recoverableDatabaseName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            recoverableDatabaseName: recoverableDatabaseName,
            options: options
        }, getOperationSpec, callback);
    };
    RecoverableManagedDatabases.prototype.listByInstanceNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByInstanceNextOperationSpec, callback);
    };
    return RecoverableManagedDatabases;
}());
export { RecoverableManagedDatabases };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByInstanceOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/recoverableDatabases",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RecoverableManagedDatabaseListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/recoverableDatabases/{recoverableDatabaseName}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.recoverableDatabaseName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion2
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RecoverableManagedDatabase
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByInstanceNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RecoverableManagedDatabaseListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
