// @azure/arm-sql/esm/operations/restorableDroppedManagedDatabases.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/restorableDroppedManagedDatabasesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a RestorableDroppedManagedDatabases. */
var RestorableDroppedManagedDatabases = /** @class */ (function () {
    /**
     * Create a RestorableDroppedManagedDatabases.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function RestorableDroppedManagedDatabases(client) {
        this.client = client;
    }
    RestorableDroppedManagedDatabases.prototype.listByInstance = function (resourceGroupName, managedInstanceName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            options: options
        }, listByInstanceOperationSpec, callback);
    };
    RestorableDroppedManagedDatabases.prototype.get = function (resourceGroupName, managedInstanceName, restorableDroppedDatabaseId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            managedInstanceName: managedInstanceName,
            restorableDroppedDatabaseId: restorableDroppedDatabaseId,
            options: options
        }, getOperationSpec, callback);
    };
    RestorableDroppedManagedDatabases.prototype.listByInstanceNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByInstanceNextOperationSpec, callback);
    };
    return RestorableDroppedManagedDatabases;
}());
export { RestorableDroppedManagedDatabases };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByInstanceOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/restorableDroppedDatabases",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RestorableDroppedManagedDatabaseListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/restorableDroppedDatabases/{restorableDroppedDatabaseId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.managedInstanceName,
        Parameters.restorableDroppedDatabaseId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RestorableDroppedManagedDatabase
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByInstanceNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RestorableDroppedManagedDatabaseListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
