// @azure/arm-sql/esm/operations/serverAzureADAdministrators.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/serverAzureADAdministratorsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ServerAzureADAdministrators. */
var ServerAzureADAdministrators = /** @class */ (function () {
    /**
     * Create a ServerAzureADAdministrators.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ServerAzureADAdministrators(client) {
        this.client = client;
    }
    /**
     * Creates a new Server Active Directory Administrator or updates an existing server Active
     * Directory Administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serverName The name of the server.
     * @param properties The required parameters for creating or updating an Active Directory
     * Administrator.
     * @param [options] The optional parameters
     * @returns Promise<Models.ServerAzureADAdministratorsCreateOrUpdateResponse>
     */
    ServerAzureADAdministrators.prototype.createOrUpdate = function (resourceGroupName, serverName, properties, options) {
        return this.beginCreateOrUpdate(resourceGroupName, serverName, properties, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    /**
     * Deletes an existing server Active Directory Administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serverName The name of the server.
     * @param [options] The optional parameters
     * @returns Promise<Models.ServerAzureADAdministratorsDeleteMethodResponse>
     */
    ServerAzureADAdministrators.prototype.deleteMethod = function (resourceGroupName, serverName, options) {
        return this.beginDeleteMethod(resourceGroupName, serverName, options)
            .then(function (lroPoller) { return lroPoller.pollUntilFinished(); });
    };
    ServerAzureADAdministrators.prototype.get = function (resourceGroupName, serverName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            options: options
        }, getOperationSpec, callback);
    };
    ServerAzureADAdministrators.prototype.listByServer = function (resourceGroupName, serverName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            options: options
        }, listByServerOperationSpec, callback);
    };
    /**
     * Creates a new Server Active Directory Administrator or updates an existing server Active
     * Directory Administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serverName The name of the server.
     * @param properties The required parameters for creating or updating an Active Directory
     * Administrator.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ServerAzureADAdministrators.prototype.beginCreateOrUpdate = function (resourceGroupName, serverName, properties, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            properties: properties,
            options: options
        }, beginCreateOrUpdateOperationSpec, options);
    };
    /**
     * Deletes an existing server Active Directory Administrator.
     * @param resourceGroupName The name of the resource group that contains the resource. You can
     * obtain this value from the Azure Resource Manager API or the portal.
     * @param serverName The name of the server.
     * @param [options] The optional parameters
     * @returns Promise<msRestAzure.LROPoller>
     */
    ServerAzureADAdministrators.prototype.beginDeleteMethod = function (resourceGroupName, serverName, options) {
        return this.client.sendLRORequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            options: options
        }, beginDeleteMethodOperationSpec, options);
    };
    return ServerAzureADAdministrators;
}());
export { ServerAzureADAdministrators };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/administrators/{administratorName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.administratorName0
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByServerOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/administrators",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ServerAdministratorListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginCreateOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/administrators/{administratorName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.administratorName0
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "properties",
        mapper: __assign(__assign({}, Mappers.ServerAzureADAdministrator), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        201: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        202: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var beginDeleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/administrators/{administratorName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.administratorName0
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        202: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        204: {
            bodyMapper: Mappers.ServerAzureADAdministrator
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
