// @azure/arm-sql/esm/operations/serverConnectionPolicies.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/serverConnectionPoliciesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ServerConnectionPolicies. */
var ServerConnectionPolicies = /** @class */ (function () {
    /**
     * Create a ServerConnectionPolicies.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ServerConnectionPolicies(client) {
        this.client = client;
    }
    ServerConnectionPolicies.prototype.createOrUpdate = function (resourceGroupName, serverName, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            parameters: parameters,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    ServerConnectionPolicies.prototype.get = function (resourceGroupName, serverName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            options: options
        }, getOperationSpec, callback);
    };
    return ServerConnectionPolicies;
}());
export { ServerConnectionPolicies };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/connectionPolicies/{connectionPolicyName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.connectionPolicyName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ServerConnectionPolicy), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ServerConnectionPolicy
        },
        201: {
            bodyMapper: Mappers.ServerConnectionPolicy
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/connectionPolicies/{connectionPolicyName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.connectionPolicyName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ServerConnectionPolicy
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
