// @azure/arm-sql/esm/operations/serverAutomaticTuningOperations.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/serverAutomaticTuningOperationsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ServerAutomaticTuningOperations. */
var ServerAutomaticTuningOperations = /** @class */ (function () {
    /**
     * Create a ServerAutomaticTuningOperations.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function ServerAutomaticTuningOperations(client) {
        this.client = client;
    }
    ServerAutomaticTuningOperations.prototype.get = function (resourceGroupName, serverName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            options: options
        }, getOperationSpec, callback);
    };
    ServerAutomaticTuningOperations.prototype.update = function (resourceGroupName, serverName, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            parameters: parameters,
            options: options
        }, updateOperationSpec, callback);
    };
    return ServerAutomaticTuningOperations;
}());
export { ServerAutomaticTuningOperations };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/automaticTuning/current",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ServerAutomaticTuning
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var updateOperationSpec = {
    httpMethod: "PATCH",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/automaticTuning/current",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion4
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.ServerAutomaticTuning), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.ServerAutomaticTuning
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
