// @azure/arm-sql/esm/operations/transparentDataEncryptionActivities.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/transparentDataEncryptionActivitiesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a TransparentDataEncryptionActivities. */
var TransparentDataEncryptionActivities = /** @class */ (function () {
    /**
     * Create a TransparentDataEncryptionActivities.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function TransparentDataEncryptionActivities(client) {
        this.client = client;
    }
    TransparentDataEncryptionActivities.prototype.listByConfiguration = function (resourceGroupName, serverName, databaseName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            databaseName: databaseName,
            options: options
        }, listByConfigurationOperationSpec, callback);
    };
    return TransparentDataEncryptionActivities;
}());
export { TransparentDataEncryptionActivities };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByConfigurationOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/transparentDataEncryption/{transparentDataEncryptionName}/operationResults",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.databaseName,
        Parameters.transparentDataEncryptionName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.TransparentDataEncryptionActivityListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
