// @azure/arm-sql/esm/operations/transparentDataEncryptions.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/transparentDataEncryptionsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a TransparentDataEncryptions. */
var TransparentDataEncryptions = /** @class */ (function () {
    /**
     * Create a TransparentDataEncryptions.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function TransparentDataEncryptions(client) {
        this.client = client;
    }
    TransparentDataEncryptions.prototype.createOrUpdate = function (resourceGroupName, serverName, databaseName, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            databaseName: databaseName,
            parameters: parameters,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    TransparentDataEncryptions.prototype.get = function (resourceGroupName, serverName, databaseName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serverName: serverName,
            databaseName: databaseName,
            options: options
        }, getOperationSpec, callback);
    };
    return TransparentDataEncryptions;
}());
export { TransparentDataEncryptions };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/transparentDataEncryption/{transparentDataEncryptionName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.databaseName,
        Parameters.transparentDataEncryptionName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.TransparentDataEncryption), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.TransparentDataEncryption
        },
        201: {
            bodyMapper: Mappers.TransparentDataEncryption
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/transparentDataEncryption/{transparentDataEncryptionName}",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.serverName,
        Parameters.databaseName,
        Parameters.transparentDataEncryptionName
    ],
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.TransparentDataEncryption
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
