// @azure/arm-sql/esm/operations/usages.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/usagesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a Usages. */
var Usages = /** @class */ (function () {
    /**
     * Create a Usages.
     * @param {SqlManagementClientContext} client Reference to the service client.
     */
    function Usages(client) {
        this.client = client;
    }
    Usages.prototype.listByInstancePool = function (resourceGroupName, instancePoolName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            instancePoolName: instancePoolName,
            options: options
        }, listByInstancePoolOperationSpec, callback);
    };
    Usages.prototype.listByInstancePoolNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByInstancePoolNextOperationSpec, callback);
    };
    return Usages;
}());
export { Usages };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByInstancePoolOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/instancePools/{instancePoolName}/usages",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.instancePoolName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.expandChildren,
        Parameters.apiVersion3
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.UsageListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listByInstancePoolNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.UsageListResult
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
