// @microsoft/teamsfx-core/build/plugins/resource/localdebug/constants.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalEnvBotKeys = exports.LocalEnvCertKeys = exports.LocalEnvAuthKeys = exports.LocalEnvBackendKeys = exports.LocalEnvFrontendKeys = exports.BotPlugin = exports.FrontendHostingPlugin = exports.AppStudioPlugin = exports.SolutionPlugin = exports.SpfxPlugin = exports.RuntimeConnectorPlugin = exports.FunctionPlugin = exports.AadPlugin = exports.LocalDebugConfigKeys = exports.ProgrammingLanguage = exports.LocalDebugCertificate = exports.LaunchBrowser = exports.LocalDebugPluginInfo = void 0;
class LocalDebugPluginInfo {
}
exports.LocalDebugPluginInfo = LocalDebugPluginInfo;
LocalDebugPluginInfo.pluginName = "fx-resource-local-debug";
LocalDebugPluginInfo.displayName = "LocalDebug Plugin";
class LaunchBrowser {
}
exports.LaunchBrowser = LaunchBrowser;
LaunchBrowser.chrome = "pwa-chrome";
LaunchBrowser.edge = "pwa-msedge";
class LocalDebugCertificate {
}
exports.LocalDebugCertificate = LocalDebugCertificate;
LocalDebugCertificate.CertFileName = "localhost.crt";
LocalDebugCertificate.KeyFileName = "localhost.key";
LocalDebugCertificate.FriendlyName = "TeamsFx Development Certificate";
var ProgrammingLanguage;
(function (ProgrammingLanguage) {
    ProgrammingLanguage["javascript"] = "javascript";
    ProgrammingLanguage["typescript"] = "typescript";
})(ProgrammingLanguage = exports.ProgrammingLanguage || (exports.ProgrammingLanguage = {}));
/**
 * Config key contract that value is provided by local debug plugin and required by other plugins.
 */
class LocalDebugConfigKeys {
}
exports.LocalDebugConfigKeys = LocalDebugConfigKeys;
LocalDebugConfigKeys.LocalAuthEndpoint = "localAuthEndpoint";
LocalDebugConfigKeys.LocalTabEndpoint = "localTabEndpoint";
LocalDebugConfigKeys.LocalTabDomain = "localTabDomain";
LocalDebugConfigKeys.TrustDevelopmentCertificate = "trustDevCert";
LocalDebugConfigKeys.LocalFunctionEndpoint = "localFunctionEndpoint";
LocalDebugConfigKeys.LocalBotEndpoint = "localBotEndpoint";
LocalDebugConfigKeys.LocalBotDomain = "localBotDomain";
class AadPlugin {
}
exports.AadPlugin = AadPlugin;
AadPlugin.Name = "fx-resource-aad-app-for-teams";
AadPlugin.LocalAppIdUri = "local_applicationIdUris";
AadPlugin.LocalClientId = "local_clientId";
AadPlugin.LocalClientSecret = "local_clientSecret";
AadPlugin.TeamsMobileDesktopAppId = "teamsMobileDesktopAppId";
AadPlugin.TeamsWebAppId = "teamsWebAppId";
class FunctionPlugin {
}
exports.FunctionPlugin = FunctionPlugin;
FunctionPlugin.FolderName = "api";
FunctionPlugin.Name = "fx-resource-function";
FunctionPlugin.DefaultFunctionName = "defaultFunctionName";
class RuntimeConnectorPlugin {
}
exports.RuntimeConnectorPlugin = RuntimeConnectorPlugin;
RuntimeConnectorPlugin.Name = "fx-resource-simple-auth";
RuntimeConnectorPlugin.FilePath = "filePath";
class SpfxPlugin {
}
exports.SpfxPlugin = SpfxPlugin;
SpfxPlugin.Name = "fx-resource-spfx";
class SolutionPlugin {
}
exports.SolutionPlugin = SolutionPlugin;
SolutionPlugin.Name = "solution";
// public static readonly SelectedPlugins: string = "selectedPlugins";
SolutionPlugin.LocalTeamsAppId = "localDebugTeamsAppId";
SolutionPlugin.RemoteTeamsAppId = "remoteTeamsAppId";
SolutionPlugin.TeamsAppTenantId = "teamsAppTenantId";
SolutionPlugin.ProgrammingLanguage = "programmingLanguage";
class AppStudioPlugin {
}
exports.AppStudioPlugin = AppStudioPlugin;
AppStudioPlugin.Name = "fx-resource-appstudio";
AppStudioPlugin.TeamsAppId = "teamsAppId";
class FrontendHostingPlugin {
}
exports.FrontendHostingPlugin = FrontendHostingPlugin;
FrontendHostingPlugin.FolderName = "tabs";
FrontendHostingPlugin.Name = "fx-resource-frontend-hosting";
class BotPlugin {
}
exports.BotPlugin = BotPlugin;
BotPlugin.FolderName = "bot";
BotPlugin.Name = "fx-resource-bot";
BotPlugin.LocalBotId = "localBotId";
BotPlugin.LocalBotPassword = "localBotPassword";
exports.LocalEnvFrontendKeys = Object.freeze({
    Browser: "FRONTEND_BROWSER",
    Https: "FRONTEND_HTTPS",
    TeamsFxEndpoint: "FRONTEND_REACT_APP_TEAMSFX_ENDPOINT",
    LoginUrl: "FRONTEND_REACT_APP_START_LOGIN_PAGE_URL",
    FuncEndpoint: "FRONTEND_REACT_APP_FUNC_ENDPOINT",
    FuncName: "FRONTEND_REACT_APP_FUNC_NAME",
    ClientId: "FRONTEND_REACT_APP_CLIENT_ID",
});
exports.LocalEnvBackendKeys = Object.freeze({
    WebJobsStorage: "BACKEND_AzureWebJobsStorage",
    FuncWorkerRuntime: "BACKEND_FUNCTIONS_WORKER_RUNTIME",
    AuthorityHost: "BACKEND_M365_AUTHORITY_HOST",
    TenantId: "BACKEND_M365_TENANT_ID",
    ClientId: "BACKEND_M365_CLIENT_ID",
    ClientSecret: "BACKEND_M365_CLIENT_SECRET",
    SqlEndpoint: "BACKEND_SQL_ENDPOINT",
    SqlDbName: "BACKEND_SQL_DATABASE_NAME",
    SqlUserName: "BACKEND_SQL_USER_NAME",
    SqlPassword: "BACKEND_SQL_PASSWORD",
    IdentityId: "BACKEND_IDENTITY_ID",
    ApiEndpoint: "BACKEND_API_ENDPOINT",
    ApplicationIdUri: "BACKEND_M365_APPLICATION_ID_URI",
    AllowedAppIds: "BACKEND_ALLOWED_APP_IDS",
});
exports.LocalEnvAuthKeys = Object.freeze({
    ClientId: "AUTH_CLIENT_ID",
    ClientSecret: "AUTH_CLIENT_SECRET",
    IdentifierUri: "AUTH_IDENTIFIER_URI",
    AadMetadataAddress: "AUTH_AAD_METADATA_ADDRESS",
    OauthAuthority: "AUTH_OAUTH_AUTHORITY",
    TabEndpoint: "AUTH_TAB_APP_ENDPOINT",
    AllowedAppIds: "AUTH_ALLOWED_APP_IDS",
    Urls: "AUTH_urls",
    ServicePath: "AUTH_SERVICE_PATH",
});
exports.LocalEnvCertKeys = Object.freeze({
    SslCrtFile: "FRONTEND_SSL_CRT_FILE",
    SslKeyFile: "FRONTEND_SSL_KEY_FILE",
});
exports.LocalEnvBotKeys = Object.freeze({
    BotId: "BOT_BOT_ID",
    BotPassword: "BOT_BOT_PASSWORD",
    ClientId: "BOT_M365_CLIENT_ID",
    ClientSecret: "BOT_M365_CLIENT_SECRET",
    TenantID: "BOT_M365_TENANT_ID",
    OauthAuthority: "BOT_M365_AUTHORITY_HOST",
    LoginEndpoint: "BOT_INITIATE_LOGIN_ENDPOINT",
    SqlEndpoint: "BOT_SQL_ENDPOINT",
    SqlDbName: "BOT_SQL_DATABASE_NAME",
    SqlUserName: "BOT_SQL_USER_NAME",
    SqlPassword: "BOT_SQL_PASSWORD",
    IdentityId: "BOT_IDENTITY_ID",
    ApiEndpoint: "BOT_API_ENDPOINT",
    ApplicationIdUri: "BOT_M365_APPLICATION_ID_URI",
});
