// @microsoft/teamsfx-core/build/plugins/resource/localdebug/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalDebugPlugin = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const typedi_1 = require("typedi");
const ResourcePluginContainer_1 = require("../../solution/fx-solution/ResourcePluginContainer");
require("./v2");
let LocalDebugPlugin = class LocalDebugPlugin {
    constructor() {
        this.name = "fx-resource-local-debug";
        this.displayName = "LocalDebug";
    }
    activate(solutionSettings) {
        return true;
    }
    async scaffold(ctx) {
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    }
    async localDebug(ctx) {
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    }
    async postLocalDebug(ctx) {
        return teamsfx_api_1.ok(teamsfx_api_1.Void);
    }
    async executeUserTask(func, ctx) {
        return teamsfx_api_1.ok(undefined);
    }
};
LocalDebugPlugin = tslib_1.__decorate([
    typedi_1.Service(ResourcePluginContainer_1.ResourcePlugins.LocalDebugPlugin)
], LocalDebugPlugin);
exports.LocalDebugPlugin = LocalDebugPlugin;
exports.default = new LocalDebugPlugin();
